---
title: "Genomics with R for biologists"
author: "Ines Lucia Patop: inespatop@brandeis.edu"
output: 
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true
    df_print: paged
    theme: united
    toc_float: true
    collapsed: false
---
```{r setup, include=FALSE}
#install.packages("tinytex")
#tinytex::install_tinytex() 
load("../macs2_analysis_p0.05/ChipSeq_ana.RData")
knitr::opts_chunk$set(echo = TRUE)
```

# Chip Sequencing Analysis

## Objectives

After this section you should be able to:

1. Explore results from ChipSeq data.
2. Modify tables and change gene names.
3. Use different ChipSeq gene annotation packages from R.
4. Find sequence motif enrichment in sequencing data.

## Introduction

__Chip-seq procedure:__

“ChIP-seq” is short for “chromatin immunoprecipitation (ChIP) followed by sequencing (seq)”. ChipSeq data is basically DNA sequencing data in which the DNA is sequenced AFTER immunoprecipitating a protein of interest. Therefore, we expect to sequence the pieces of DNA that are interacting with the precipitated protein.

If the protein of interest is a transcription factor (TF) for example, we expect to get the regions of DNA close to the genes that are being regulated by this particular TF (ie. the promoters of the genes). If we immunoprecipitate RNA polymerase 2, we expect to get the regions of the DNA that are being transcribed to RNA. 

But, how do we know if certain region of the genome is really interreacting with the protein of interest? As in any analysis, it is important to have a control to see the probability to have a signal in a particular region just by chance. Usually as a control we use the DNA sequencing BEFORE immunoprecipitation (IP). This data is usually called input (INP), as it is the input material used for the IP. What we are looking then will be "enrichment" of each region of the genome in the IP compared with the INP.

The biochemical procedure usually goes as follows:

```{r, out.width = "500px", echo=FALSE, fig.align="center", fig.cap="Commonly used ChipSeq procedure. DNA is fragmented followed by immunoprecipitation for the protein of interest. Finally, DNA library is prepared and sequenced. As you can see there is an enrichment of the DNA regions bonded to the protein of interest."}
knitr::include_graphics("../macs2_analysis_p0.05/chipseqproc.png")
```

To study this type of data we will re-analyze the data generated in [Meireles-Filho  et al 2014](https://www.cell.com/current-biology/fulltext/S0960-9822(13)01403-6?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0960982213014036%3Fshowall%3Dtrue1).

In this work the authors study genome-wide the binding of two transcription factors (Clk and Cyc) in the fruit fly (*Drosophila melanogaster*). To do so they do Chip sequencing.

These two transcription factors are core components of the circadian clock in animals. Just as an introduction these two transcription factors bind to specific regions on the genome and activate transcription. Clk oscilates over the day due to a set of feedback loops with other core-clock components. These set of molecular interactions generate an autonomous oscilating molecular clock with a period of almost 24 hr. This clock can be re-set by light input. We will study RNA oscilations over the day in next chapter.

```{r, out.width = "500px", echo=FALSE, fig.align="center",fig.cap="Circadian clock at the molecular level. The molecular feedback loop is formed by the negative feedback of Period (PER) and Timeless (TIM) on their own transcription. Figure adapted from Dubowy et al 2017"}
knitr::include_graphics("../macs2_analysis_p0.05/clkcyc.png")
```

```{r, out.width = "500px", echo=FALSE, fig.align="center",fig.cap="Cyc, Clk and Time levels over the day. Figure adapted from Alves Meireles-Filho et al 2013"}
knitr::include_graphics("../macs2_analysis_p0.05/clkcycosc.png")
```

```{r, out.width = "500px", echo=FALSE, fig.align="center",fig.cap="Light control of the circadian clock. Figure adapted from Alves Meireles-Filho et al 2013"}
knitr::include_graphics("../macs2_analysis_p0.05/clkcycosc.png")
```


## Data pre-processing (peak-calling): 

Our goal is then to quantify signal enrichment in certain part of the genome in IP vs INP sequencing. This is achieved by comparing the ALIGNEMENT of the reads of IP vs INP. 

This process is called “peak calling” because it is trying to determine the “pileup” of reads along the genome forming "peaks". As any peak the algorithm used will report high of the peak, summit location, width and finally a pvalue corresponding to the comparison of the signal in that region in IP vs INP.

Here we can see how the pileup reads look like in IGV. We are looking here at the promoter of the genes *tim*, a known target of CLK protein. As you can see, the peak is clear in the IP comapred with the INP.

```{r, out.width = "900px", echo=FALSE, fig.align="center",fig.cap="IGV pileup traks over tim. Upper line IP, lower line INP."}
knitr::include_graphics("../macs2_analysis_p0.05/IGVtim.png")
```

To do this analysis genome wide, we need a tool that can do this for all the genom. There are few tools available, we use here [MACS](https://github.com/taoliu/MACS) algorithm to assess the peaks from the sequencing data.

Basically, this analysis relies in the aligned data (.bam files). It compares the IP with the INP control. 

For this analysis I aligned the data from [GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40467) with [Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml), an aligner that is used for DNA data (not splice aware). 

Finally ran the MACS command in the **terminal** as follows:

```
macs2 callpeak -t head_clk_ip_rep1.sorted.bam head_clk_ip_rep2.sorted.bam  -c ./head_clk_input_rep1.sorted.bam ./head_clk_input_rep2.sorted.bam -g dm -n PeakAna_clk_Ip_vs_INP -B -p 0.05
```
Where, -t indicates that the following arguments are the IP samples, -c the control samples, -g the genome name in this case "dm", -n the name that all the output files will have, -B indicates to the program to store visualization files (in [.bdg](https://software.broadinstitute.org/software/igv/bedgraph) format usefull to visualize in IGV).

The **output** after running MACS on the shell:

|File name | Description | 
|----------|-------------------------------------------|
|PeakAna_clk_Ip_vs_INP_model.r | An R script for producing a PDF illustrating the peak model |
|PeakAna_clk_Ip_vs_INP_model.pdf |The PDF image of the read distribution in model peaks and fragment size estimation |
|PeakAna_clk_Ip_vs_INP_peaks.xls | Key parameters used by MACS and detailed information of every peak identified by MACS|
|PeakAna_clk_Ip_vs_INP_peaks.bed | Peak locations in BED format |
|PeakAna_clk_Ip_vs_INP_peaks.subpeaks.bed | Subpeak locations in BED-like format. This file is generated by PeakSplitter, which is called by MACS |
|PeakAna_clk_Ip_vs_INP_summits.bed | Summit locations of the peaks in BED format |
|PeakAna_clk_Ip_vs_INP_MACS_bedGraph | Directory where the BedGraph files are generated. For each control or ChIP-seq sample, a BedGraph file describes the read distribution along the whole genome |

## Exploring the results:

We will see the first part of the output we get in the "NAME_peaks.xls":

We can see here that we have all the columns we expect and we cqn try to see places we know are interesting for us.


```{r }
#knitr::include_graphics('images/hex-rmarkdown.png')
macs.res<-read.table("../macs2_analysis_p0.05/PeakAna_clk_Ip_vs_INP_peaks_bed.txt",header = T)
head(macs.res)
```

## Annotating the peaks.

With this information we want to be able to conclude something about this transcription factor regulation. 

The first thing to see is genes that might be regulated. This is basically looking which is the closest gene to each peak. 

To do this we will just looking at the annotation file that have the information of which gene is in each region of the genome (we can think of this as the "map" of the genome).

>What problems you can think that this approach might have? Think about enhancers for example.

For this we will use the following libraries: 

```{r, message=FALSE,warning=FALSE}
library(GenomicFeatures) #This is an R package to deal with genomic data. You can read more at https://bioconductor.org/packages/release/bioc/vignettes/GenomicFeatures/inst/doc/GenomicFeatures.pdf
library(ChIPpeakAnno) #This is a package to annotate Chip-Peaks data https://bioconductor.org/packages/release/bioc/vignettes/ChIPpeakAnno/inst/doc/ChIPpeakAnno.html
```

### Read the annotation file 

In class we saw what were annotation files. Those were files that contained the "map" of the genome. This map shows us where each gene start and end and where each exon, intron and UTR is in that gene.

We will work with a GTF file. Remember that annotation files can be in many formats. They can be downloaded from different places as UCSC genome browser, Ensemble, etc. 

To read the GTF file we will use the `makeTxDbFromGFF` function. We are doing this because it will allow us to change it to another format, sqlite. 


```{r message=FALSE,warning=FALSE}
txdb <- makeTxDbFromGFF(file="../macs2_analysis_p0.05/annot2.chr.gtf", format = "gtf", dataSource="dm6", organism="Drosophila melanogaster") #this funciton is actually reading the GTF file and creating a data base. Try reading the help documentation of this funciton.

#Now we need to save this as an sqlite:
saveDb(txdb, file="../macs2_analysis_p0.05/dm6.sqlite") #Look at your working directory, what happened?
txdb <- loadDb("../macs2_analysis_p0.05/dm6.sqlite") #this is loading the sqlite directly without creating the database
```

### Get the genes from this annotation data base
The function `genes` actually extract the genomic ranges (chr start, end, strand, etc) from the genes in the annotation data base we are working with `txdb`.

```{r message=FALSE,warning=FALSE}
ge <- genes(txdb, columns=c("tx_name", "gene_id", "tx_type")) #get the genes genomic ranges, what do you think are the options we are passing to this function? (read the help manual to check).

#lets look at the genes:
as.data.frame(head(ge))
```

### Get the genomic ranges of the peaks from MACS and find which gene is closer to them (ie. Annotate them)
Now we will use the function `GRanges` to actually get the genomic ranges of the peaks from MACS. And we will annotate them with the function `annotatedPeak` and the `genes` function to map them to the closer gene.

Basically we are using the information about the genes (start, end, etc) to see which is the closest one to each peak.


```{r message=FALSE,warning=FALSE}
peaksGR<-GRanges(macs.res) #create the GRanges object of the macs-table.
annotatedPeak <- annotatePeakInBatch(peaksGR, AnnotationData=genes(txdb))
#look at the begining of the table
as.data.frame(head(annotatedPeak))
```

### Change the gene format.

What do you notice about the table annotatedPeak? Can you recognize any of the genes? Try to google some of them. 

The gene names are now in a format that is general for fly genes. This format is the one we have from flyblase.org, a database for fly genes. These names are useful for databases purposes but are not indicative of their function. Therefore, we want now to put the gene-names that we all know (no the Fb...), to do so we need to use the following packages:


```{r message=FALSE,warning=FALSE}
library(AnnotationDbi) #annotation data base package: https://bioconductor.org/packages/release/bioc/vignettes/AnnotationDbi/inst/doc/IntroToAnnotationPackages.pdf
library(org.Dm.eg.db) #Drosophila melanogaster (Dm) data base
library(dplyr) #this package allows us to manipulate better the data.frames
```

#### Get the names of the genes

What we need now is to build some kind of dictionary that will be able to translate between both name types: FB to Symbol.

So, we first create an object with the `GeneNames` in both types. To do so we will use the `select` function. 

As you will notice if you try the first line of comand `select(org.Dm.eg.db, c(as.character(annotatedPeak$feature)), "SYMBOL", keytype = "ENSEMBL")`, there is a problem with this function. Basically this comes from the fact that both `AnnotationDbi` and `dplyr` have a function named `select`. R will call the function from the last package executed. To tell R which package we want to use we have to write: `Package::function`, in this case: `AnnotationDbi::select`.

Other solutions to this "function name collision" issue are:

1. execute the package from which we want to use the conflicting function in the last place. 
2. Rename the conflicting function after running the first package. In this case:

`library(AnnotationDbi) ##annotation data base package`
`library(org.Dm.eg.db) ##Drosophila melanogaster (Dm) data base`
`dm.select=select`
`library(dplyr)`

```{r }
GeneNames = AnnotationDbi::select(org.Dm.eg.db, c(as.character(annotatedPeak$feature)), "SYMBOL", keytype = "ENSEMBL")
GeneNames$feature=GeneNames$ENSEMBL
```
>How does this looks like?

```{r }
head(GeneNames)
```
```{r }
class(GeneNames)
```

```{r }
GeneNames=GeneNames[,-1] #what is this doing? 
```

To see what happened we can look at the object again after that
```{r }
head(GeneNames)
```

#### Create a data frame with the peak annotation and the NEW names
We will then put the names into the first table. To this end we will use `dplyr` package function by doing a `left_join`. This is actually a type of merging in which we keep everything that is in the first object. Why do you think we do this? 

We can explore the join options, here are some graphic explanation, and if you want to know more you can go [to the manual](https://www.rdocumentation.org/packages/dplyr/versions/0.7.8/topics/join) and try the examples. Also, I would recommend you to read the [cheatsheet](https://rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf).


```{r, out.width = "300px", echo=FALSE, fig.align="center", fig.cap="Different joint options from dplyr. Adapted from cheatsheet."}
knitr::include_graphics("../macs2_analysis_p0.05/dplyr-joins.png")
```

```{r }
class(annotatedPeak) # this tells us the class, which type is it?
annotatedPeak_df=as.data.frame(annotatedPeak) # we will change it to data.frame to have an object we can manipulate better and write to a table
annotatedPeak_df=left_join(annotatedPeak_df,GeneNames, by="feature") #What is this doing?
```

### Explore and export the table
We can now see how the table looks like and export it as a .txt table so we can explore it outside R.

```{r }
as.data.frame(head(annotatedPeak_df))#how it looks?
write.table(annotatedPeak_df,file = "../macs2_analysis_p0.05/ChipPeakAnno_results.xls",row.names = T,col.names = T,sep="\t") #this is writing the table, look in your directory and try opening it with excel.
```

## Annotate with `ChIPseeker`
We are going to use now [ChIPseeker](https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html), another chip-seq annotation package. We need to do again many things, like extracting the Gene-names. Luckily we already did part of the work.

```{r message=FALSE,warning=FALSE}
library(ChIPseeker)

#we will use the Genomic Ranges object created previously and the txdb also
peakAnno <- annotatePeak(peaksGR, TxDb=txdb, verbose=FALSE)
peakAnno_df <- as.data.frame(peakAnno)

GeneNames = AnnotationDbi::select(org.Dm.eg.db, c(as.character(peakAnno_df$geneId)), "SYMBOL", keytype = "ENSEMBL")
GeneNames$geneId=GeneNames$ENSEMBL
head(GeneNames)
```

```{r message=FALSE,warning=FALSE}
GeneNames=GeneNames[,-1]

peakAnno_df=peakAnno_df %>% left_join(GeneNames, by="geneId") #this is merging the two tables by the column Geneld
```

### Explore and Export the output
If we look at the head of the file we can see that we have a __different annotation format__ for the peaks and that we actually have a distance to the TSS.

We will save this file as well. 

```{r message=FALSE,warning=FALSE}
head(peakAnno_df)
write.table(peakAnno_df,file = "ChIPseekerAnno_results.xls",row.names = T,col.names = T,sep="\t") #try looking at it on excel or any other spreadsheet editor, what are the differences you see with the previous one?
```

### Positive controls. 

We can now explore the data back at [IGV](https://igv.org/app/) and see if the genes we get a significantly enriched for this TF are actually enriched.

Lets think about some controls we can do with the data. Which is the protein we are analyzing? Is there any gene we already know that protein might be regulating?

Well, yes. We are analyzing __Clk__, that happens to regulate __tim__.

So, making it more general: Other option to see if our analysis and the data in general make sense is to go to previous literature and see what genes we know that are regulated by this TF. Then, we expect them to be enriched in our analysis.


```{r }
peakAnno_df_tim=peakAnno_df[which(peakAnno_df$SYMBOL=="tim"),] #What we are doing here?, remember Lab 2!
peakAnno_df_tim=peakAnno_df_tim[order(peakAnno_df_tim$X.log10.pvalue.,decreasing = T),] # we order them by log10(pval), what that really means?

as.data.frame(head(peakAnno_df_tim))
```

### Plots

This package has some useful plot functions so we can explore the results more easily. 

First we can see the peaks over the chromosomes. For that we use the function `covplot`. Its name comes from "coverage plot".

What do you think about the results? Do they make sense?

```{r message=FALSE,warning=FALSE, fig.cap="Coverage plot."}
covplot(peaksGR, weightCol="X.log10.pvalue.",chrs=c("chr2L","chr2R","chr3L","chr3R","chr4","chrX","chrY"))
```

And then a general distribution of the peaks in relation with he transcription start site (TSS). For that we use the function `peakHeatmap` and `plotAvgProf2`.

```{r message=FALSE,warning=FALSE,fig.cap="Peak heatmap."}
peakHeatmap(peaksGR, TxDb=txdb, upstream=1000, downstream=1000, color="darkviolet")
```

```{r message=FALSE,warning=FALSE,fig.cap="Average coverage plot."}
plotAvgProf2(peaksGR, TxDb=txdb, upstream=1000, downstream=1000, xlab="Genomic Region (5'->3')", ylab = "Read Count Frequency") #Try exploring the options of this function. 
```

## Sequence Motif analysis

Another goal now is to analyze this peaks and see a pattern enrichment and see if we can find a __Binding Motif__ for Clk in fly head.

This is basically DNA analysis, so we will have to figure out the regular DNA motifs and see if there is any specific enrichment.

For this we will use the following libraries: 

```{r message=FALSE,warning=FALSE}
library(Biostrings); library(seqLogo); library(BCRANK); library(gridExtra)
```

### Put the peak sequences in fasta format

To parse the corresponding sequences from the reference genome, the `getSeq` function from the Biostrings package can be used. The following example parses the sequences for each peak set and saves the results to separate FASTA files, one for each peak set. In addition, the sequences in the FASTA files are ranked (sorted) by increasing p-values as expected by some motif discovery tools, such as BCRANK.

```{r message=FALSE,warning=FALSE}
library("Rsamtools") #we will use the Rsamtools package for this.

peaksGR <- peaksGR[order(peaksGR$X.log10.pvalue., decreasing=TRUE)] #order the peaks according to pvalue.

as.data.frame(head(peaksGR))
```

```{r message=FALSE,warning=FALSE}
pseq <- getSeq(FaFile(file = "../macs2_analysis_p0.05/dm6.fa"), peaksGR) #extract the sequence of the peaks using the genome fasta file dm6.fa and the genome-ranges in the object peaksGR
```

We can now see how the `pseq` object looks like.

```{r message=FALSE,warning=FALSE}
as.data.frame(head(pseq))
```

```{r message=FALSE,warning=FALSE}
writeXStringSet(pseq, filepath = "peaksGR.fasta") #use the names and save the peaks in fasta format, how this looks like? open it outside R! you can shange the filepath to save it any place you want.
```

### Find the MOTIFS

The Bioconductor package [BCRANK](https://www.bioconductor.org/packages//2.12/bioc/vignettes/BCRANK/inst/doc/BCRANK.pdf) is one of the many available for de novo discovery of DNA binding motifs in peak regions of ChIP-Seq. 

The main function we are using here is bcrank. when looking at the documentation we see that: 

"BCRANK uses a heuristic search strategy. First a score is computed for an initial short consensus sequence, typically selected at random. The score considers both the number of consensus occurrences and the rank of the genomic regions. Then all consensus sequences in a neighborhood of the start guess are evaluated and the one with highest score is kept as the starting point for the next iteration. When a local optimum is found, the algorithm is terminated and the locally optimal consensus is reported as a result. In order to increase the chance of detecting the globally optimal solution, the algorithm may be restarted several times using different random starting points. Alternatively, BCRANK can be used for assigning scores to previously established consensus sequences. The sections below describe in more detail how the neighborhood, scoring function and start guess are implemented."

We will use 25 restarts and a penalty for both repetitive sequences and non-canonical letters. 

 P1 - Penalty on non-specific bases. Let l be the length of the consensus sequence and b the
total number of fixed bases (A, C, G, T) in the sequence. If there are no fixed bases, b is set
to 0.5. The penalty is then defined as P1 = b/l.
 P2 - Penalty on repetitive motifs. Let rn, n ∈ 1, 2 be the number of input DNA regions that
contain at least n occurrences of the consensus. Then P2 = 1 − (r2/r1).

```
set.seed(0)
BCRANKout <- bcrank("peaksGR.fasta", restarts=25, use.P1=TRUE, use.P2=TRUE)
```

```{r message=FALSE,warning=FALSE,include=FALSE,echo=FALSE}
#set.seed(0)
#BCRANKout <- bcrank("peaksGR.fasta", restarts=5, use.P1=TRUE, use.P2=TRUE)
```

### Explore the results

Then we can also see the results, we will se the results that are in the top.

We can first explore the results: the matrix shows the proportion of each letter in each possition for thar specific motif.
 
```{r }
head(toptable(BCRANKout))
a<-BCRANKout@toplist
a[[1]]
```

### Plot the results
We can play around with the matrixes and plot them in the known DNA-Logo standard. To do this we will first store the most top motif from `toptable`. Then we will use the `pwa` function from BCRANK and the `seqLogo` function from seqLogo package.

```{r ,fig.cap="Most significant sequence logo."}
topMotif <- toptable(BCRANKout, i=1)
weightMatrix <- pwm(topMotif, normalize = FALSE)
weightMatrixNormalized <- pwm(topMotif, normalize = TRUE)
seqLogo(weightMatrixNormalized,ic.scale = T)
```

We can now do the same with a different function just to show you that you can do the same things with different approaches. This is using function form the `seqLogo` package. 

```{r ,fig.cap="Most significant sequence logo."}
p<-makePWM(as.data.frame(weightMatrixNormalized))
p@consensus
seqLogo(p)
```
We will create a table to be able to analyze the data
```{r }
topt<-as.data.frame(toptable(BCRANKout)) #extract the results table
head(topt)
```

### Plotting with Ggplot2
Now with ggplot we will use a package called ggseqlogo. 
This will be nice to learn how to install packages from github (you have to use the package devtools and then the function install_github).
http://www.bioconductor.org/packages//2.11/bioc/vignettes/seqLogo/inst/doc/seqLogo.pdf 

```{r }
library(ggplot2)
#install.packages("devtools")
library("devtools") #this library is literally called "developers tools" and allow us to do more complex things as downloading packages from github
#install_github("omarwagih/ggseqlogo")
library(ggseqlogo)
```

#### Create the matrixes for ggplot2
We will first store all the matrixes in a list so we can then just plot them all together and not only one by one.

```{r message=FALSE,warning=FALSE}
# Some example DNA sequences, lets try to go together over what is this loop doing
seqs_list=list() #first we create an empty list that we will fill up over the loop
for(e in 1:16){ #we will loop 16 times
  #print(e) #we will print every loop
  topMotif <- toptable(BCRANKout, i=e) #we get the motif number e (ie, the iteration will let us go from top motif 1 to 16)
weightMatrixNormalized <- pwm(topMotif, normalize = TRUE) #do some process with the motif
seqs_list[[e]]<-weightMatrixNormalized #add the processed motif to the list in position e (ie. the loop allows us to fill up the list)
}
```

#### Plot
We can now, either plot one by one for each position in the list. 

```{r,fig.cap="Most significant sequence logo in probability and percentage of information." }
# Get first set of sequences
seqs_dna = seqs_list[[1]]

# Plot a sequence logo with the 2 different methods available
p1 = ggseqlogo( seqs_dna, method = 'prob' )
p2 = ggseqlogo( seqs_dna, method = 'bits' ) +labs(y="Information Content") #this changed the name of the `y` axes
grid.arrange(p1, p2)

````

Or we can plot all together using the facet function to actually separate them.

```{r ,fig.cap="Many significant sequence logo in probability and percentage of information."}
#Plot all the first 12 sequences
ggplot() + geom_logo(seqs_list,method = 'prob') + theme_logo() + 
  facet_wrap(~seq_group, ncol=4)


```

## Extra: How to solve some common annotation issues.  

Lets see some regular problem and how to solve it.

I will show you first a regular problem we can have: annotation files do not correlate with the one previously used for annotation.


```{r message=FALSE,warning=FALSE}
txdb <- makeTxDbFromGFF(file="../macs2_analysis_p0.05/Drosophila_melanogaster.BDGP6.92.gtf", format = "gtf", dataSource="dm6", organism="Drosophila melanogaster")
saveDb(txdb, file="./dm6.sqlite")
txdb <- loadDb("./dm6.sqlite")
class(txdb)
```
Now we can actually extract the gene location from the annotation that we created
```{r message=FALSE}
ge <- genes(txdb, columns=c("tx_name", "gene_id", "tx_type")) #get the genes genomic ranges
as.data.frame(head(ge))
```
Now we will use the function `annotatePeakInBach`
```{r message=FALSE,warning=FALSE}
?annotatePeakInBatch() #to undestand what are the features we need
peaksGR<-GRanges(macs.res) # create the GRanges object of the macs-table 
annotatedPeak <- annotatePeakInBatch(peaksGR, AnnotationData=genes(txdb))
head(annotatedPeak)
```
Here we see that the peaks have a lot of NAs so we did something wrong. If we compare the `ge` object and the annotated peaks object we can see the problem.
```{r }
as.data.frame(head(ge))
as.data.frame(head(annotatedPeak))

```
## Extra: Detailed explanation of the MACS output files

> NAME_peaks.xls is a tabular file which contains information about called peaks. You can open it in excel and sort/filter using excel functions. Information include:

```
  chromosome name
  start position of peak
  end position of peak
  length of peak region
  absolute peak summit position
  pileup height at peak summit, -log10(pvalue) for the peak summit (e.g. pvalue =1e-10, then this value should be 10)
  fold enrichment for this peak summit against random Poisson distribution with local lambda, -log10(qvalue) at peak summit
  Coordinates in XLS is 1-based which is different with BED format.
```

> NAME_peaks.narrowPeak is BED6+4 format file which contains the peak locations together with peak summit, pvalue and qvalue. You can load it to UCSC genome browser. Definition of some specific columns are:

```
  5th: integer score for display calculated as int(-10*log10qvalue). Please note that currently this value might be out of the [0-1000] range defined in UCSC Encode narrowPeak [format](https://genome.ucsc.edu/FAQ/FAQformat.html##format12)
  7th: fold-change
  8th: -log10pvalue
  9th: -log10qvalue
  10th: relative summit position to peak start
  The file can be loaded directly to UCSC genome browser. Remove the beginning track line if you want to analyze it by other tools.
```

> NAME_summits.bed is in BED format, which contains the peak summits locations for every peaks. The 5th column in this file is -log10pvalue the same as NAME_peaks.bed. If you want to find the motifs at the binding sites, this file is recommended. The file can be loaded directly to UCSC genome browser. Remove the beginning track line if you want to analyze it by other tools.

> NAME_peaks.broadPeak is in BED6+3 format which is similar to narrowPeak file, except for missing the 10th column for annotating peak summits.

> NAME_peaks.gappedPeak is in BED12+3 format which contains both the broad region and narrow peaks. The 5th column is 10-log10qvalue, to be more compatible to show grey levels on UCSC browser. Tht 7th is the start of the first narrow peak in the region, and the 8th column is the end. The 9th column should be RGB color key, however, we keep 0 here to use the default color, so change it if you want. The 10th column tells how many blocks including the starting 1bp and ending 1bp of broad regions. The 11th column shows the length of each blocks, and 12th for the starts of each blocks. 13th: fold-change, 14th: -log10pvalue, 15th: -log10qvalue. The file can be loaded directly to UCSC genome browser.

> NAME_model.r is an R script which you can use to produce a PDF image about the model based on your data. Load it to R by:

`$ Rscript NAME_model.r`

Then a pdf file NAME_model.pdf will be generated in your current directory. Note, R is required to draw this figure.

> The .bdg files are in bedGraph format which can be imported to UCSC genome browser or be converted into even smaller bigWig files. There are two kinds of bdg files: treat_pileup, and control_lambda.

## Activity

> Slect one gene of interest and find the enrichment score. 
> Load the aligned data in IGV and look at the peaks. Data is available in the book repository.
> Do the same analysis in, Cyc, the other protein of interest in the paper. Data is available in the book repository. 

## Resources and Bibliography

Dubowy C, Sehgal A. Circadian Rhythms and Sleep in Drosophila melanogaster. Genetics. 2017;205(4):1373-1397. doi:10.1534/genetics.115.185157

MEIRELES-FILHO, Antonio Carlos Alves  and  KYRIACOU, Charalambos Panayiotis. Circadian rhythms in insect disease vectors. Mem. Inst. Oswaldo Cruz [online]. 2013, vol.108, suppl.1 [cited  2020-07-08], pp.48-58.

A.C.A. Meireles-Filho, A.F. Bardet, J.O. Yáñez-Cuna, G. Stampfel, A. Stark
cis-regulatory requirements for tissue-specific programs of the circadian clock
Curr Biol, 24 (2014), pp. 1-10

Zhang, Y., Liu, T., Meyer, C.A. et al. Model-based Analysis of ChIP-Seq (MACS). Genome Biol 9, R137 (2008). https://doi.org/10.1186/gb-2008-9-9-r137

Feng, J., Liu, T., Qin, B. et al. Identifying ChIP-seq enrichment using MACS. Nat Protoc 7, 1728–1740 (2012). https://doi.org/10.1038/nprot.2012.101

https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html

https://www.bioconductor.org/packages/release/bioc/vignettes/ChIPpeakAnno/inst/doc/pipeline.html

## Session info: all the packages installed.

This is useful when we want to de-bug a code or share a code or a problem with someone. 

```{r message=FALSE,warning=FALSE}
toLatex(sessionInfo())
````