<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Lab 2 Data manipulation | Genomics with R for biologists</title>
  <meta name="description" content="This is the index for the coursebook" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="Lab 2 Data manipulation | Genomics with R for biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is the index for the coursebook" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Lab 2 Data manipulation | Genomics with R for biologists" />
  
  <meta name="twitter:description" content="This is the index for the coursebook" />
  

<meta name="author" content="Ines Lucia Patop inespatop@brandeis.edu" />


<meta name="date" content="2020-08-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction-to-r-and-rstudio.html"/>
<link rel="next" href="visualization-and-statistical-test.html"/>
<script src="libs/header-attrs-2.3/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="include/psyteachr.css" type="text/css" />
<link rel="stylesheet" href="include/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Genomics with R for biologists</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Overview</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#course-aims"><i class="fa fa-check"></i><b>0.1</b> Course Aims</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#intended-learning-outcomes"><i class="fa fa-check"></i><b>0.2</b> Intended Learning Outcomes</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#next-steps"><i class="fa fa-check"></i><b>0.3</b> Next steps</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#data-availability"><i class="fa fa-check"></i><b>0.4</b> Data availability</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#acknowledgments"><i class="fa fa-check"></i><b>0.5</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html"><i class="fa fa-check"></i><b>1</b> Introduction to R and RStudio</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>1.1</b> Objectives</a></li>
<li class="chapter" data-level="1.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>1.2</b> Introduction</a></li>
<li class="chapter" data-level="1.3" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#rstudio-world"><i class="fa fa-check"></i><b>1.3</b> RStudio world</a></li>
<li class="chapter" data-level="1.4" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#create-a-project"><i class="fa fa-check"></i><b>1.4</b> Create a project</a></li>
<li class="chapter" data-level="1.5" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#useful-shortcuts-in-rstudio"><i class="fa fa-check"></i><b>1.5</b> Useful shortcuts in RStudio:</a></li>
<li class="chapter" data-level="1.6" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#installing-and-loading-packages"><i class="fa fa-check"></i><b>1.6</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.7" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#creating-objects"><i class="fa fa-check"></i><b>1.7</b> Creating objects</a></li>
<li class="chapter" data-level="1.8" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#r-as-a-calculator"><i class="fa fa-check"></i><b>1.8</b> R as a calculator</a></li>
<li class="chapter" data-level="1.9" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#functions"><i class="fa fa-check"></i><b>1.9</b> Functions</a></li>
<li class="chapter" data-level="1.10" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#loops"><i class="fa fa-check"></i><b>1.10</b> Loops</a></li>
<li class="chapter" data-level="1.11" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#conditions"><i class="fa fa-check"></i><b>1.11</b> Conditions</a></li>
<li class="chapter" data-level="1.12" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#character-objects"><i class="fa fa-check"></i><b>1.12</b> Character objects</a></li>
<li class="chapter" data-level="1.13" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#identify"><i class="fa fa-check"></i><b>1.13</b> Identify</a></li>
<li class="chapter" data-level="1.14" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#modify"><i class="fa fa-check"></i><b>1.14</b> Modify</a></li>
<li class="chapter" data-level="1.15" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#useful-resources"><i class="fa fa-check"></i><b>1.15</b> Useful resources</a></li>
<li class="chapter" data-level="1.16" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#activity"><i class="fa fa-check"></i><b>1.16</b> Activity:</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-manipulation.html"><a href="data-manipulation.html"><i class="fa fa-check"></i><b>2</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>2.1</b> Objectives</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>2.2</b> Introduction</a></li>
<li class="chapter" data-level="2.3" data-path="data-manipulation.html"><a href="data-manipulation.html#load-data"><i class="fa fa-check"></i><b>2.3</b> Load data</a></li>
<li class="chapter" data-level="2.4" data-path="data-manipulation.html"><a href="data-manipulation.html#data-exploration"><i class="fa fa-check"></i><b>2.4</b> Data exploration</a></li>
<li class="chapter" data-level="2.5" data-path="data-manipulation.html"><a href="data-manipulation.html#subsetting"><i class="fa fa-check"></i><b>2.5</b> Subsetting</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#activity"><i class="fa fa-check"></i><b>2.6</b> Activity:</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html"><i class="fa fa-check"></i><b>3</b> Visualization and Statistical test</a>
<ul>
<li class="chapter" data-level="3.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#objectives-1"><i class="fa fa-check"></i><b>3.1</b> Objectives</a></li>
<li class="chapter" data-level="3.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#introduction-1"><i class="fa fa-check"></i><b>3.2</b> Introduction</a></li>
<li class="chapter" data-level="3.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#plots"><i class="fa fa-check"></i><b>3.3</b> Plots</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#line-and-points"><i class="fa fa-check"></i><b>3.3.1</b> Line and points</a></li>
<li class="chapter" data-level="3.3.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#historgam-and-density-plots"><i class="fa fa-check"></i><b>3.3.2</b> Historgam and density plots</a></li>
<li class="chapter" data-level="3.3.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#boxplot"><i class="fa fa-check"></i><b>3.3.3</b> Boxplot</a></li>
<li class="chapter" data-level="3.3.4" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#saving-plots"><i class="fa fa-check"></i><b>3.3.4</b> Saving plots</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#statistical-test"><i class="fa fa-check"></i><b>3.4</b> Statistical test</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#descriptive-statistics"><i class="fa fa-check"></i><b>3.4.1</b> Descriptive statistics</a></li>
<li class="chapter" data-level="3.4.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#t-testwilcoxon"><i class="fa fa-check"></i><b>3.4.2</b> T-test/Wilcoxon</a></li>
<li class="chapter" data-level="3.4.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#anovakruskalwallis"><i class="fa fa-check"></i><b>3.4.3</b> Anova/Kruskalâ€“Wallis</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#activity-1"><i class="fa fa-check"></i><b>3.5</b> Activity:</a></li>
<li class="chapter" data-level="3.6" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#resources"><i class="fa fa-check"></i><b>3.6</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html"><i class="fa fa-check"></i><b>4</b> Sequencing techniques and preprocessing</a>
<ul>
<li class="chapter" data-level="4.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>4.1</b> Objectives</a></li>
<li class="chapter" data-level="4.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#different-type-of-data-dna-rna-splicing-single-end-paired-end"><i class="fa fa-check"></i><b>4.3</b> Different type of data (DNA, RNA, splicing, single-end paired-end)</a></li>
<li class="chapter" data-level="4.4" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#from-the-sequencer-machine-to-the-sequence"><i class="fa fa-check"></i><b>4.4</b> From the sequencer machine to the sequence</a></li>
<li class="chapter" data-level="4.5" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#general-processing"><i class="fa fa-check"></i><b>4.5</b> General processing</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#visualization-of-alignment-igv"><i class="fa fa-check"></i><b>4.5.1</b> Visualization of alignment (IGV)</a></li>
<li class="chapter" data-level="4.5.2" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#gene-counts-read-esat-feature-counts-peak-calling"><i class="fa fa-check"></i><b>4.5.2</b> Gene counts / read (esat feature counts) /peak calling</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#using-the-terminal"><i class="fa fa-check"></i><b>4.6</b> Using the terminal</a></li>
<li class="chapter" data-level="4.7" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#a-bit-of-shell-commands"><i class="fa fa-check"></i><b>4.7</b> A bit of shell commands</a></li>
<li class="chapter" data-level="4.8" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#resources-and-bibliography"><i class="fa fa-check"></i><b>4.8</b> Resources and Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html"><i class="fa fa-check"></i><b>5</b> Chip Sequencing Analysis</a>
<ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>5.1</b> Objectives</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>5.2</b> Introduction</a></li>
<li class="chapter" data-level="5.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#data-pre-processing-peak-calling"><i class="fa fa-check"></i><b>5.3</b> Data pre-processing (peak-calling):</a></li>
<li class="chapter" data-level="5.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#exploring-the-results"><i class="fa fa-check"></i><b>5.4</b> Exploring the results:</a></li>
<li class="chapter" data-level="5.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#annotating-the-peaks."><i class="fa fa-check"></i><b>5.5</b> Annotating the peaks.</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#read-the-annotation-file"><i class="fa fa-check"></i><b>5.5.1</b> Read the annotation file</a></li>
<li class="chapter" data-level="5.5.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#get-the-genes-from-this-annotation-data-base"><i class="fa fa-check"></i><b>5.5.2</b> Get the genes from this annotation data base</a></li>
<li class="chapter" data-level="5.5.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#get-the-genomic-ranges-of-the-peaks-from-macs-and-find-which-gene-is-closer-to-them-ie.-annotate-them"><i class="fa fa-check"></i><b>5.5.3</b> Get the genomic ranges of the peaks from MACS and find which gene is closer to them (ie. Annotate them)</a></li>
<li class="chapter" data-level="5.5.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#change-the-gene-format."><i class="fa fa-check"></i><b>5.5.4</b> Change the gene format.</a></li>
<li class="chapter" data-level="5.5.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-and-export-the-table"><i class="fa fa-check"></i><b>5.5.5</b> Explore and export the table</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#annotate-with-chipseeker"><i class="fa fa-check"></i><b>5.6</b> Annotate with <code>ChIPseeker</code></a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-and-export-the-output"><i class="fa fa-check"></i><b>5.6.1</b> Explore and Export the output</a></li>
<li class="chapter" data-level="5.6.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#positive-controls."><i class="fa fa-check"></i><b>5.6.2</b> Positive controls.</a></li>
<li class="chapter" data-level="5.6.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#plots"><i class="fa fa-check"></i><b>5.6.3</b> Plots</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#sequence-motif-analysis"><i class="fa fa-check"></i><b>5.7</b> Sequence Motif analysis</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#put-the-peak-sequences-in-fasta-format"><i class="fa fa-check"></i><b>5.7.1</b> Put the peak sequences in fasta format</a></li>
<li class="chapter" data-level="5.7.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#find-the-motifs"><i class="fa fa-check"></i><b>5.7.2</b> Find the MOTIFS</a></li>
<li class="chapter" data-level="5.7.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-the-results"><i class="fa fa-check"></i><b>5.7.3</b> Explore the results</a></li>
<li class="chapter" data-level="5.7.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#plot-the-results"><i class="fa fa-check"></i><b>5.7.4</b> Plot the results</a></li>
<li class="chapter" data-level="5.7.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>5.7.5</b> Plotting with Ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#extra-how-to-solve-some-common-annotation-issues."><i class="fa fa-check"></i><b>5.8</b> Extra: How to solve some common annotation issues.</a></li>
<li class="chapter" data-level="5.9" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#extra-detailed-explanation-of-the-macs-output-files"><i class="fa fa-check"></i><b>5.9</b> Extra: Detailed explanation of the MACS output files</a></li>
<li class="chapter" data-level="5.10" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#activity"><i class="fa fa-check"></i><b>5.10</b> Activity</a></li>
<li class="chapter" data-level="5.11" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#resources-and-bibliography"><i class="fa fa-check"></i><b>5.11</b> Resources and Bibliography</a></li>
<li class="chapter" data-level="5.12" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#session-info-all-the-packages-installed."><i class="fa fa-check"></i><b>5.12</b> Session info: all the packages installed.</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html"><i class="fa fa-check"></i><b>6</b> RNA-seq analysis</a>
<ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>6.1</b> Objectives</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>6.2</b> Introduction</a></li>
<li class="chapter" data-level="6.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#differential-gene-expression"><i class="fa fa-check"></i><b>6.3</b> Differential gene expression</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#loading-the-tables"><i class="fa fa-check"></i><b>6.3.1</b> Loading the tables</a></li>
<li class="chapter" data-level="6.3.2" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#create-the-deseq2-count-matrix"><i class="fa fa-check"></i><b>6.3.2</b> Create the DeSeq2 count matrix</a></li>
<li class="chapter" data-level="6.3.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#meta-data-coldata-preparation"><i class="fa fa-check"></i><b>6.3.3</b> Meta-data (colData) preparation</a></li>
<li class="chapter" data-level="6.3.4" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#quality-filtering-distance-matrix-and-principal-component-analysis-pca"><i class="fa fa-check"></i><b>6.3.4</b> Quality filtering: Distance matrix and principal component analysis (PCA)</a></li>
<li class="chapter" data-level="6.3.5" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#differential-gene-expression-analysis-between-zt."><i class="fa fa-check"></i><b>6.3.5</b> Differential gene expression analysis between ZT.</a></li>
<li class="chapter" data-level="6.3.6" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#explore-results"><i class="fa fa-check"></i><b>6.3.6</b> Explore results</a></li>
<li class="chapter" data-level="6.3.7" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#plot-counts-using-ggplot"><i class="fa fa-check"></i><b>6.3.7</b> Plot counts using ggplot</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#gene-ontology-go-term-analysis"><i class="fa fa-check"></i><b>6.4</b> Gene Ontology (GO) term analysis:</a></li>
<li class="chapter" data-level="6.5" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#extra-negative-binomial-model"><i class="fa fa-check"></i><b>6.5</b> Extra: negative binomial model</a></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#activity"><i class="fa fa-check"></i><b>6.6</b> Activity</a></li>
<li class="chapter" data-level="6.7" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#resources-and-bibliography"><i class="fa fa-check"></i><b>6.7</b> Resources and Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="circadian-analysis.html"><a href="circadian-analysis.html"><i class="fa fa-check"></i><b>7</b> Circadian analysis</a>
<ul>
<li class="chapter" data-level="7.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#objectives-1"><i class="fa fa-check"></i><b>7.1</b> Objectives</a></li>
<li class="chapter" data-level="7.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#introduction-1"><i class="fa fa-check"></i><b>7.2</b> Introduction</a></li>
<li class="chapter" data-level="7.3" data-path="circadian-analysis.html"><a href="circadian-analysis.html#cycling-genes-analysis"><i class="fa fa-check"></i><b>7.3</b> Cycling genes analysis</a></li>
<li class="chapter" data-level="7.4" data-path="circadian-analysis.html"><a href="circadian-analysis.html#cycling-genes-analysis-normalizing-by-maximum"><i class="fa fa-check"></i><b>7.4</b> Cycling genes analysis normalizing by maximum</a></li>
<li class="chapter" data-level="7.5" data-path="circadian-analysis.html"><a href="circadian-analysis.html#plot-the-data."><i class="fa fa-check"></i><b>7.5</b> Plot the data.</a></li>
<li class="chapter" data-level="7.6" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#activity-1"><i class="fa fa-check"></i><b>7.6</b> Activity</a></li>
<li class="chapter" data-level="7.7" data-path="circadian-analysis.html"><a href="circadian-analysis.html#resources-and-bibliography-1"><i class="fa fa-check"></i><b>7.7</b> Resources and Bibliography</a></li>
</ul></li>
<li class="appendix"><span><b>Appendices</b></span></li>
<li class="divider"></li>
<li><a rel="license" href="https://creativecommons.org/licenses/by-sa/4.0/" 
    target="blank"><img alt="Creative Commons License" 
    style="border-width:0" 
    src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><li>
<li><a href="https://psyteachr.github.io/" target="blank">PsyTeachR</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Genomics with R for biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="data-manipulation" class="section level1" number="2">
<h1><span class="header-section-number">Lab 2</span> Data manipulation</h1>
<div id="objectives" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Objectives</h2>
<p>After this section you should be able to:</p>
<ol style="list-style-type: decimal">
<li>Load, explore and manipulate data in R</li>
</ol>
</div>
<div id="introduction" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Introduction</h2>
<p>One of the main uses of R is for data manipulation and plot. This is similar to what many of us do in any regular table editor as excel or google spread sheet.</p>
<p>We will use the following packages. You can read in detail the manual of each of them.</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="data-manipulation.html#cb70-1"></a><span class="co">#Install packages</span></span>
<span id="cb70-2"><a href="data-manipulation.html#cb70-2"></a><span class="co">#install.packages(&quot;ggplot2&quot;)</span></span>
<span id="cb70-3"><a href="data-manipulation.html#cb70-3"></a><span class="co">#install.packages(&quot;dplyr&quot;)</span></span>
<span id="cb70-4"><a href="data-manipulation.html#cb70-4"></a><span class="co">#install.packages(&quot;plyr&quot;)</span></span>
<span id="cb70-5"><a href="data-manipulation.html#cb70-5"></a></span>
<span id="cb70-6"><a href="data-manipulation.html#cb70-6"></a><span class="co">#Load the package</span></span>
<span id="cb70-7"><a href="data-manipulation.html#cb70-7"></a><span class="kw">library</span>(<span class="st">&quot;ggplot2&quot;</span>)</span>
<span id="cb70-8"><a href="data-manipulation.html#cb70-8"></a><span class="kw">library</span>(<span class="st">&quot;dplyr&quot;</span>)</span>
<span id="cb70-9"><a href="data-manipulation.html#cb70-9"></a><span class="kw">library</span>(<span class="st">&quot;plyr&quot;</span>)</span>
<span id="cb70-10"><a href="data-manipulation.html#cb70-10"></a><span class="kw">library</span>(RColorBrewer)</span>
<span id="cb70-11"><a href="data-manipulation.html#cb70-11"></a><span class="kw">library</span>(car)</span>
<span id="cb70-12"><a href="data-manipulation.html#cb70-12"></a></span>
<span id="cb70-13"><a href="data-manipulation.html#cb70-13"></a><span class="co">#Manuals</span></span>
<span id="cb70-14"><a href="data-manipulation.html#cb70-14"></a><span class="co">#vignette(&quot;dplyr&quot;)</span></span>
<span id="cb70-15"><a href="data-manipulation.html#cb70-15"></a><span class="co">#?ggplot2</span></span>
<span id="cb70-16"><a href="data-manipulation.html#cb70-16"></a><span class="co">#?plyr</span></span></code></pre></div>
</div>
<div id="load-data" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Load data</h2>
<p>There are many ways to load data. In the following chapters we will use a diverse set of functions to read the data from files. Some of them are:</p>
<pre><code>read.table() #general to any type of table
read.csv() #specific for comma sepparated tables
read.delim() #specific for tab delimited tables</code></pre>
<p>Some of the important options of these function are:</p>
<pre><code>read.table(file = &quot;location/of/your/file.txt&quot;,sep = &quot;.&quot;,header = T or F)</code></pre>
<p>Where the separator can be a comma, dot, etc. You can see more details using: <code>?read.table</code></p>
<p>In this case we will use data that is already available in R. The package <a href="https://stat.ethz.ch/R-manual/R-devel/library/datasets/html/00Index.html">datasets</a> provides a handful set of data to analyze.</p>
<p>We will use the <a href="https://stat.ethz.ch/R-manual/R-devel/library/datasets/html/ChickWeight.html">ChickWeight</a> dataset. This is data set of weight in chickens with age an different diet.</p>
<p>This will allow us to visualize the data and to do some statistic tests.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="data-manipulation.html#cb73-1"></a><span class="co"># Install the package</span></span>
<span id="cb73-2"><a href="data-manipulation.html#cb73-2"></a><span class="co">#install.packages(&quot;datasets&quot;)</span></span>
<span id="cb73-3"><a href="data-manipulation.html#cb73-3"></a><span class="co"># For a full list of these datasets, type library(help = &quot;datasets&quot;)</span></span>
<span id="cb73-4"><a href="data-manipulation.html#cb73-4"></a><span class="co"># Load the library and dataset</span></span>
<span id="cb73-5"><a href="data-manipulation.html#cb73-5"></a><span class="kw">library</span>(datasets)</span>
<span id="cb73-6"><a href="data-manipulation.html#cb73-6"></a><span class="kw">data</span>(ChickWeight) <span class="co">#What happens in the Environment section of RStudio?</span></span></code></pre></div>
</div>
<div id="data-exploration" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> Data exploration</h2>
<p>It is important to understand the data before heading into the analysis. We will go over some techniques for this.</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="data-manipulation.html#cb74-1"></a><span class="co"># To see the table, you can click on the environment part or run this...</span></span>
<span id="cb74-2"><a href="data-manipulation.html#cb74-2"></a><span class="co">#View(ChickWeight)</span></span>
<span id="cb74-3"><a href="data-manipulation.html#cb74-3"></a></span>
<span id="cb74-4"><a href="data-manipulation.html#cb74-4"></a><span class="co"># As you can see this is a table, just in case we want to convert it to a data.frame</span></span>
<span id="cb74-5"><a href="data-manipulation.html#cb74-5"></a>ChickWeight&lt;-<span class="kw">as.data.frame</span>(ChickWeight)</span></code></pre></div>
<p>To see only the beginning, we can use the head function:</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="data-manipulation.html#cb75-1"></a><span class="kw">head</span>(ChickWeight)</span></code></pre></div>
<pre><code>##   weight Time Chick Diet
## 1     42    0     1    1
## 2     51    2     1    1
## 3     59    4     1    1
## 4     64    6     1    1
## 5     76    8     1    1
## 6     93   10     1    1</code></pre>
<blockquote>
<p>What is n doing?</p>
</blockquote>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="data-manipulation.html#cb77-1"></a><span class="kw">head</span>(ChickWeight,<span class="dt">n =</span> <span class="dv">20</span>) </span></code></pre></div>
<pre><code>##    weight Time Chick Diet
## 1      42    0     1    1
## 2      51    2     1    1
## 3      59    4     1    1
## 4      64    6     1    1
## 5      76    8     1    1
## 6      93   10     1    1
## 7     106   12     1    1
## 8     125   14     1    1
## 9     149   16     1    1
## 10    171   18     1    1
## 11    199   20     1    1
## 12    205   21     1    1
## 13     40    0     2    1
## 14     49    2     2    1
## 15     58    4     2    1
## 16     72    6     2    1
## 17     84    8     2    1
## 18    103   10     2    1
## 19    122   12     2    1
## 20    138   14     2    1</code></pre>
<blockquote>
<p>What is the structure of the data.frame?</p>
</blockquote>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="data-manipulation.html#cb79-1"></a><span class="kw">str</span>(ChickWeight) </span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:	578 obs. of  4 variables:
##  $ weight: num  42 51 59 64 76 93 106 125 149 171 ...
##  $ Time  : num  0 2 4 6 8 10 12 14 16 18 ...
##  $ Chick : Ord.factor w/ 50 levels &quot;18&quot;&lt;&quot;16&quot;&lt;&quot;15&quot;&lt;..: 15 15 15 15 15 15 15 15 15 15 ...
##  $ Diet  : Factor w/ 4 levels &quot;1&quot;,&quot;2&quot;,&quot;3&quot;,&quot;4&quot;: 1 1 1 1 1 1 1 1 1 1 ...
##  - attr(*, &quot;formula&quot;)=Class &#39;formula&#39;  language weight ~ Time | Chick
##   .. ..- attr(*, &quot;.Environment&quot;)=&lt;environment: R_EmptyEnv&gt; 
##  - attr(*, &quot;outer&quot;)=Class &#39;formula&#39;  language ~Diet
##   .. ..- attr(*, &quot;.Environment&quot;)=&lt;environment: R_EmptyEnv&gt; 
##  - attr(*, &quot;labels&quot;)=List of 2
##   ..$ x: chr &quot;Time&quot;
##   ..$ y: chr &quot;Body weight&quot;
##  - attr(*, &quot;units&quot;)=List of 2
##   ..$ x: chr &quot;(days)&quot;
##   ..$ y: chr &quot;(gm)&quot;</code></pre>
<p>With the <strong>$ operator</strong> we can explore the columns</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="data-manipulation.html#cb81-1"></a><span class="kw">class</span>(ChickWeight<span class="op">$</span>weight)</span></code></pre></div>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<p>We can see the <strong>dimensions of the table</strong> for example: how many <strong>rows</strong> it has?</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="data-manipulation.html#cb83-1"></a><span class="kw">nrow</span>(ChickWeight)</span></code></pre></div>
<pre><code>## [1] 578</code></pre>
<p>How many <strong>columns</strong>?</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="data-manipulation.html#cb85-1"></a><span class="kw">ncol</span>(ChickWeight)</span></code></pre></div>
<pre><code>## [1] 4</code></pre>
<p>The <strong>names</strong> of columns</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="data-manipulation.html#cb87-1"></a><span class="kw">names</span>(ChickWeight)</span></code></pre></div>
<pre><code>## [1] &quot;weight&quot; &quot;Time&quot;   &quot;Chick&quot;  &quot;Diet&quot;</code></pre>
<p>With the <strong>[]</strong> we can access the individual elements</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="data-manipulation.html#cb89-1"></a><span class="kw">names</span>(ChickWeight)[<span class="dv">3</span>]</span></code></pre></div>
<pre><code>## [1] &quot;Chick&quot;</code></pre>
<p>We can see the levels of a factor</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="data-manipulation.html#cb91-1"></a><span class="kw">levels</span>(ChickWeight<span class="op">$</span>Diet)[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>## [1] &quot;1&quot; &quot;2&quot; &quot;3&quot;</code></pre>
<p>What is the difference if we just print the column?</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="data-manipulation.html#cb93-1"></a>ChickWeight<span class="op">$</span>Diet[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</span></code></pre></div>
<pre><code>## [1] 1 1 1
## Levels: 1 2 3 4</code></pre>
<p>Can we see the levels of a numeric vector? This is a reminder that the data type is important.</p>
<pre><code>levels(ChickWeight$weight) # nop</code></pre>
<p>We can now get different basic statistics now:</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="data-manipulation.html#cb96-1"></a><span class="kw">mean</span>(ChickWeight<span class="op">$</span>weight)</span></code></pre></div>
<pre><code>## [1] 121.8183</code></pre>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="data-manipulation.html#cb98-1"></a><span class="kw">summary</span>(ChickWeight<span class="op">$</span>weight)</span></code></pre></div>
<pre><code>##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##    35.0    63.0   103.0   121.8   163.8   373.0</code></pre>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="data-manipulation.html#cb100-1"></a><span class="kw">summary</span>(ChickWeight)</span></code></pre></div>
<pre><code>##      weight           Time           Chick     Diet   
##  Min.   : 35.0   Min.   : 0.00   13     : 12   1:220  
##  1st Qu.: 63.0   1st Qu.: 4.00   9      : 12   2:120  
##  Median :103.0   Median :10.00   20     : 12   3:120  
##  Mean   :121.8   Mean   :10.72   10     : 12   4:118  
##  3rd Qu.:163.8   3rd Qu.:16.00   17     : 12          
##  Max.   :373.0   Max.   :21.00   19     : 12          
##                                  (Other):506</code></pre>
<p>To see what is this exactly doing, just go to the help page:
<code>?summary</code></p>
<p>To save this summary table we can create an object with just the result of the summary</p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="data-manipulation.html#cb102-1"></a>chick_sumary&lt;-<span class="kw">summary</span>(ChickWeight)</span>
<span id="cb102-2"><a href="data-manipulation.html#cb102-2"></a>chick_sumary</span></code></pre></div>
<pre><code>##      weight           Time           Chick     Diet   
##  Min.   : 35.0   Min.   : 0.00   13     : 12   1:220  
##  1st Qu.: 63.0   1st Qu.: 4.00   9      : 12   2:120  
##  Median :103.0   Median :10.00   20     : 12   3:120  
##  Mean   :121.8   Mean   :10.72   10     : 12   4:118  
##  3rd Qu.:163.8   3rd Qu.:16.00   17     : 12          
##  Max.   :373.0   Max.   :21.00   19     : 12          
##                                  (Other):506</code></pre>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="data-manipulation.html#cb104-1"></a><span class="kw">class</span>(chick_sumary)</span></code></pre></div>
<pre><code>## [1] &quot;table&quot;</code></pre>
<p>We can change the data kind, and assign it to a different object</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="data-manipulation.html#cb106-1"></a>chick_sumary_df&lt;-<span class="kw">as.data.frame</span>(chick_sumary)</span></code></pre></div>
<p>This is not that useful as you can see if you inspect the data in using View(chick_sumary_df) this is because it is a complicated format, we better just save the table.
We will see other ways to save data in R in the future chapters.
You can see more details using: <code>?write.table</code></p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="data-manipulation.html#cb107-1"></a><span class="kw">write.table</span>(chick_sumary, <span class="st">&quot;mydata.txt&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">row.names =</span> F,<span class="dt">col.names =</span> T)</span>
<span id="cb107-2"><a href="data-manipulation.html#cb107-2"></a><span class="co">#this is clearly no perfect but for the important part, the numeric and integer columns, we have the stat</span></span></code></pre></div>
</div>
<div id="subsetting" class="section level2" number="2.5">
<h2><span class="header-section-number">2.5</span> Subsetting</h2>
<p>Subsetting means extracting part of the data. There are many different ways to do this. One important notion for tables and data frames is that dimensions go as follows:
<code>data[row,column]</code></p>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="data-manipulation.html#cb108-1"></a><span class="co">#we can see specific columns and rows</span></span>
<span id="cb108-2"><a href="data-manipulation.html#cb108-2"></a>ChickWeight[<span class="dv">1</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>] <span class="co">#row 1, column 1:3</span></span></code></pre></div>
<pre><code>##   weight Time Chick
## 1     42    0     1</code></pre>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="data-manipulation.html#cb110-1"></a>ChickWeight[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>,<span class="dv">1</span>] <span class="co">#col 1, row 1:3</span></span></code></pre></div>
<pre><code>## [1] 42 51 59</code></pre>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="data-manipulation.html#cb112-1"></a>ChickWeight[<span class="dv">1</span>,<span class="dv">1</span>] <span class="co">#row1, col1</span></span></code></pre></div>
<pre><code>## [1] 42</code></pre>
<p>If we want to know for example only the data from the chickens taking the diet 4</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="data-manipulation.html#cb114-1"></a><span class="kw">head</span>(ChickWeight[ChickWeight<span class="op">$</span>Diet<span class="op">==</span><span class="dv">4</span>,])</span></code></pre></div>
<pre><code>##     weight Time Chick Diet
## 461     42    0    41    4
## 462     51    2    41    4
## 463     66    4    41    4
## 464     85    6    41    4
## 465    103    8    41    4
## 466    124   10    41    4</code></pre>
<blockquote>
<p>Why == and no =?</p>
</blockquote>
<p>Remember in R, = is an assignment, as the &lt;-, while the == is for comparison.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="data-manipulation.html#cb116-1"></a><span class="kw">head</span>(ChickWeight<span class="op">$</span>Diet<span class="op">==</span><span class="dv">4</span>)</span></code></pre></div>
<pre><code>## [1] FALSE FALSE FALSE FALSE FALSE FALSE</code></pre>
<p>Lets explore the class:</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="data-manipulation.html#cb118-1"></a><span class="kw">class</span>(ChickWeight<span class="op">$</span>Diet<span class="op">==</span><span class="dv">4</span>)</span></code></pre></div>
<pre><code>## [1] &quot;logical&quot;</code></pre>
<p>So, when we do ChickWeight[ChickWeight$Diet==4,], R is just showing the ChickWeight for which ChickWeight$Diet==4 is TRUE</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="data-manipulation.html#cb120-1"></a><span class="kw">head</span>(<span class="kw">which</span>(ChickWeight<span class="op">$</span>Diet<span class="op">==</span><span class="dv">4</span>))</span></code></pre></div>
<pre><code>## [1] 461 462 463 464 465 466</code></pre>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="data-manipulation.html#cb122-1"></a><span class="kw">head</span>(ChickWeight[ChickWeight<span class="op">$</span>Diet<span class="op">==</span><span class="dv">4</span>,])</span></code></pre></div>
<pre><code>##     weight Time Chick Diet
## 461     42    0    41    4
## 462     51    2    41    4
## 463     66    4    41    4
## 464     85    6    41    4
## 465    103    8    41    4
## 466    124   10    41    4</code></pre>
<p>And for more conditions, we can use AND (&amp;) to integrate them.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="data-manipulation.html#cb124-1"></a><span class="kw">head</span>(ChickWeight[ChickWeight<span class="op">$</span>Diet<span class="op">==</span><span class="dv">4</span> <span class="op">&amp;</span><span class="st"> </span>ChickWeight<span class="op">$</span>Time<span class="op">&gt;</span><span class="dv">6</span>,])</span></code></pre></div>
<pre><code>##     weight Time Chick Diet
## 465    103    8    41    4
## 466    124   10    41    4
## 467    155   12    41    4
## 468    153   14    41    4
## 469    175   16    41    4
## 470    184   18    41    4</code></pre>
<p>Other option is OR (|).</p>
<p>Remember, computers will read as things come
<span class="math display">\[
condition-A AND condition-B OR condition-C
condition-A &amp; condition-B | condition-C
\]</span>
Is not the same as</p>
<p><span class="math display">\[
condition A &amp; (condition B | condition C)
\]</span></p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="data-manipulation.html#cb126-1"></a><span class="kw">head</span>(ChickWeight[ChickWeight<span class="op">$</span>Diet<span class="op">==</span><span class="dv">4</span> <span class="op">&amp;</span><span class="st"> </span>ChickWeight<span class="op">$</span>Time<span class="op">&gt;</span><span class="dv">6</span> <span class="op">&amp;</span><span class="st"> </span>ChickWeight<span class="op">$</span>Time<span class="op">&lt;</span><span class="dv">20</span>,])</span></code></pre></div>
<pre><code>##     weight Time Chick Diet
## 465    103    8    41    4
## 466    124   10    41    4
## 467    155   12    41    4
## 468    153   14    41    4
## 469    175   16    41    4
## 470    184   18    41    4</code></pre>
<p>And if we just want the weights of theseâ€¦</p>
<pre><code>ChickWeight$weight[ChickWeight$Diet==4 &amp; ChickWeight$Time&gt;6 &amp; ChickWeight$Time&lt;20,] </code></pre>
<blockquote>
<p>why this gives an error?</p>
</blockquote>
<p>Because we only have one dimension now, not 2.
ChickWeight$weight is one dimention object, so we have to use [ ], not [ , ].</p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="data-manipulation.html#cb129-1"></a><span class="kw">head</span>(ChickWeight<span class="op">$</span>weight[ChickWeight<span class="op">$</span>Diet<span class="op">==</span><span class="dv">4</span> <span class="op">&amp;</span><span class="st"> </span>ChickWeight<span class="op">$</span>Time<span class="op">&gt;</span><span class="dv">6</span> <span class="op">&amp;</span><span class="st"> </span>ChickWeight<span class="op">$</span>Time<span class="op">&lt;</span><span class="dv">20</span>])</span></code></pre></div>
<pre><code>## [1] 103 124 155 153 175 184</code></pre>
</div>
<div id="activity" class="section level2" number="2.6">
<h2><span class="header-section-number">2.6</span> Activity:</h2>
<p>This activity integrates knowledge from the previous chapter.</p>
<p><strong>1. Remove the first and last row of the ChickWeight data frame</strong>
<strong>2. Create a vector with the second column from the data frame</strong></p>
</div>
</div>
<div class="psyteachr_footer">
  
</div>
<script>

/* update total correct if #total_correct exists */
update_total_correct = function() {
  if (t = document.getElementById("total_correct")) {
    t.innerHTML =
      document.getElementsByClassName("correct").length + " of " +
      document.getElementsByClassName("solveme").length + " correct";
  }
}

/* solution button toggling function */
b_func = function() {
  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "");
  }
  
  if (my_answer !== "" & real_answers.includes(my_answer)) {
    cl.add("correct");
  } else {
    cl.remove("correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol){
    var tol = JSON.parse(this.dataset.tol);  
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("correct");
    } else {
      cl.remove("correct");
    }  
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("correct");
    }  
  }
  
  update_total_correct();
}

window.onload = function() {
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('solution')) {
      buttons[i].onclick = b_func;
    }
  }
  
  /* set up solveme inputs */
  var solveme = document.getElementsByClassName("solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off"); 
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";
    
    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;
    
    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;
  }
  
  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="introduction-to-r-and-rstudio.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="visualization-and-statistical-test.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"],
"google": false
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": {},
"toc": {
"collapse": "section",
"scroll_highlight": true
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
