<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Lab 6 RNA seq analysis | Genomics with R for biologists</title>
  <meta name="description" content="This is the index for the coursebook" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Lab 6 RNA seq analysis | Genomics with R for biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is the index for the coursebook" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Lab 6 RNA seq analysis | Genomics with R for biologists" />
  
  <meta name="twitter:description" content="This is the index for the coursebook" />
  

<meta name="author" content="Ines Lucia Patop inespatop@brandeis.edu" />


<meta name="date" content="2020-07-25" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="chip-sequencing-analysis.html"/>

<script src="libs/header-attrs-2.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="include/psyteachr.css" type="text/css" />
<link rel="stylesheet" href="include/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Genomics with R for biologists</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Overview</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#course-aims"><i class="fa fa-check"></i><b>0.1</b> Course Aims</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#intended-learning-outcomes"><i class="fa fa-check"></i><b>0.2</b> Intended Learning Outcomes</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#next-steps"><i class="fa fa-check"></i><b>0.3</b> Next steps</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#data-availability"><i class="fa fa-check"></i><b>0.4</b> Data availability</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#acknowledgments"><i class="fa fa-check"></i><b>0.5</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html"><i class="fa fa-check"></i><b>1</b> Introduction to R and RStudio</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>1.1</b> Objectives</a></li>
<li class="chapter" data-level="1.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>1.2</b> Introduction</a></li>
<li class="chapter" data-level="1.3" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#rstudio-world"><i class="fa fa-check"></i><b>1.3</b> RStudio world</a></li>
<li class="chapter" data-level="1.4" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#create-a-project"><i class="fa fa-check"></i><b>1.4</b> Create a project</a></li>
<li class="chapter" data-level="1.5" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#useful-shortcuts-in-rstudio"><i class="fa fa-check"></i><b>1.5</b> Useful shortcuts in RStudio:</a></li>
<li class="chapter" data-level="1.6" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#installing-and-loading-packages"><i class="fa fa-check"></i><b>1.6</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.7" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#creating-objects"><i class="fa fa-check"></i><b>1.7</b> Creating objects</a></li>
<li class="chapter" data-level="1.8" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#r-as-a-calculator"><i class="fa fa-check"></i><b>1.8</b> R as a calculator</a></li>
<li class="chapter" data-level="1.9" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#functions"><i class="fa fa-check"></i><b>1.9</b> Functions</a></li>
<li class="chapter" data-level="1.10" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#loops"><i class="fa fa-check"></i><b>1.10</b> Loops</a></li>
<li class="chapter" data-level="1.11" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#conditions"><i class="fa fa-check"></i><b>1.11</b> Conditions</a></li>
<li class="chapter" data-level="1.12" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#character-objects"><i class="fa fa-check"></i><b>1.12</b> Character objects</a></li>
<li class="chapter" data-level="1.13" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#identify"><i class="fa fa-check"></i><b>1.13</b> Identify</a></li>
<li class="chapter" data-level="1.14" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#modify"><i class="fa fa-check"></i><b>1.14</b> Modify</a></li>
<li class="chapter" data-level="1.15" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#useful-resources"><i class="fa fa-check"></i><b>1.15</b> Useful resources</a></li>
<li class="chapter" data-level="1.16" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#activity"><i class="fa fa-check"></i><b>1.16</b> Activity:</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-manipulation.html"><a href="data-manipulation.html"><i class="fa fa-check"></i><b>2</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-manipulation.html"><a href="data-manipulation.html#objectives-1"><i class="fa fa-check"></i><b>2.1</b> Objectives</a></li>
<li class="chapter" data-level="2.2" data-path="data-manipulation.html"><a href="data-manipulation.html#introduction-1"><i class="fa fa-check"></i><b>2.2</b> Introduction</a></li>
<li class="chapter" data-level="2.3" data-path="data-manipulation.html"><a href="data-manipulation.html#load-data"><i class="fa fa-check"></i><b>2.3</b> Load data</a></li>
<li class="chapter" data-level="2.4" data-path="data-manipulation.html"><a href="data-manipulation.html#data-exploration"><i class="fa fa-check"></i><b>2.4</b> Data exploration</a></li>
<li class="chapter" data-level="2.5" data-path="data-manipulation.html"><a href="data-manipulation.html#subsetting"><i class="fa fa-check"></i><b>2.5</b> Subsetting</a></li>
<li class="chapter" data-level="2.6" data-path="data-manipulation.html"><a href="data-manipulation.html#activity-1"><i class="fa fa-check"></i><b>2.6</b> Activity:</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html"><i class="fa fa-check"></i><b>3</b> Visualization and Statistical test</a>
<ul>
<li class="chapter" data-level="3.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#objectives-2"><i class="fa fa-check"></i><b>3.1</b> Objectives</a></li>
<li class="chapter" data-level="3.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#introduction-2"><i class="fa fa-check"></i><b>3.2</b> Introduction</a></li>
<li class="chapter" data-level="3.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#plots"><i class="fa fa-check"></i><b>3.3</b> Plots</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#line-and-points"><i class="fa fa-check"></i><b>3.3.1</b> Line and points</a></li>
<li class="chapter" data-level="3.3.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#historgam-and-density-plots"><i class="fa fa-check"></i><b>3.3.2</b> Historgam and density plots</a></li>
<li class="chapter" data-level="3.3.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#boxplot"><i class="fa fa-check"></i><b>3.3.3</b> Boxplot</a></li>
<li class="chapter" data-level="3.3.4" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#saving-plots"><i class="fa fa-check"></i><b>3.3.4</b> Saving plots</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#statistical-test"><i class="fa fa-check"></i><b>3.4</b> Statistical test</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#descriptive-statistics"><i class="fa fa-check"></i><b>3.4.1</b> Descriptive statistics</a></li>
<li class="chapter" data-level="3.4.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#t-testwilcoxon"><i class="fa fa-check"></i><b>3.4.2</b> T-test/Wilcoxon</a></li>
<li class="chapter" data-level="3.4.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#anovakruskalwallis"><i class="fa fa-check"></i><b>3.4.3</b> Anova/Kruskal–Wallis</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#activity-2"><i class="fa fa-check"></i><b>3.5</b> Activity:</a></li>
<li class="chapter" data-level="3.6" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#resources"><i class="fa fa-check"></i><b>3.6</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html"><i class="fa fa-check"></i><b>4</b> Sequencing techniques and preprocessing</a>
<ul>
<li class="chapter" data-level="4.1" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#objectives-3"><i class="fa fa-check"></i><b>4.1</b> Objectives</a></li>
<li class="chapter" data-level="4.2" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#introduction-3"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#different-type-of-data-dna-rna-splicing-single-end-paired-end"><i class="fa fa-check"></i><b>4.3</b> Different type of data (DNA, RNA, splicing, single-end paired-end)</a></li>
<li class="chapter" data-level="4.4" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#from-the-sequencer-machine-to-the-sequence"><i class="fa fa-check"></i><b>4.4</b> From the sequencer machine to the sequence</a></li>
<li class="chapter" data-level="4.5" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#general-processing"><i class="fa fa-check"></i><b>4.5</b> General processing</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#visualization-of-alignment-igv"><i class="fa fa-check"></i><b>4.5.1</b> Visualization of alignment (IGV)</a></li>
<li class="chapter" data-level="4.5.2" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#gene-counts-read-esat-feature-counts-peak-calling"><i class="fa fa-check"></i><b>4.5.2</b> Gene counts / read (esat feature counts) /peak calling</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#using-the-terminal"><i class="fa fa-check"></i><b>4.6</b> Using the terminal</a></li>
<li class="chapter" data-level="4.7" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#a-bit-of-shell-commands"><i class="fa fa-check"></i><b>4.7</b> A bit of shell commands</a></li>
<li class="chapter" data-level="4.8" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#resources-and-bibliography"><i class="fa fa-check"></i><b>4.8</b> Resources and Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html"><i class="fa fa-check"></i><b>5</b> Chip Sequencing Analysis</a>
<ul>
<li class="chapter" data-level="5.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#objectives-4"><i class="fa fa-check"></i><b>5.1</b> Objectives</a></li>
<li class="chapter" data-level="5.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#introduction-4"><i class="fa fa-check"></i><b>5.2</b> Introduction</a></li>
<li class="chapter" data-level="5.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#data-pre-processing-peak-calling"><i class="fa fa-check"></i><b>5.3</b> Data pre-processing (peak-calling):</a></li>
<li class="chapter" data-level="5.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#exploring-the-results"><i class="fa fa-check"></i><b>5.4</b> Exploring the results:</a></li>
<li class="chapter" data-level="5.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#annotating-the-peaks."><i class="fa fa-check"></i><b>5.5</b> Annotating the peaks.</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#read-the-annotation-file"><i class="fa fa-check"></i><b>5.5.1</b> Read the annotation file</a></li>
<li class="chapter" data-level="5.5.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#get-the-genes-from-this-annotation-data-base"><i class="fa fa-check"></i><b>5.5.2</b> Get the genes from this annotation data base</a></li>
<li class="chapter" data-level="5.5.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#get-the-genomic-ranges-of-the-peaks-from-macs-and-find-which-gene-is-closer-to-them-ie.-annotate-them"><i class="fa fa-check"></i><b>5.5.3</b> Get the genomic ranges of the peaks from MACS and find which gene is closer to them (ie. Annotate them)</a></li>
<li class="chapter" data-level="5.5.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#change-the-gene-format."><i class="fa fa-check"></i><b>5.5.4</b> Change the gene format.</a></li>
<li class="chapter" data-level="5.5.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-and-export-the-table"><i class="fa fa-check"></i><b>5.5.5</b> Explore and export the table</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#annotate-with-chipseeker"><i class="fa fa-check"></i><b>5.6</b> Annotate with <code>ChIPseeker</code></a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-and-export-the-output"><i class="fa fa-check"></i><b>5.6.1</b> Explore and Export the output</a></li>
<li class="chapter" data-level="5.6.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#positive-controls."><i class="fa fa-check"></i><b>5.6.2</b> Positive controls.</a></li>
<li class="chapter" data-level="5.6.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#plots-1"><i class="fa fa-check"></i><b>5.6.3</b> Plots</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#sequence-motif-analysis"><i class="fa fa-check"></i><b>5.7</b> Sequence Motif analysis</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#put-the-peak-sequences-in-fasta-format"><i class="fa fa-check"></i><b>5.7.1</b> Put the peak sequences in fasta format</a></li>
<li class="chapter" data-level="5.7.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#find-the-motifs"><i class="fa fa-check"></i><b>5.7.2</b> Find the MOTIFS</a></li>
<li class="chapter" data-level="5.7.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-the-results"><i class="fa fa-check"></i><b>5.7.3</b> Explore the results</a></li>
<li class="chapter" data-level="5.7.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#plot-the-results"><i class="fa fa-check"></i><b>5.7.4</b> Plot the results</a></li>
<li class="chapter" data-level="5.7.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>5.7.5</b> Plotting with Ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#extra-how-to-solve-some-common-annotation-issues."><i class="fa fa-check"></i><b>5.8</b> Extra: How to solve some common annotation issues.</a></li>
<li class="chapter" data-level="5.9" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#extra-detailed-explanation-of-the-macs-output-files"><i class="fa fa-check"></i><b>5.9</b> Extra: Detailed explanation of the MACS output files</a></li>
<li class="chapter" data-level="5.10" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#activity-3"><i class="fa fa-check"></i><b>5.10</b> Activity</a></li>
<li class="chapter" data-level="5.11" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#resources-and-bibliography-1"><i class="fa fa-check"></i><b>5.11</b> Resources and Bibliography</a></li>
<li class="chapter" data-level="5.12" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#session-info-all-the-packages-installed."><i class="fa fa-check"></i><b>5.12</b> Session info: all the packages installed.</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html"><i class="fa fa-check"></i><b>6</b> RNA seq analysis</a>
<ul>
<li class="chapter" data-level="6.1" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#objectives-5"><i class="fa fa-check"></i><b>6.1</b> Objectives</a></li>
<li class="chapter" data-level="6.2" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#introduction-5"><i class="fa fa-check"></i><b>6.2</b> Introduction</a></li>
<li class="chapter" data-level="6.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#differential-gene-expression"><i class="fa fa-check"></i><b>6.3</b> Differential gene expression</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#loading-the-tables"><i class="fa fa-check"></i><b>6.3.1</b> Loading the tables</a></li>
<li class="chapter" data-level="6.3.2" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#create-the-deseq2-count-matrix"><i class="fa fa-check"></i><b>6.3.2</b> Create the DeSeq2 count matrix</a></li>
<li class="chapter" data-level="6.3.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#coldata-preparation"><i class="fa fa-check"></i><b>6.3.3</b> colData preparation</a></li>
<li class="chapter" data-level="6.3.4" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#heat-map-pca-analysis"><i class="fa fa-check"></i><b>6.3.4</b> Heat-map PCA analysis</a></li>
<li class="chapter" data-level="6.3.5" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#differential-gene-expression-analysis-between-zt."><i class="fa fa-check"></i><b>6.3.5</b> Differential gene expression analysis between ZT.</a></li>
<li class="chapter" data-level="6.3.6" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#plot-counts-using-ggplot"><i class="fa fa-check"></i><b>6.3.6</b> Plot counts using ggplot</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#cycling-genes-analysis"><i class="fa fa-check"></i><b>6.4</b> Cycling genes analysis</a></li>
<li class="chapter" data-level="6.5" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#activity-4"><i class="fa fa-check"></i><b>6.5</b> Activity</a></li>
<li class="chapter" data-level="6.6" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#extra-heatmap-and-pca-in-the-other-celltypes."><i class="fa fa-check"></i><b>6.6</b> Extra: Heatmap and PCA in the other celltypes.</a></li>
<li class="chapter" data-level="6.7" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#extra-advanced-cycling-genes-analysis"><i class="fa fa-check"></i><b>6.7</b> Extra: Advanced cycling genes analysis</a></li>
<li class="chapter" data-level="6.8" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#extra-negative-binomial-model"><i class="fa fa-check"></i><b>6.8</b> Extra: negative binomial model</a></li>
<li class="chapter" data-level="6.9" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#resources-and-bibliography-2"><i class="fa fa-check"></i><b>6.9</b> Resources and Bibliography</a></li>
</ul></li>
<li class="appendix"><span><b>Appendices</b></span></li>
<li class="divider"></li>
<li><a rel="license" href="https://creativecommons.org/licenses/by-sa/4.0/" 
    target="blank"><img alt="Creative Commons License" 
    style="border-width:0" 
    src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><li>
<li><a href="https://psyteachr.github.io/" target="blank">PsyTeachR</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Genomics with R for biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="rna-seq-analysis" class="section level1" number="6">
<h1><span class="header-section-number">Lab 6</span> RNA seq analysis</h1>
<div id="objectives-5" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Objectives</h2>
<p>After this section you should be able to:</p>
<ol style="list-style-type: decimal">
<li>Understand RNA sequencing counts.</li>
<li>Perform differential gene expresion.</li>
<li>Do GO term analysis.</li>
<li>Analyze circadian data.</li>
</ol>
</div>
<div id="introduction-5" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Introduction</h2>
<p>We will re-analyze some data from <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006613">Abruzzi et al, 2017</a>. These are data that explore gene expression in different neuronal cell types in a circadian layout.</p>
<p>This time we will just take the process data from the paper, and no directly the Raw data like in the previous ChipAnalisys. The data was downloaded from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE77451">GEO</a>. The only pre-processing to the data was done to change column names.</p>
<p>As you can see in the methods section of the paper they have different cell-types and timepoints. Lets try to understand what type of system and data we have in hands.</p>
<p><strong>Circadian behavior</strong></p>
<p>Most living organisms change their behavior and metabolism over the day. Some animals like us have more activity over the day while others like mice have more activity over night. Flies in particular have a peak of activity the first hours of the morning and the last hours of the afternoon with a “siesta” (nap in Spanish) at the middle of the day.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-2"></span>
<img src="RNAseq/circadianact.png" alt="Fly activity over the day. Adapted from Nicholas R. J Glossop et al 2011. To meassure fly activity each individual fly is monitored either by counting when they cross a red-light bim (regular activity monitors) of by software tracking (flyboxes)." width="500px" />
<p class="caption">
Figure 4.1: Fly activity over the day. Adapted from Nicholas R. J Glossop et al 2011. To meassure fly activity each individual fly is monitored either by counting when they cross a red-light bim (regular activity monitors) of by software tracking (flyboxes).
</p>
</div>
<p>These changes in behaviour are govern by changes in gene expression in particular cell-types in the brain called the pace-makers. If we look at gene expression in these cells over the day we will notice that some genes oscillate. Like <strong>Clk</strong> and <strong>tim</strong>, the proteins we analyzed in the previous Lab.</p>
<p>In the circadian field time is reported as “hours after the light is ON”. This is zeitgeber time (ZT) and it is usefull to have a unified meassurement of time. Most animals in labs are kept in 12 hours of light followed by 12 hours of dark. Therfore, ZT0 is the “sunrise” or lights on and ZT12 is lights off in a light/dark (LD)12:12 cycle time.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-3"></span>
<img src="RNAseq/cycwithzt.png" alt="Fly activity over the day with ZT scale. Figure adapted from Dubowy et al 2017" width="500px" />
<p class="caption">
Figure 4.2: Fly activity over the day with ZT scale. Figure adapted from Dubowy et al 2017
</p>
</div>
<p>As stated before, circadian behaviors is generated by a small subset of neurons.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-4"></span>
<img src="RNAseq/circadiancells.png" alt="Schematic representation of the circadian neural network. Four small ventrolateral neurons (s‐LNvs, red), the 5th s‐LNv (dark violet), four large ventrolateral neurons (l‐LNvs, brown), six dorsolateral neurons (LNds, orange), three lateral posterior neurons (LPN, green), and ca. 60 neurons per hemisphere in three dorsal groups (DN1–3, lilac, cyan, blue, respectively). Adapted from Schubert et al 2018" width="500px" />
<p class="caption">
Figure 4.3: Schematic representation of the circadian neural network. Four small ventrolateral neurons (s‐LNvs, red), the 5th s‐LNv (dark violet), four large ventrolateral neurons (l‐LNvs, brown), six dorsolateral neurons (LNds, orange), three lateral posterior neurons (LPN, green), and ca. 60 neurons per hemisphere in three dorsal groups (DN1–3, lilac, cyan, blue, respectively). Adapted from Schubert et al 2018
</p>
</div>
<p>In the data analzed in this lab, the authors collected RNA samples over 6 times of the day (ZT3, 7, 11, 15, 19, 23) from 4 celltypes: LNv, LNd, DN1 and Dopaminergic neurons (TH cells). This last group of cells are not part of the circadian core clock but in this study they see oscilating genes!</p>
<p><strong>Packages:</strong></p>
<p>We will use the following packages. Do not forget to executing them every time you open a new R session. Before runing them for the first time, you will have to install them (go back to Lab1 if you have any doubt).</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="rna-seq-analysis.html#cb262-1"></a><span class="kw">library</span>(DESeq2) <span class="co">#for differential expression analysis</span></span>
<span id="cb262-2"><a href="rna-seq-analysis.html#cb262-2"></a><span class="kw">library</span>(ggplot2) <span class="co">#for 2D graph</span></span>
<span id="cb262-3"><a href="rna-seq-analysis.html#cb262-3"></a><span class="kw">library</span>(ggrepel) <span class="co">#to get the names in ggplot graph </span></span>
<span id="cb262-4"><a href="rna-seq-analysis.html#cb262-4"></a><span class="kw">library</span>(gridExtra) <span class="co">#to arrange the plots</span></span>
<span id="cb262-5"><a href="rna-seq-analysis.html#cb262-5"></a><span class="kw">library</span>(factoextra) <span class="co">#extra plots</span></span>
<span id="cb262-6"><a href="rna-seq-analysis.html#cb262-6"></a><span class="kw">library</span>(plotly) <span class="co">#for 3D graphics</span></span>
<span id="cb262-7"><a href="rna-seq-analysis.html#cb262-7"></a><span class="kw">library</span>(plyr) <span class="co">#table manipulation</span></span>
<span id="cb262-8"><a href="rna-seq-analysis.html#cb262-8"></a><span class="kw">library</span>(dplyr)<span class="co">#table manipulation</span></span>
<span id="cb262-9"><a href="rna-seq-analysis.html#cb262-9"></a><span class="kw">library</span>(tidyr)<span class="co">#table manipulation</span></span>
<span id="cb262-10"><a href="rna-seq-analysis.html#cb262-10"></a><span class="kw">library</span>(<span class="st">&quot;RColorBrewer&quot;</span>) <span class="co">#extra color palettes</span></span>
<span id="cb262-11"><a href="rna-seq-analysis.html#cb262-11"></a><span class="kw">library</span>(<span class="st">&quot;pheatmap&quot;</span>) <span class="co">#nice heatmaps</span></span>
<span id="cb262-12"><a href="rna-seq-analysis.html#cb262-12"></a><span class="kw">library</span>(org.Dm.eg.db) <span class="co">#get annotation</span></span>
<span id="cb262-13"><a href="rna-seq-analysis.html#cb262-13"></a><span class="kw">library</span>(MetaCycle) <span class="co"># to identify cycling genes, the successor of JTK according to JTK authors</span></span>
<span id="cb262-14"><a href="rna-seq-analysis.html#cb262-14"></a><span class="kw">library</span>(pcaExplorer) <span class="co">#more PCA analysis</span></span></code></pre></div>
</div>
<div id="differential-gene-expression" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Differential gene expression</h2>
<p>Gene expression data is made of integer numbers by genes. Each number represents exactly how many reads aligned to the DNA sequence that is assigned to each gene.</p>
<table style="width:100%;">
<caption>Gene counts table</caption>
<colgroup>
<col width="9%" />
<col width="21%" />
<col width="23%" />
<col width="21%" />
<col width="23%" />
</colgroup>
<thead>
<tr class="header">
<th align="right">Gene</th>
<th align="right">Condition 1 replicate 1</th>
<th align="right">Condition 1 replicate 2</th>
<th align="right">Condition 2 replicate 1</th>
<th align="right">Condition 2 replicate 2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">A</td>
<td align="right">20</td>
<td align="right">400</td>
<td align="right">60</td>
<td align="right">40</td>
</tr>
<tr class="even">
<td align="right">B</td>
<td align="right">2</td>
<td align="right">50</td>
<td align="right">2000</td>
<td align="right">2500</td>
</tr>
<tr class="odd">
<td align="right">C</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">D</td>
<td align="right">5</td>
<td align="right">2050</td>
<td align="right">150</td>
<td align="right">144</td>
</tr>
</tbody>
</table>
<p>Ultimatelly, what we want to do is to compare gene expression between conditions. To do so we have to make some considerations. One is the fact that we want to make conslusions about the whole transciptome of the organism we are working with. This means around 12000 comparisons (assuming 12000 genes). This implies a huge multiple-comparison correction. Another important consideration is the difference in sequencing depth, to overcame this issue the reads are normalized using an stimate that includes the total amount of reads in that sample. Finally, genes that are too lowlly expressed or have a huge dispersion have to be removed from the analysis as they are too noisy to get a trustable conclusion from them.</p>
<p>To perform differential gene expression we have to build a model that explains the differences in gene expression using the condition we are testing, fit the data to that model and see if it can explain the differences. This is done using generalized linear models (GLM) modeled using a negative binomial distribution (NB). Why is this? Becasue the data are counts, ie. integers (there are not 12.5 reads, you either have 12 or 13), to model discrete counts, we use either poisson or NB distribution. NB is ussed because it allows for expression and variance to be unlinked.</p>
<p>Luckily for us, there are many packages dedicated to solve this problems. We will use <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DeSeq2</a></p>
<p>The steps performed by the DESeq function are documented in the manual page of DESeq2; briefly, they are:</p>
<p>1 estimation of size factors by estimateSizeFactors (this is for normalization)
2 estimation of dispersion by estimateDispersions (calculate the gene expression dispersion)
3 negative binomial GLM fitting and Wald statistics by nbinomWaldTest (this is the actuall differential gene expression step)</p>
<p>The only requirement for Deseq2 is to have the tables of data in a proper format and a meta-data object that indicates what is each sample.</p>
<p>Imagine we have a wildtype and a mutant, DeSeq2 requieres
Table: <strong>1: Data matrix with Gene name as ROWNAMES</strong></p>
<table>
<caption><strong>2: MetaData (colData) in which ROWNAMES are the COLNAMES of the Data matrix and the conditions are column names</strong></caption>
<thead>
<tr class="header">
<th align="right">Row Names</th>
<th align="right">WildType.1</th>
<th align="right">WildType.2</th>
<th align="right">Mutant.1</th>
<th align="right">Mutant.2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">A</td>
<td align="right">20</td>
<td align="right">400</td>
<td align="right">60</td>
<td align="right">40</td>
</tr>
<tr class="even">
<td align="right">B</td>
<td align="right">2</td>
<td align="right">50</td>
<td align="right">2000</td>
<td align="right">2500</td>
</tr>
<tr class="odd">
<td align="right">C</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">D</td>
<td align="right">5</td>
<td align="right">2050</td>
<td align="right">150</td>
<td align="right">144</td>
</tr>
</tbody>
</table>
<table>
<thead>
<tr class="header">
<th align="right">Row Names</th>
<th align="right">Genotype</th>
<th align="right">replicate</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">WildType.1</td>
<td align="right">WT</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="right">WildType.2</td>
<td align="right">WT</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="right">Mutant.1</td>
<td align="right">actinKD</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="right">Mutant.2</td>
<td align="right">actinKD</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<p>The we can use the Genotype as a condition to compare gene expresion for example.</p>
<p>At this time we will start by comparing gene expression at two timepoints in one celltype. This is good to have a first impresion of the data but it is not the way to go if your idea is to analyze circadian cycling behaviour.</p>
<div id="loading-the-tables" class="section level3" number="6.3.1">
<h3><span class="header-section-number">6.3.1</span> Loading the tables</h3>
<p>This time we have 4 gene expression tables: one for each cell type we are analyzing. Therefore, some of the processing will be done using for loops. This might be confusing at the beginning but is a really useful tool.</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="rna-seq-analysis.html#cb263-1"></a><span class="co">#We can list the files that are present in our working directory, or any path we specify. Remeber the path will change depending where you have your data.</span></span>
<span id="cb263-2"><a href="rna-seq-analysis.html#cb263-2"></a><span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;../RNAseq/&quot;</span>)</span>
<span id="cb263-3"><a href="rna-seq-analysis.html#cb263-3"></a><span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;../RNAseq/&quot;</span>,<span class="dt">pattern =</span> <span class="st">&quot;expression&quot;</span>) <span class="co">#what do you think patter is doing? remember to go to the help page if you need to.</span></span>
<span id="cb263-4"><a href="rna-seq-analysis.html#cb263-4"></a><span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;../RNAseq/&quot;</span>,<span class="dt">pattern =</span> <span class="st">&quot;expression&quot;</span>,<span class="dt">full.names =</span> T) <span class="co">#what happened now?</span></span>
<span id="cb263-5"><a href="rna-seq-analysis.html#cb263-5"></a></span>
<span id="cb263-6"><a href="rna-seq-analysis.html#cb263-6"></a>paper_tables&lt;-<span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;../RNAseq/&quot;</span>,<span class="dt">pattern =</span> <span class="st">&quot;xpression&quot;</span>,<span class="dt">full.names =</span> T) <span class="co">#what we are doing here?, check your enviroment, do you see something new?</span></span>
<span id="cb263-7"><a href="rna-seq-analysis.html#cb263-7"></a></span>
<span id="cb263-8"><a href="rna-seq-analysis.html#cb263-8"></a><span class="co">#If we try with one</span></span>
<span id="cb263-9"><a href="rna-seq-analysis.html#cb263-9"></a>TH_gene_expression&lt;-<span class="kw">read.table</span>(<span class="st">&quot;../RNAseq/TH_gene_expression.txt&quot;</span>,<span class="dt">header =</span> T)</span></code></pre></div>
<pre><code>##  [1] &quot;all_tables.rdata&quot;        &quot;circadianact.png&quot;       
##  [3] &quot;circadianbrain.png&quot;      &quot;circadiancells.png&quot;     
##  [5] &quot;cycDN1_reults.txt&quot;       &quot;cycDN1.csv&quot;             
##  [7] &quot;cycwithzt.png&quot;           &quot;DN1_gene_expression.txt&quot;
##  [9] &quot;F1.large.jpg&quot;            &quot;howcircadian.png&quot;       
## [11] &quot;LNd_gene_expression.txt&quot; &quot;LNv_gene_expression.txt&quot;
## [13] &quot;RNASeq_DeSeq2.html&quot;      &quot;RNASeq_DeSeq2.rmd&quot;      
## [15] &quot;RNAseqDataPrep.html&quot;     &quot;TableMod.rmd&quot;           
## [17] &quot;TH_gene_expression.txt&quot; 
## [1] &quot;DN1_gene_expression.txt&quot; &quot;LNd_gene_expression.txt&quot;
## [3] &quot;LNv_gene_expression.txt&quot; &quot;TH_gene_expression.txt&quot; 
## [1] &quot;../RNAseq//DN1_gene_expression.txt&quot; &quot;../RNAseq//LNd_gene_expression.txt&quot;
## [3] &quot;../RNAseq//LNv_gene_expression.txt&quot; &quot;../RNAseq//TH_gene_expression.txt&quot;</code></pre>
<p>How can we make it work for all the tables? We want to change the name, we basically want to remove the “.txt” ending. That can be done using gsub. We want to remove “.txt”, so we replace it for an empty string.</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="rna-seq-analysis.html#cb265-1"></a><span class="kw">gsub</span>(<span class="dt">x =</span> <span class="st">&quot;../RNAseq/TH_gene_expression.txt&quot;</span>,<span class="dt">pattern =</span><span class="st">&quot;.txt&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;../RNAseq/TH_gene_expression&quot;</code></pre>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="rna-seq-analysis.html#cb267-1"></a><span class="co">#We also need to remove the path file</span></span>
<span id="cb267-2"><a href="rna-seq-analysis.html#cb267-2"></a><span class="kw">gsub</span>(<span class="dt">x =</span> <span class="st">&quot;../RNAseq/TH_gene_expression.txt&quot;</span>,<span class="dt">pattern =</span><span class="st">&quot;../RNAseq/&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;TH_gene_expression.txt&quot;</code></pre>
<p>Now we can try to do it for all of them, just trying to see if changing the names is working but NOT applying it to anything until we know it works.</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="rna-seq-analysis.html#cb269-1"></a><span class="cf">for</span> (e <span class="cf">in</span> (paper_tables)){</span>
<span id="cb269-2"><a href="rna-seq-analysis.html#cb269-2"></a>  <span class="kw">print</span>(<span class="st">&quot;original name:          &quot;</span>)</span>
<span id="cb269-3"><a href="rna-seq-analysis.html#cb269-3"></a>  <span class="kw">print</span>(e) <span class="co"># this is printing e</span></span>
<span id="cb269-4"><a href="rna-seq-analysis.html#cb269-4"></a>  n&lt;-<span class="kw">gsub</span>(<span class="dt">pattern =</span><span class="st">&quot;.txt&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>,<span class="dt">x =</span> e)</span>
<span id="cb269-5"><a href="rna-seq-analysis.html#cb269-5"></a>  n&lt;-<span class="kw">gsub</span>(<span class="dt">pattern =</span><span class="st">&quot;../RNAseq//&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>,<span class="dt">x =</span> n)</span>
<span id="cb269-6"><a href="rna-seq-analysis.html#cb269-6"></a>  <span class="kw">print</span>(<span class="st">&quot;final name:        &quot;</span>)</span>
<span id="cb269-7"><a href="rna-seq-analysis.html#cb269-7"></a>  <span class="kw">print</span>(n)</span>
<span id="cb269-8"><a href="rna-seq-analysis.html#cb269-8"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;original name:          &quot;
## [1] &quot;../RNAseq//DN1_gene_expression.txt&quot;
## [1] &quot;final name:        &quot;
## [1] &quot;DN1_gene_expression&quot;
## [1] &quot;original name:          &quot;
## [1] &quot;../RNAseq//LNd_gene_expression.txt&quot;
## [1] &quot;final name:        &quot;
## [1] &quot;LNd_gene_expression&quot;
## [1] &quot;original name:          &quot;
## [1] &quot;../RNAseq//LNv_gene_expression.txt&quot;
## [1] &quot;final name:        &quot;
## [1] &quot;LNv_gene_expression&quot;
## [1] &quot;original name:          &quot;
## [1] &quot;../RNAseq//TH_gene_expression.txt&quot;
## [1] &quot;final name:        &quot;
## [1] &quot;TH_gene_expression&quot;</code></pre>
<blockquote>
<p>What is this for loop doing? Do you think it is working? Why?</p>
</blockquote>
<p>Now we will read the tables and put them in the names we created. To do this we will use the function assign, that actually assign anything to a given name. The names is stored in “n”. What we want to assign is the “read.delim” of the file name stored in e.</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="rna-seq-analysis.html#cb271-1"></a><span class="cf">for</span> (e <span class="cf">in</span> (paper_tables)){</span>
<span id="cb271-2"><a href="rna-seq-analysis.html#cb271-2"></a>  n&lt;-<span class="kw">gsub</span>(<span class="dt">pattern =</span><span class="st">&quot;.txt&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>,<span class="dt">x =</span> e)</span>
<span id="cb271-3"><a href="rna-seq-analysis.html#cb271-3"></a>  n&lt;-<span class="kw">gsub</span>(<span class="dt">pattern =</span><span class="st">&quot;../RNAseq//&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>,<span class="dt">x =</span> n)</span>
<span id="cb271-4"><a href="rna-seq-analysis.html#cb271-4"></a>  <span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&quot;I am reading: &quot;</span>,n))</span>
<span id="cb271-5"><a href="rna-seq-analysis.html#cb271-5"></a>  <span class="kw">assign</span>(n,<span class="kw">read.delim</span>(e))</span>
<span id="cb271-6"><a href="rna-seq-analysis.html#cb271-6"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;I am reading: DN1_gene_expression&quot;
## [1] &quot;I am reading: LNd_gene_expression&quot;
## [1] &quot;I am reading: LNv_gene_expression&quot;
## [1] &quot;I am reading: TH_gene_expression&quot;</code></pre>
</div>
<div id="create-the-deseq2-count-matrix" class="section level3" number="6.3.2">
<h3><span class="header-section-number">6.3.2</span> Create the DeSeq2 count matrix</h3>
<p>DeSeq2 expects to have a data.frame with the gene-names as <code>row.names</code> and then just the counts.</p>
<p>I will first show it for one of the tables and then do a for loop to do it for all of them.</p>
<div class="sourceCode" id="cb273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="rna-seq-analysis.html#cb273-1"></a><span class="co">#We do not want to overwrite the raw tables. So, we will create new objects with the same data and we will modify these new objects and preserve the original data. Is like doing &quot;save as&quot; in word.</span></span>
<span id="cb273-2"><a href="rna-seq-analysis.html#cb273-2"></a></span>
<span id="cb273-3"><a href="rna-seq-analysis.html#cb273-3"></a>ds2_DN1_gene_expression_try&lt;-DN1_gene_expression <span class="co">#what is this doing?</span></span>
<span id="cb273-4"><a href="rna-seq-analysis.html#cb273-4"></a></span>
<span id="cb273-5"><a href="rna-seq-analysis.html#cb273-5"></a><span class="co">#We do now the loop for all of them to create the `ds2_` objects</span></span>
<span id="cb273-6"><a href="rna-seq-analysis.html#cb273-6"></a><span class="kw">ls</span>(<span class="dt">pattern =</span> <span class="st">&quot;gene_expression&quot;</span>) <span class="co">#this list the objects in the environment that has the selected pattern in its name</span></span></code></pre></div>
<pre><code>## [1] &quot;DN1_gene_expression&quot;         &quot;ds2_DN1_gene_expression_try&quot;
## [3] &quot;LNd_gene_expression&quot;         &quot;LNv_gene_expression&quot;        
## [5] &quot;TH_gene_expression&quot;</code></pre>
<div class="sourceCode" id="cb275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb275-1"><a href="rna-seq-analysis.html#cb275-1"></a><span class="co">#The new name we want is just adding &quot;ds2_&quot; to the begining of the object name. We can do this with the paste0 funciton.</span></span>
<span id="cb275-2"><a href="rna-seq-analysis.html#cb275-2"></a><span class="kw">paste0</span>(<span class="st">&quot;ds2_&quot;</span>,<span class="st">&quot;DN1_gene_expression&quot;</span>)</span>
<span id="cb275-3"><a href="rna-seq-analysis.html#cb275-3"></a></span>
<span id="cb275-4"><a href="rna-seq-analysis.html#cb275-4"></a><span class="cf">for</span> (e <span class="cf">in</span> <span class="kw">ls</span>(<span class="dt">pattern =</span> <span class="st">&quot;gene_expression&quot;</span>)){ </span>
<span id="cb275-5"><a href="rna-seq-analysis.html#cb275-5"></a>  x =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;ds2&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>,e) <span class="co">#this is the new name</span></span>
<span id="cb275-6"><a href="rna-seq-analysis.html#cb275-6"></a>  <span class="kw">print</span>(<span class="st">&quot;new name is:&quot;</span>)</span>
<span id="cb275-7"><a href="rna-seq-analysis.html#cb275-7"></a>  <span class="kw">print</span>(x) </span>
<span id="cb275-8"><a href="rna-seq-analysis.html#cb275-8"></a>  <span class="kw">assign</span>(x, <span class="kw">get</span>(e)) <span class="co">#get is a function that extracts the data from an object name</span></span>
<span id="cb275-9"><a href="rna-seq-analysis.html#cb275-9"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;ds2_DN1_gene_expression&quot;
## [1] &quot;new name is:&quot;
## [1] &quot;ds2_DN1_gene_expression&quot;
## [1] &quot;new name is:&quot;
## [1] &quot;ds2_ds2_DN1_gene_expression_try&quot;
## [1] &quot;new name is:&quot;
## [1] &quot;ds2_LNd_gene_expression&quot;
## [1] &quot;new name is:&quot;
## [1] &quot;ds2_LNv_gene_expression&quot;
## [1] &quot;new name is:&quot;
## [1] &quot;ds2_TH_gene_expression&quot;</code></pre>
<p>Now we have to do some manipulation in the tables for them to work with Deseq2.</p>
<div class="sourceCode" id="cb277"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb277-1"><a href="rna-seq-analysis.html#cb277-1"></a>ds2_DN1_gene_expression_try &lt;-<span class="st"> </span>ds2_DN1_gene_expression</span>
<span id="cb277-2"><a href="rna-seq-analysis.html#cb277-2"></a><span class="kw">row.names</span>(ds2_DN1_gene_expression_try) =<span class="st"> </span>ds2_DN1_gene_expression<span class="op">$</span>Symbol <span class="co"># we put the symbols as row.names</span></span>
<span id="cb277-3"><a href="rna-seq-analysis.html#cb277-3"></a>ds2_DN1_gene_expression_try&lt;-ds2_DN1_gene_expression_try[,<span class="op">-</span><span class="dv">1</span>] <span class="co"># we take the symbols column out</span></span>
<span id="cb277-4"><a href="rna-seq-analysis.html#cb277-4"></a><span class="kw">head</span>(ds2_DN1_gene_expression_try) <span class="co"># we can see that now the table is what we need</span></span>
<span id="cb277-5"><a href="rna-seq-analysis.html#cb277-5"></a><span class="kw">rm</span>(ds2_DN1_gene_expression_try) <span class="co">#remove it becasue you do not need it anymore and because it can create problems</span></span></code></pre></div>
<pre><code>##             DN1_ZT3_1 DN1_ZT7_1 DN1_ZT11_1 DN1_ZT15_1 DN1_ZT19_1 DN1_ZT23_1
## FBtr0070202         0         0          0          0          0          0
## FBtr0070207         0         0          0          0         14          0
## FBtr0070238         0         0          0          0         21          0
## FBtr0070251         0         0          0          0          0          0
## FBtr0070484         0         0          0          0          0          0
## FBtr0070489         0         0          0          0          0          0
##             DN1_ZT3_2 DN1_ZT7_2 DN1_ZT11_2 DN1_ZT15_2 DN1_ZT19_2 DN1_ZT23_2
## FBtr0070202         0         0          0          0          0          0
## FBtr0070207         0         0          0          0         16          7
## FBtr0070238         0         0          0          0         15          9
## FBtr0070251         0         0          1          0          2          0
## FBtr0070484         0         0          0          0          0          1
## FBtr0070489         0         0          0          0          0          0</code></pre>
<p>We can add this step to the for loop, or even better we can create a funciton and practice something new in R.</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="rna-seq-analysis.html#cb279-1"></a><span class="co">#We create a small funciton to remember</span></span>
<span id="cb279-2"><a href="rna-seq-analysis.html#cb279-2"></a>sum2&lt;-<span class="cf">function</span>(x){</span>
<span id="cb279-3"><a href="rna-seq-analysis.html#cb279-3"></a> y=x<span class="op">+</span><span class="dv">2</span> </span>
<span id="cb279-4"><a href="rna-seq-analysis.html#cb279-4"></a> <span class="kw">return</span>(y)</span>
<span id="cb279-5"><a href="rna-seq-analysis.html#cb279-5"></a>}</span>
<span id="cb279-6"><a href="rna-seq-analysis.html#cb279-6"></a></span>
<span id="cb279-7"><a href="rna-seq-analysis.html#cb279-7"></a><span class="kw">sum2</span>(<span class="dv">4</span>) </span></code></pre></div>
<pre><code>## [1] 6</code></pre>
<blockquote>
<p>what is this doing? why?</p>
</blockquote>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="rna-seq-analysis.html#cb281-1"></a><span class="co">#Now we create a function that takes the data frame and does all the transformation we did before</span></span>
<span id="cb281-2"><a href="rna-seq-analysis.html#cb281-2"></a>create_ds2=<span class="st"> </span><span class="cf">function</span>(df_name){</span>
<span id="cb281-3"><a href="rna-seq-analysis.html#cb281-3"></a>  df =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">get</span>(df_name))</span>
<span id="cb281-4"><a href="rna-seq-analysis.html#cb281-4"></a>  <span class="kw">row.names</span>(df) =<span class="st"> </span><span class="kw">as.character</span>(df[,<span class="dv">1</span>])</span>
<span id="cb281-5"><a href="rna-seq-analysis.html#cb281-5"></a>  df=df[,<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb281-6"><a href="rna-seq-analysis.html#cb281-6"></a>  <span class="kw">return</span>(df) <span class="co"># this actually tells the function to return the transformed data.frame</span></span>
<span id="cb281-7"><a href="rna-seq-analysis.html#cb281-7"></a>}</span>
<span id="cb281-8"><a href="rna-seq-analysis.html#cb281-8"></a></span>
<span id="cb281-9"><a href="rna-seq-analysis.html#cb281-9"></a><span class="co">#Let&#39;s try the function in one of the ds2 objects</span></span>
<span id="cb281-10"><a href="rna-seq-analysis.html#cb281-10"></a><span class="co">#head(create_ds2(ds2_LNd_gene_expression))[,c(1:3)]</span></span>
<span id="cb281-11"><a href="rna-seq-analysis.html#cb281-11"></a><span class="cf">for</span> (e <span class="cf">in</span> <span class="kw">ls</span>(<span class="dt">pattern =</span> <span class="st">&quot;ds2_&quot;</span>)){</span>
<span id="cb281-12"><a href="rna-seq-analysis.html#cb281-12"></a>  <span class="kw">print</span>(e)</span>
<span id="cb281-13"><a href="rna-seq-analysis.html#cb281-13"></a>  <span class="kw">assign</span>(e,<span class="kw">create_ds2</span>(e)) <span class="co">#this should actually run the function and apply it to the data frame</span></span>
<span id="cb281-14"><a href="rna-seq-analysis.html#cb281-14"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;ds2_DN1_gene_expression&quot;
## [1] &quot;ds2_ds2_DN1_gene_expression_try&quot;
## [1] &quot;ds2_LNd_gene_expression&quot;
## [1] &quot;ds2_LNv_gene_expression&quot;
## [1] &quot;ds2_TH_gene_expression&quot;</code></pre>
</div>
<div id="coldata-preparation" class="section level3 tabset tabset-fade tabset-pills" number="6.3.3">
<h3><span class="header-section-number">6.3.3</span> colData preparation</h3>
<p>We now need to have a data.frame with the mapping between the columns and the ZT, the program call this colData. Is what is usually called “meta data”. Is a file that explains what each object/column is.</p>
<p>We will do it by hand just changing the names. No more loops for now. After you execute each step I recommend you to inspect the object to see what is happening.</p>
<p><strong>DN1</strong></p>
<div class="sourceCode" id="cb283"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb283-1"><a href="rna-seq-analysis.html#cb283-1"></a><span class="co">#takes the names in the data frame and put them in the new objects. We want to get the information for each </span></span>
<span id="cb283-2"><a href="rna-seq-analysis.html#cb283-2"></a>colData.DN1 =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">names</span>(ds2_DN1_gene_expression)) </span>
<span id="cb283-3"><a href="rna-seq-analysis.html#cb283-3"></a><span class="co">#now we put them as row.names</span></span>
<span id="cb283-4"><a href="rna-seq-analysis.html#cb283-4"></a><span class="kw">rownames</span>(colData.DN1) =<span class="st"> </span>colData.DN1[,<span class="dv">1</span>]</span>
<span id="cb283-5"><a href="rna-seq-analysis.html#cb283-5"></a><span class="co">#now we take the columns out </span></span>
<span id="cb283-6"><a href="rna-seq-analysis.html#cb283-6"></a>colData.DN1 =colData.DN1[,<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb283-7"><a href="rna-seq-analysis.html#cb283-7"></a><span class="co">#this is taking the last part of the rownames and put them in a column named `ZT`</span></span>
<span id="cb283-8"><a href="rna-seq-analysis.html#cb283-8"></a>colData.DN1<span class="op">$</span>ZT =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.DN1),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">2</span>)</span>
<span id="cb283-9"><a href="rna-seq-analysis.html#cb283-9"></a>colData.DN1<span class="op">$</span>rep =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.DN1),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">3</span>)</span>
<span id="cb283-10"><a href="rna-seq-analysis.html#cb283-10"></a></span>
<span id="cb283-11"><a href="rna-seq-analysis.html#cb283-11"></a><span class="kw">head</span>(colData.DN1)</span></code></pre></div>
<pre><code>##              ZT rep
## DN1_ZT3_1   ZT3   1
## DN1_ZT7_1   ZT7   1
## DN1_ZT11_1 ZT11   1
## DN1_ZT15_1 ZT15   1
## DN1_ZT19_1 ZT19   1
## DN1_ZT23_1 ZT23   1</code></pre>
<p><strong>LNv</strong></p>
<div class="sourceCode" id="cb285"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb285-1"><a href="rna-seq-analysis.html#cb285-1"></a>colData.LNv =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">names</span>(ds2_LNv_gene_expression)) </span>
<span id="cb285-2"><a href="rna-seq-analysis.html#cb285-2"></a><span class="co">#now we put them as row.names</span></span>
<span id="cb285-3"><a href="rna-seq-analysis.html#cb285-3"></a><span class="kw">rownames</span>(colData.LNv) =<span class="st"> </span>colData.LNv[,<span class="dv">1</span>]</span>
<span id="cb285-4"><a href="rna-seq-analysis.html#cb285-4"></a><span class="co">#now we take the columns out </span></span>
<span id="cb285-5"><a href="rna-seq-analysis.html#cb285-5"></a>colData.LNv =colData.LNv[,<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb285-6"><a href="rna-seq-analysis.html#cb285-6"></a><span class="co">#this is taking the last part of the rownames and put them in a column named `ZT`</span></span>
<span id="cb285-7"><a href="rna-seq-analysis.html#cb285-7"></a>colData.LNv<span class="op">$</span>ZT =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.LNv),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">2</span>)</span>
<span id="cb285-8"><a href="rna-seq-analysis.html#cb285-8"></a>colData.LNv<span class="op">$</span>rep =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.LNv),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">3</span>)</span></code></pre></div>
<p><strong>LNd</strong></p>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="rna-seq-analysis.html#cb286-1"></a>colData.LNd =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">names</span>(ds2_LNd_gene_expression)) </span>
<span id="cb286-2"><a href="rna-seq-analysis.html#cb286-2"></a><span class="co">#now we put them as row.names</span></span>
<span id="cb286-3"><a href="rna-seq-analysis.html#cb286-3"></a><span class="kw">rownames</span>(colData.LNd) =<span class="st"> </span>colData.LNd[,<span class="dv">1</span>]</span>
<span id="cb286-4"><a href="rna-seq-analysis.html#cb286-4"></a><span class="co">#now we take the columns out </span></span>
<span id="cb286-5"><a href="rna-seq-analysis.html#cb286-5"></a>colData.LNd =colData.LNd[,<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb286-6"><a href="rna-seq-analysis.html#cb286-6"></a><span class="co">#this is taking the last part of the rownames and put them in a column named `ZT`</span></span>
<span id="cb286-7"><a href="rna-seq-analysis.html#cb286-7"></a>colData.LNd<span class="op">$</span>ZT =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.LNd),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">2</span>)</span>
<span id="cb286-8"><a href="rna-seq-analysis.html#cb286-8"></a>colData.LNd<span class="op">$</span>rep =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.LNd),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">3</span>)</span></code></pre></div>
<p><strong>TH</strong></p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="rna-seq-analysis.html#cb287-1"></a>colData.TH =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">names</span>(ds2_TH_gene_expression)) </span>
<span id="cb287-2"><a href="rna-seq-analysis.html#cb287-2"></a><span class="co">#now we put them as row.names</span></span>
<span id="cb287-3"><a href="rna-seq-analysis.html#cb287-3"></a><span class="kw">rownames</span>(colData.TH) =<span class="st"> </span>colData.TH[,<span class="dv">1</span>]</span>
<span id="cb287-4"><a href="rna-seq-analysis.html#cb287-4"></a><span class="co">#now we take the columns out and </span></span>
<span id="cb287-5"><a href="rna-seq-analysis.html#cb287-5"></a>colData.TH =colData.TH[,<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb287-6"><a href="rna-seq-analysis.html#cb287-6"></a>colData.TH<span class="op">$</span>ZT =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.TH),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">2</span>)</span>
<span id="cb287-7"><a href="rna-seq-analysis.html#cb287-7"></a>colData.TH<span class="op">$</span>rep =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.TH),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">3</span>)</span></code></pre></div>
</div>
<div id="heat-map-pca-analysis" class="section level3 tabset tabset-fade tabset-pills" number="6.3.4">
<h3><span class="header-section-number">6.3.4</span> Heat-map PCA analysis</h3>
<p>Now we will initialize the data and see the differences between them.</p>
<p>We first need to initiate the object. This imply telling the package which is the matrix of read counts, the colData (meta deta) and the design. The design is what the program uses to do the differential expression. Is simple in this case: only the ZT.</p>
<p>Then we will do a HeatMap and a principal component analysis (PCA) to see the differences between samples. This is a regular procedure in differential gene expression analysis as it helps us to identify samples that might be outliers.</p>
<p><strong>DN1</strong></p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="rna-seq-analysis.html#cb288-1"></a><span class="co">#This is actually the main function of the package and prepares everything to run the differential expression analysis.</span></span>
<span id="cb288-2"><a href="rna-seq-analysis.html#cb288-2"></a>dds.DN1 &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> <span class="kw">as.matrix</span>(ds2_DN1_gene_expression), <span class="dt">colData =</span> colData.DN1,<span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>ZT)</span>
<span id="cb288-3"><a href="rna-seq-analysis.html#cb288-3"></a></span>
<span id="cb288-4"><a href="rna-seq-analysis.html#cb288-4"></a><span class="co">#class of dds.DN1</span></span>
<span id="cb288-5"><a href="rna-seq-analysis.html#cb288-5"></a><span class="kw">class</span>(dds.DN1)</span></code></pre></div>
<pre><code>## [1] &quot;DESeqDataSet&quot;
## attr(,&quot;package&quot;)
## [1] &quot;DESeq2&quot;</code></pre>
<p>We will first create the heat map with the unnormalized counts.</p>
<p>First we need the distance matrix. A distance matrix is a way to quantify differences. In this case each column of the data matrix will be compared with all the others.
You can read more <a href="https://www.displayr.com/what-is-a-distance-matrix/">here</a> or <a href="http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/120-correspondence-analysis-theory-and-practice/">here</a></p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="rna-seq-analysis.html#cb290-1"></a>sampleDists &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">t</span>(<span class="kw">assay</span>(dds.DN1)))</span>
<span id="cb290-2"><a href="rna-seq-analysis.html#cb290-2"></a></span>
<span id="cb290-3"><a href="rna-seq-analysis.html#cb290-3"></a><span class="co">#Then we take this in a matrix and create the heatmap</span></span>
<span id="cb290-4"><a href="rna-seq-analysis.html#cb290-4"></a>sampleDistMatrix &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(sampleDists)</span>
<span id="cb290-5"><a href="rna-seq-analysis.html#cb290-5"></a><span class="kw">rownames</span>(sampleDistMatrix) &lt;-<span class="st"> </span><span class="kw">paste</span>(dds.DN1<span class="op">$</span>ZT,dds.DN1<span class="op">$</span>rep,<span class="dt">sep=</span><span class="st">&quot;_rep_&quot;</span>)</span>
<span id="cb290-6"><a href="rna-seq-analysis.html#cb290-6"></a><span class="kw">colnames</span>(sampleDistMatrix) &lt;-<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb290-7"><a href="rna-seq-analysis.html#cb290-7"></a>colors &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>( <span class="kw">rev</span>(<span class="kw">brewer.pal</span>(<span class="dv">9</span>, <span class="st">&quot;Blues&quot;</span>)) )(<span class="dv">255</span>)<span class="co">#select color </span></span>
<span id="cb290-8"><a href="rna-seq-analysis.html#cb290-8"></a><span class="co">#Heatmap fucntion</span></span>
<span id="cb290-9"><a href="rna-seq-analysis.html#cb290-9"></a><span class="kw">pheatmap</span>(sampleDistMatrix,</span>
<span id="cb290-10"><a href="rna-seq-analysis.html#cb290-10"></a>         <span class="dt">clustering_distance_rows=</span>sampleDists,</span>
<span id="cb290-11"><a href="rna-seq-analysis.html#cb290-11"></a>         <span class="dt">clustering_distance_cols=</span>sampleDists,</span>
<span id="cb290-12"><a href="rna-seq-analysis.html#cb290-12"></a>                  <span class="dt">annotation_row =</span> <span class="kw">rownames</span>(sampleDists),</span>
<span id="cb290-13"><a href="rna-seq-analysis.html#cb290-13"></a>         <span class="dt">annotation_col =</span> <span class="kw">rownames</span>(sampleDists),</span>
<span id="cb290-14"><a href="rna-seq-analysis.html#cb290-14"></a>         <span class="dt">cutree_rows =</span> <span class="dv">2</span>,</span>
<span id="cb290-15"><a href="rna-seq-analysis.html#cb290-15"></a>         <span class="dt">cutree_cols =</span> <span class="dv">2</span>,</span>
<span id="cb290-16"><a href="rna-seq-analysis.html#cb290-16"></a>         <span class="dt">col=</span>colors)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-21"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-21-1.png" alt="Heatmap representing distance between samples" width="100%" />
<p class="caption">
Figure 6.1: Heatmap representing distance between samples
</p>
</div>
<p>Now we do a logaritmig transformation to see how the data looks like after minimizing differences between samples for rows with small counts, and which normalizes with respect to library size. You can read the help excecuting <code>?rlog</code>.</p>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="rna-seq-analysis.html#cb291-1"></a>rdl.DN1 &lt;-<span class="st"> </span><span class="kw">rlog</span>(dds.DN1) </span></code></pre></div>
<p>now we can plot the PCA analysis</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="rna-seq-analysis.html#cb292-1"></a><span class="kw">plotPCA</span>(rdl.DN1, <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-23"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-23-1.png" alt="PCA coloring by ZT" width="100%" />
<p class="caption">
Figure 6.2: PCA coloring by ZT
</p>
</div>
<p>And also we can work with the plot by taking the data out and using ggplot2:</p>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="rna-seq-analysis.html#cb293-1"></a>data &lt;-<span class="st"> </span><span class="kw">plotPCA</span>(rdl.DN1, <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>), <span class="dt">returnData=</span><span class="ot">TRUE</span>)<span class="co">#extract the data</span></span>
<span id="cb293-2"><a href="rna-seq-analysis.html#cb293-2"></a>percentVar &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="dv">100</span> <span class="op">*</span><span class="st"> </span><span class="kw">attr</span>(data, <span class="st">&quot;percentVar&quot;</span>)) <span class="co">#calculate the percentage of variation explained by each PC</span></span>
<span id="cb293-3"><a href="rna-seq-analysis.html#cb293-3"></a></span>
<span id="cb293-4"><a href="rna-seq-analysis.html#cb293-4"></a><span class="kw">ggplot</span>(data, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>name)) <span class="op">+</span></span>
<span id="cb293-5"><a href="rna-seq-analysis.html#cb293-5"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb293-6"><a href="rna-seq-analysis.html#cb293-6"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> name),</span>
<span id="cb293-7"><a href="rna-seq-analysis.html#cb293-7"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb293-8"><a href="rna-seq-analysis.html#cb293-8"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb293-9"><a href="rna-seq-analysis.html#cb293-9"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb293-10"><a href="rna-seq-analysis.html#cb293-10"></a><span class="st">  </span><span class="kw">theme_minimal</span>()<span class="op">+</span></span>
<span id="cb293-11"><a href="rna-seq-analysis.html#cb293-11"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC1: &quot;</span>,percentVar[<span class="dv">1</span>],<span class="st">&quot;% variance&quot;</span>)) <span class="op">+</span></span>
<span id="cb293-12"><a href="rna-seq-analysis.html#cb293-12"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC2: &quot;</span>,percentVar[<span class="dv">2</span>],<span class="st">&quot;% variance&quot;</span>)) <span class="op">+</span></span>
<span id="cb293-13"><a href="rna-seq-analysis.html#cb293-13"></a><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title=</span><span class="st">&quot;PC1 vs PC2&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-24"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-24-1.png" alt="PCA coloring by ZT adding label by name" width="100%" />
<p class="caption">
Figure 5.2: PCA coloring by ZT adding label by name
</p>
</div>
<p>This analyze how much variation are we seeing in each principal component (PC)</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="rna-seq-analysis.html#cb294-1"></a><span class="co">#We can even create 3D plots for all the PC we want</span></span>
<span id="cb294-2"><a href="rna-seq-analysis.html#cb294-2"></a>ntop &lt;-<span class="st"> </span><span class="dv">1000</span></span>
<span id="cb294-3"><a href="rna-seq-analysis.html#cb294-3"></a>rv &lt;-<span class="st"> </span><span class="kw">rowVars</span>(<span class="kw">assay</span>(rdl.DN1))</span>
<span id="cb294-4"><a href="rna-seq-analysis.html#cb294-4"></a>select &lt;-<span class="st"> </span><span class="kw">order</span>(rv, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="kw">seq_len</span>(<span class="kw">min</span>(ntop, <span class="kw">length</span>(rv)))]</span>
<span id="cb294-5"><a href="rna-seq-analysis.html#cb294-5"></a>mat &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">assay</span>(rdl.DN1)[select, ])</span>
<span id="cb294-6"><a href="rna-seq-analysis.html#cb294-6"></a>pc &lt;-<span class="st"> </span><span class="kw">prcomp</span>(mat)</span>
<span id="cb294-7"><a href="rna-seq-analysis.html#cb294-7"></a></span>
<span id="cb294-8"><a href="rna-seq-analysis.html#cb294-8"></a><span class="kw">fviz_eig</span>(pc, <span class="dt">addlabels =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-25"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-25-1.png" alt="Variance explained by each PC" width="100%" />
<p class="caption">
Figure 5.3: Variance explained by each PC
</p>
</div>
<p>Now we can plot more than 2 PCs</p>
<div class="sourceCode" id="cb295"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="rna-seq-analysis.html#cb295-1"></a>pc_df=<span class="kw">as.data.frame</span>(pc<span class="op">$</span>x)</span>
<span id="cb295-2"><a href="rna-seq-analysis.html#cb295-2"></a>pc_df=<span class="kw">merge</span>(pc_df,colData.DN1,<span class="dt">by=</span><span class="st">&quot;row.names&quot;</span>,<span class="dt">all=</span>T)</span>
<span id="cb295-3"><a href="rna-seq-analysis.html#cb295-3"></a><span class="kw">rownames</span>(pc_df)=pc_df<span class="op">$</span>Row.names</span>
<span id="cb295-4"><a href="rna-seq-analysis.html#cb295-4"></a></span>
<span id="cb295-5"><a href="rna-seq-analysis.html#cb295-5"></a></span>
<span id="cb295-6"><a href="rna-seq-analysis.html#cb295-6"></a>p1=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT)) <span class="op">+</span></span>
<span id="cb295-7"><a href="rna-seq-analysis.html#cb295-7"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>)  <span class="op">+</span></span>
<span id="cb295-8"><a href="rna-seq-analysis.html#cb295-8"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb295-9"><a href="rna-seq-analysis.html#cb295-9"></a></span>
<span id="cb295-10"><a href="rna-seq-analysis.html#cb295-10"></a></span>
<span id="cb295-11"><a href="rna-seq-analysis.html#cb295-11"></a>p2=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC3, PC2, <span class="dt">color=</span>ZT)) <span class="op">+</span></span>
<span id="cb295-12"><a href="rna-seq-analysis.html#cb295-12"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb295-13"><a href="rna-seq-analysis.html#cb295-13"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb295-14"><a href="rna-seq-analysis.html#cb295-14"></a></span>
<span id="cb295-15"><a href="rna-seq-analysis.html#cb295-15"></a></span>
<span id="cb295-16"><a href="rna-seq-analysis.html#cb295-16"></a><span class="co">#and both together</span></span>
<span id="cb295-17"><a href="rna-seq-analysis.html#cb295-17"></a><span class="kw">grid.arrange</span>(p1,p2, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-261"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-26-1.png" alt="Ploting many principal components" width="100%" />
<p class="caption">
Figure 6.3: Ploting many principal components
</p>
</div>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="rna-seq-analysis.html#cb296-1"></a>p1=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>Row.names)) <span class="op">+</span></span>
<span id="cb296-2"><a href="rna-seq-analysis.html#cb296-2"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb296-3"><a href="rna-seq-analysis.html#cb296-3"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Row.names),</span>
<span id="cb296-4"><a href="rna-seq-analysis.html#cb296-4"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb296-5"><a href="rna-seq-analysis.html#cb296-5"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb296-6"><a href="rna-seq-analysis.html#cb296-6"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb296-7"><a href="rna-seq-analysis.html#cb296-7"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb296-8"><a href="rna-seq-analysis.html#cb296-8"></a></span>
<span id="cb296-9"><a href="rna-seq-analysis.html#cb296-9"></a></span>
<span id="cb296-10"><a href="rna-seq-analysis.html#cb296-10"></a>p2=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC3, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>Row.names)) <span class="op">+</span></span>
<span id="cb296-11"><a href="rna-seq-analysis.html#cb296-11"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb296-12"><a href="rna-seq-analysis.html#cb296-12"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Row.names),</span>
<span id="cb296-13"><a href="rna-seq-analysis.html#cb296-13"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb296-14"><a href="rna-seq-analysis.html#cb296-14"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb296-15"><a href="rna-seq-analysis.html#cb296-15"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb296-16"><a href="rna-seq-analysis.html#cb296-16"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb296-17"><a href="rna-seq-analysis.html#cb296-17"></a></span>
<span id="cb296-18"><a href="rna-seq-analysis.html#cb296-18"></a></span>
<span id="cb296-19"><a href="rna-seq-analysis.html#cb296-19"></a><span class="co">#and both together</span></span>
<span id="cb296-20"><a href="rna-seq-analysis.html#cb296-20"></a><span class="kw">grid.arrange</span>(p1,p2, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-262"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-26-2.png" alt="Ploting many principal components" width="100%" />
<p class="caption">
Figure 6.4: Ploting many principal components
</p>
</div>
<pre><code>#We can do some fancy 3D plots also
plot_ly(pc_df, x = ~PC1, y = ~PC2, z = ~PC3, color = ~ZT,colorscale = c(&#39;#FFE1A1&#39;, &#39;#683531&#39;))

plot_ly(pc_df, x = ~PC2, y = ~PC3, z = ~PC4, color = ~ZT,colorscale = c(&#39;#FFE1A1&#39;, &#39;#683531&#39;))
</code></pre>
</div>
<div id="differential-gene-expression-analysis-between-zt." class="section level3 tabset tabset-fade tabset-pills" number="6.3.5">
<h3><span class="header-section-number">6.3.5</span> Differential gene expression analysis between ZT.</h3>
<p>We already saw that for DN1: ZT19 and ZT23 looks more similar between them and also ZT3, 7, 15 and 11. We will now explore particular differences. <code>DESeq</code> is the main function here it acts over the object created previously. It executes the normalization and differential gene expression analysis.</p>
<p><strong>DN1</strong></p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="rna-seq-analysis.html#cb298-1"></a>dds.DN1 =<span class="st"> </span><span class="kw">DESeq</span>(dds.DN1)</span></code></pre></div>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="rna-seq-analysis.html#cb305-1"></a><span class="co">#The resultNames is taking the names of the comparisons done by the DESeq function</span></span>
<span id="cb305-2"><a href="rna-seq-analysis.html#cb305-2"></a><span class="kw">resultsNames</span>(dds.DN1)</span></code></pre></div>
<pre><code>## [1] &quot;Intercept&quot;       &quot;ZT_ZT15_vs_ZT11&quot; &quot;ZT_ZT19_vs_ZT11&quot; &quot;ZT_ZT23_vs_ZT11&quot;
## [5] &quot;ZT_ZT3_vs_ZT11&quot;  &quot;ZT_ZT7_vs_ZT11&quot;</code></pre>
<p>The <code>results</code> function extract this result, we will compare ZT3 vs ZT15. Use ?results to explore this function further.</p>
<div class="sourceCode" id="cb307"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb307-1"><a href="rna-seq-analysis.html#cb307-1"></a>res.ZT3vsZT15.DN1 &lt;-<span class="st"> </span><span class="kw">results</span>( dds.DN1, <span class="dt">contrast =</span> <span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>,<span class="st">&quot;ZT3&quot;</span>, <span class="st">&quot;ZT15&quot;</span>),<span class="dt">alpha=</span><span class="fl">0.05</span>)</span>
<span id="cb307-2"><a href="rna-seq-analysis.html#cb307-2"></a></span>
<span id="cb307-3"><a href="rna-seq-analysis.html#cb307-3"></a><span class="kw">head</span>(res.ZT3vsZT15.DN1)</span></code></pre></div>
<pre><code>## log2 fold change (MLE): ZT ZT3 vs ZT15 
## Wald test p-value: ZT ZT3 vs ZT15 
## DataFrame with 6 rows and 6 columns
##                      baseMean log2FoldChange            lfcSE      stat
##                     &lt;numeric&gt;      &lt;numeric&gt;        &lt;numeric&gt; &lt;numeric&gt;
## FBtr0070202                 0             NA               NA        NA
## FBtr0070207   2.1168359707544              0 5.24448449612991         0
## FBtr0070238  2.68357323289241              0  4.8861753047644         0
## FBtr0070251 0.304433817751997              0 5.39799191011387         0
## FBtr0070484 0.151419583982656              0 5.39799191011387         0
## FBtr0070489                 0             NA               NA        NA
##                pvalue      padj
##             &lt;numeric&gt; &lt;numeric&gt;
## FBtr0070202        NA        NA
## FBtr0070207         1         1
## FBtr0070238         1         1
## FBtr0070251         1        NA
## FBtr0070484         1        NA
## FBtr0070489        NA        NA</code></pre>
<blockquote>
<p>What is each column?</p>
</blockquote>
<div class="sourceCode" id="cb309"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb309-1"><a href="rna-seq-analysis.html#cb309-1"></a><span class="kw">mcols</span>(res.ZT3vsZT15.DN1)<span class="op">$</span>description</span></code></pre></div>
<pre><code>## [1] &quot;mean of normalized counts for all samples&quot;
## [2] &quot;log2 fold change (MLE): ZT ZT3 vs ZT15&quot;   
## [3] &quot;standard error: ZT ZT3 vs ZT15&quot;           
## [4] &quot;Wald statistic: ZT ZT3 vs ZT15&quot;           
## [5] &quot;Wald test p-value: ZT ZT3 vs ZT15&quot;        
## [6] &quot;BH adjusted p-values&quot;</code></pre>
<p>We can see the distribution.</p>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="rna-seq-analysis.html#cb311-1"></a><span class="kw">summary</span>(res.ZT3vsZT15.DN1)</span></code></pre></div>
<pre><code>## 
## out of 25337 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)       : 79, 0.31%
## LFC &lt; 0 (down)     : 55, 0.22%
## outliers [1]       : 0, 0%
## low counts [2]     : 3419, 13%
## (mean count &lt; 2)
## [1] see &#39;cooksCutoff&#39; argument of ?results
## [2] see &#39;independentFiltering&#39; argument of ?results</code></pre>
<p>We can now do an histogram to explore how is the distribution of the padj.</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="rna-seq-analysis.html#cb313-1"></a><span class="co">#</span></span>
<span id="cb313-2"><a href="rna-seq-analysis.html#cb313-2"></a><span class="kw">hist</span>(res.ZT3vsZT15.DN1<span class="op">$</span>padj, <span class="dt">breaks=</span><span class="dv">100</span>, <span class="dt">col=</span><span class="st">&quot;skyblue&quot;</span>, <span class="dt">border=</span><span class="st">&quot;slateblue&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-31"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-31-1.png" alt="Histogram of psjusted value distribution" width="100%" />
<p class="caption">
Figure 6.5: Histogram of psjusted value distribution
</p>
</div>
<p>Other way to see significance is to use MA plots which shows log2 fold changes (on the y-axis) versus the mean of normalized counts (on the x-axis).</p>
<div class="sourceCode" id="cb314"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb314-1"><a href="rna-seq-analysis.html#cb314-1"></a><span class="kw">plotMA</span>(res.ZT3vsZT15.DN1,<span class="dt">alpha=</span><span class="fl">0.05</span>) <span class="co">#In red everything with padj &lt; 0.05</span></span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-32"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-32-1.png" alt="MA plot: log2 fold changes vs normalized counts" width="100%" />
<p class="caption">
Figure 6.6: MA plot: log2 fold changes vs normalized counts
</p>
</div>
<p>We can create a data frame of the results and play around.</p>
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="rna-seq-analysis.html#cb315-1"></a>res.ZT3vsZT15.DN1_df&lt;-<span class="kw">as.data.frame</span>(res.ZT3vsZT15.DN1)</span>
<span id="cb315-2"><a href="rna-seq-analysis.html#cb315-2"></a>res.ZT3vsZT15.DN1_df=res.ZT3vsZT15.DN1_df[<span class="kw">complete.cases</span>(res.ZT3vsZT15.DN1_df),]<span class="co">#remove NAs</span></span>
<span id="cb315-3"><a href="rna-seq-analysis.html#cb315-3"></a>res.ZT3vsZT15.DN1_df<span class="op">$</span>transcript_name=<span class="kw">rownames</span>(res.ZT3vsZT15.DN1_df)</span>
<span id="cb315-4"><a href="rna-seq-analysis.html#cb315-4"></a></span>
<span id="cb315-5"><a href="rna-seq-analysis.html#cb315-5"></a><span class="kw">head</span>(res.ZT3vsZT15.DN1_df)</span></code></pre></div>
<pre><code>##              baseMean log2FoldChange     lfcSE      stat     pvalue padj
## FBtr0070207  2.116836      0.0000000 5.2444845  0.000000 1.00000000    1
## FBtr0070238  2.683573      0.0000000 4.8861753  0.000000 1.00000000    1
## FBtr0070658  2.762362      0.0000000 5.3979919  0.000000 1.00000000    1
## FBtr0071170 16.483738      2.1943705 1.1682234  1.878383 0.06032885    1
## FBtr0071173  4.674766      0.0000000 5.3502754  0.000000 1.00000000    1
## FBtr0071619 49.515374     -0.8457701 0.6368886 -1.327972 0.18418743    1
##             transcript_name
## FBtr0070207     FBtr0070207
## FBtr0070238     FBtr0070238
## FBtr0070658     FBtr0070658
## FBtr0071170     FBtr0071170
## FBtr0071173     FBtr0071173
## FBtr0071619     FBtr0071619</code></pre>
<p>As you can see, the names of the genes are in a nomenclature that is not intuitive. We can then use some functions that will allow us to change the names for some more informative.</p>
<p>Lets take one example: FBtr0070207, is a trsncript name (not a gene name, but one of the possible variants of the gene) <a href="http://flybase.org/reports/FBtr0070207" class="uri">http://flybase.org/reports/FBtr0070207</a>. We want for this analysis the gene name. For that we will use the package <code>org.Dm.eg.db</code> (each model organism has one of this: <a href="https://bioconductor.org/packages/release/BiocViews.html#___AnnotationData" class="uri">https://bioconductor.org/packages/release/BiocViews.html#___AnnotationData</a>)</p>
<div class="sourceCode" id="cb317"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb317-1"><a href="rna-seq-analysis.html#cb317-1"></a><span class="co">#And add the names of the genes</span></span>
<span id="cb317-2"><a href="rna-seq-analysis.html#cb317-2"></a><span class="kw">keytypes</span>(org.Dm.eg.db)</span>
<span id="cb317-3"><a href="rna-seq-analysis.html#cb317-3"></a><span class="co">#We have here the ENSEMBLE transcript names, so we will use that keytype</span></span>
<span id="cb317-4"><a href="rna-seq-analysis.html#cb317-4"></a>genenames &lt;-<span class="st"> </span><span class="kw">mapIds</span>(org.Dm.eg.db,<span class="dt">keys =</span> <span class="kw">rownames</span>(dds.DN1),<span class="dt">column =</span> <span class="st">&quot;SYMBOL&quot;</span>,<span class="dt">keytype=</span><span class="st">&quot;ENSEMBLTRANS&quot;</span>)</span>
<span id="cb317-5"><a href="rna-seq-analysis.html#cb317-5"></a>annotation_DN1 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gene_name =</span> genenames, <span class="dt">transcript_name=</span><span class="kw">rownames</span>(dds.DN1), <span class="dt">row.names =</span> <span class="kw">rownames</span>(dds.DN1), <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</span>
<span id="cb317-6"><a href="rna-seq-analysis.html#cb317-6"></a><span class="kw">head</span>(annotation_DN1) </span></code></pre></div>
<pre><code>##  [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot;  &quot;ENSEMBLTRANS&quot;
##  [6] &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;     &quot;EVIDENCEALL&quot;  &quot;FLYBASE&quot;     
## [11] &quot;FLYBASECG&quot;    &quot;FLYBASEPROT&quot;  &quot;GENENAME&quot;     &quot;GO&quot;           &quot;GOALL&quot;       
## [16] &quot;MAP&quot;          &quot;ONTOLOGY&quot;     &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PMID&quot;        
## [21] &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UNIGENE&quot;      &quot;UNIPROT&quot;     
##             gene_name transcript_name
## FBtr0070202   CG14625     FBtr0070202
## FBtr0070207    fs(1)N     FBtr0070207
## FBtr0070238   CG11409     FBtr0070238
## FBtr0070251   CG11384     FBtr0070251
## FBtr0070484   fs(1)Yb     FBtr0070484
## FBtr0070489   CG12498     FBtr0070489</code></pre>
<p>Now that we have a “mapping” we can use it.</p>
<div class="sourceCode" id="cb319"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb319-1"><a href="rna-seq-analysis.html#cb319-1"></a>res.ZT3vsZT15.DN1_df=<span class="kw">merge</span>(res.ZT3vsZT15.DN1_df,annotation_DN1,<span class="dt">by=</span><span class="st">&quot;transcript_name&quot;</span>)</span>
<span id="cb319-2"><a href="rna-seq-analysis.html#cb319-2"></a><span class="kw">head</span>(res.ZT3vsZT15.DN1_df)</span></code></pre></div>
<pre><code>##   transcript_name  baseMean log2FoldChange     lfcSE       stat    pvalue padj
## 1     FBtr0005088 404.52656     -0.3962867 0.4939109 -0.8023445 0.4223537    1
## 2     FBtr0005673   2.35263      1.7199349 2.8262384  0.6085598 0.5428163    1
## 3     FBtr0005674   2.35263      1.7199349 2.8262384  0.6085598 0.5428163    1
## 4     FBtr0070000  44.02153     -0.7589507 1.1324434 -0.6701886 0.5027375    1
## 5     FBtr0070002   4.30420      3.7116759 2.5555706  1.4523863 0.1463942    1
## 6     FBtr0070006  23.06487     -1.1075368 1.0502903 -1.0545054 0.2916516    1
##   gene_name
## 1  Pp2A-29B
## 2      &lt;NA&gt;
## 3      &lt;NA&gt;
## 4      Nep3
## 5    CG9570
## 6    CG9572</code></pre>
<p>Now lets look for the genes that are changing with than Log2foldChange &gt; 1 (which is a fold change of 2)</p>
<div class="sourceCode" id="cb321"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb321-1"><a href="rna-seq-analysis.html#cb321-1"></a>sig_down=<span class="kw">unique</span>(res.ZT3vsZT15.DN1_df<span class="op">$</span>gene_name[(res.ZT3vsZT15.DN1_df<span class="op">$</span>padj<span class="op">&lt;</span><span class="fl">0.05</span><span class="op">&amp;</span>res.ZT3vsZT15.DN1_df<span class="op">$</span>log2FoldChange<span class="op">&lt;</span>(<span class="op">-</span><span class="dv">1</span>))])</span>
<span id="cb321-2"><a href="rna-seq-analysis.html#cb321-2"></a>sig_down</span>
<span id="cb321-3"><a href="rna-seq-analysis.html#cb321-3"></a></span>
<span id="cb321-4"><a href="rna-seq-analysis.html#cb321-4"></a>sig_up=<span class="kw">unique</span>(res.ZT3vsZT15.DN1_df<span class="op">$</span>gene_name[(res.ZT3vsZT15.DN1_df<span class="op">$</span>padj<span class="op">&lt;</span><span class="fl">0.05</span><span class="op">&amp;</span>res.ZT3vsZT15.DN1_df<span class="op">$</span>log2FoldChange<span class="op">&gt;</span><span class="dv">1</span>)])</span></code></pre></div>
<pre><code>##  [1] NA         &quot;CG14089&quot;  &quot;CG13041&quot;  &quot;CG32137&quot;  &quot;Pdp1&quot;     &quot;tim&quot;     
##  [7] &quot;vri&quot;      &quot;CG18302&quot;  &quot;rdo&quot;      &quot;CG6959&quot;   &quot;GstD2&quot;    &quot;CG31183&quot; 
## [13] &quot;TwdlW&quot;    &quot;CG4000&quot;   &quot;Usp8&quot;     &quot;CG31324&quot;  &quot;RYa-R&quot;    &quot;CG11825&quot; 
## [19] &quot;CG6362&quot;   &quot;CG34247&quot;  &quot;TrissinR&quot; &quot;Nep5&quot;     &quot;CG17777&quot;  &quot;CG13071&quot; 
## [25] &quot;beat-VI&quot;</code></pre>
<p>Lets plot them</p>
<div class="sourceCode" id="cb323"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb323-1"><a href="rna-seq-analysis.html#cb323-1"></a><span class="kw">plotCounts</span>(dds.DN1, <span class="dt">gene=</span><span class="kw">which.min</span>(res.ZT3vsZT15.DN1<span class="op">$</span>padj), <span class="dt">intgroup=</span><span class="st">&quot;ZT&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-361"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-36-1.png" alt="Plot gene expression." width="100%" />
<p class="caption">
Figure 3.3: Plot gene expression.
</p>
</div>
<div class="sourceCode" id="cb324"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb324-1"><a href="rna-seq-analysis.html#cb324-1"></a><span class="kw">plotCounts</span>(dds.DN1, <span class="dt">gene=</span><span class="st">&quot;FBtr0077567&quot;</span>, <span class="dt">intgroup=</span><span class="st">&quot;ZT&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-362"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-36-2.png" alt="Plot gene expression." width="100%" />
<p class="caption">
Figure 3.4: Plot gene expression.
</p>
</div>
<p>We can do this in ggplot2</p>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="rna-seq-analysis.html#cb325-1"></a>d &lt;-<span class="st"> </span><span class="kw">plotCounts</span>(dds.DN1, <span class="dt">gene=</span><span class="st">&quot;FBtr0077567&quot;</span>, <span class="dt">intgroup=</span><span class="st">&quot;ZT&quot;</span>, <span class="dt">returnData=</span><span class="ot">TRUE</span>)</span>
<span id="cb325-2"><a href="rna-seq-analysis.html#cb325-2"></a><span class="kw">ggplot</span>(d, <span class="kw">aes</span>(<span class="dt">x=</span>ZT, <span class="dt">y=</span>count)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w=</span><span class="fl">0.1</span>,<span class="dt">h=</span><span class="dv">0</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-371"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-37-1.png" alt="Plot gene expression using ggplot." width="100%" />
<p class="caption">
Figure 3.5: Plot gene expression using ggplot.
</p>
</div>
<div class="sourceCode" id="cb326"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb326-1"><a href="rna-seq-analysis.html#cb326-1"></a>trans_down=<span class="kw">unique</span>(res.ZT3vsZT15.DN1_df<span class="op">$</span>transcript_name[(res.ZT3vsZT15.DN1_df<span class="op">$</span>padj<span class="op">&lt;</span><span class="fl">0.05</span><span class="op">&amp;</span>res.ZT3vsZT15.DN1_df<span class="op">$</span>log2FoldChange<span class="op">&lt;</span>(<span class="op">-</span><span class="dv">1</span>))])</span>
<span id="cb326-2"><a href="rna-seq-analysis.html#cb326-2"></a></span>
<span id="cb326-3"><a href="rna-seq-analysis.html#cb326-3"></a><span class="co">#We can do this for the genes differentially expressed.</span></span>
<span id="cb326-4"><a href="rna-seq-analysis.html#cb326-4"></a>e=<span class="dv">0</span></span>
<span id="cb326-5"><a href="rna-seq-analysis.html#cb326-5"></a><span class="cf">for</span> (i <span class="cf">in</span> trans_down[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]){</span>
<span id="cb326-6"><a href="rna-seq-analysis.html#cb326-6"></a>  d &lt;-<span class="st"> </span><span class="kw">plotCounts</span>(dds.DN1, <span class="dt">gene=</span>i, <span class="dt">intgroup=</span><span class="st">&quot;ZT&quot;</span>, <span class="dt">returnData=</span>T)</span>
<span id="cb326-7"><a href="rna-seq-analysis.html#cb326-7"></a>  e=e<span class="op">+</span><span class="dv">1</span> <span class="co">#this makes e greater each time</span></span>
<span id="cb326-8"><a href="rna-seq-analysis.html#cb326-8"></a>  n=<span class="kw">paste0</span>(<span class="st">&quot;p&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>,e)</span>
<span id="cb326-9"><a href="rna-seq-analysis.html#cb326-9"></a> <span class="kw">assign</span>(n,<span class="kw">ggplot</span>(d, <span class="kw">aes</span>(<span class="dt">x=</span>ZT, <span class="dt">y=</span>count)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w=</span><span class="fl">0.1</span>,<span class="dt">h=</span><span class="dv">0</span>)) <span class="op">+</span><span class="st"> </span><span class="kw">labs</span> (<span class="dt">title=</span>(res.ZT3vsZT15.DN1_df<span class="op">$</span>gene_name[res.ZT3vsZT15.DN1_df<span class="op">$</span>transcript_name<span class="op">==</span>i]))) </span>
<span id="cb326-10"><a href="rna-seq-analysis.html#cb326-10"></a>  }</span>
<span id="cb326-11"><a href="rna-seq-analysis.html#cb326-11"></a><span class="kw">grid.arrange</span>(p_<span class="dv">4</span>,p_<span class="dv">2</span>, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-372"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-37-2.png" alt="Plot gene expression using ggplot." width="100%" />
<p class="caption">
Figure 3.6: Plot gene expression using ggplot.
</p>
</div>
<div class="sourceCode" id="cb327"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb327-1"><a href="rna-seq-analysis.html#cb327-1"></a><span class="co">#Or for the clock genes like tim, per, clk</span></span>
<span id="cb327-2"><a href="rna-seq-analysis.html#cb327-2"></a><span class="co">#For this we will use the transcript_names for each of them</span></span>
<span id="cb327-3"><a href="rna-seq-analysis.html#cb327-3"></a>tim_trans&lt;-<span class="kw">na.omit</span>(annotation_DN1<span class="op">$</span>transcript_name[annotation_DN1<span class="op">$</span>gene_name<span class="op">==</span><span class="st">&quot;tim&quot;</span>])</span>
<span id="cb327-4"><a href="rna-seq-analysis.html#cb327-4"></a><span class="kw">c</span>(tim_trans)</span>
<span id="cb327-5"><a href="rna-seq-analysis.html#cb327-5"></a>Clk_trans&lt;-<span class="kw">na.omit</span>(annotation_DN1<span class="op">$</span>transcript_name[annotation_DN1<span class="op">$</span>gene_name<span class="op">==</span><span class="st">&quot;Clk&quot;</span>])</span>
<span id="cb327-6"><a href="rna-seq-analysis.html#cb327-6"></a>per_trans&lt;-<span class="kw">na.omit</span>(annotation_DN1<span class="op">$</span>transcript_name[annotation_DN1<span class="op">$</span>gene_name<span class="op">==</span><span class="st">&quot;per&quot;</span>])</span>
<span id="cb327-7"><a href="rna-seq-analysis.html#cb327-7"></a></span>
<span id="cb327-8"><a href="rna-seq-analysis.html#cb327-8"></a><span class="co">#Take into account that this loop will overwrite the previous plots you should change the names if you want to preserve the precious ones</span></span>
<span id="cb327-9"><a href="rna-seq-analysis.html#cb327-9"></a>e=<span class="dv">0</span></span>
<span id="cb327-10"><a href="rna-seq-analysis.html#cb327-10"></a><span class="cf">for</span> (i <span class="cf">in</span> tim_trans){</span>
<span id="cb327-11"><a href="rna-seq-analysis.html#cb327-11"></a>  d &lt;-<span class="st"> </span><span class="kw">plotCounts</span>(dds.DN1, <span class="dt">gene=</span>i, <span class="dt">intgroup=</span><span class="st">&quot;ZT&quot;</span>, <span class="dt">returnData=</span>T)</span>
<span id="cb327-12"><a href="rna-seq-analysis.html#cb327-12"></a>  e=e<span class="op">+</span><span class="dv">1</span> <span class="co">#this makes e greater each time</span></span>
<span id="cb327-13"><a href="rna-seq-analysis.html#cb327-13"></a>  n=<span class="kw">paste0</span>(<span class="st">&quot;p&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>,e)</span>
<span id="cb327-14"><a href="rna-seq-analysis.html#cb327-14"></a>  t=<span class="kw">paste0</span>(annotation_DN1<span class="op">$</span>gene_name[annotation_DN1<span class="op">$</span>transcript_name<span class="op">==</span>i],<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>,i) <span class="co">#title for the plot</span></span>
<span id="cb327-15"><a href="rna-seq-analysis.html#cb327-15"></a> <span class="kw">assign</span>(n,<span class="kw">ggplot</span>(d, <span class="kw">aes</span>(<span class="dt">x=</span>ZT, <span class="dt">y=</span>count)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w=</span><span class="fl">0.1</span>,<span class="dt">h=</span><span class="dv">0</span>)) <span class="op">+</span><span class="st"> </span><span class="kw">labs</span> (<span class="dt">title=</span>t) )</span>
<span id="cb327-16"><a href="rna-seq-analysis.html#cb327-16"></a>  }</span>
<span id="cb327-17"><a href="rna-seq-analysis.html#cb327-17"></a></span>
<span id="cb327-18"><a href="rna-seq-analysis.html#cb327-18"></a><span class="kw">grid.arrange</span>(p_<span class="dv">1</span>,p_<span class="dv">2</span>,p_<span class="dv">4</span>,p_<span class="dv">5</span>, <span class="dt">ncol=</span><span class="dv">2</span>,<span class="dt">top=</span><span class="st">&quot;Tim transcripts&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-373"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-37-3.png" alt="Plot gene expression using ggplot." width="100%" />
<p class="caption">
Figure 3.7: Plot gene expression using ggplot.
</p>
</div>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb328-1"><a href="rna-seq-analysis.html#cb328-1"></a><span class="co">#Notice that will overwrite everything if you want to preserve the plots you need to actually change the object names</span></span>
<span id="cb328-2"><a href="rna-seq-analysis.html#cb328-2"></a>e=<span class="dv">0</span></span>
<span id="cb328-3"><a href="rna-seq-analysis.html#cb328-3"></a><span class="cf">for</span> (i <span class="cf">in</span> Clk_trans){</span>
<span id="cb328-4"><a href="rna-seq-analysis.html#cb328-4"></a>  d &lt;-<span class="st"> </span><span class="kw">plotCounts</span>(dds.DN1, <span class="dt">gene=</span>i, <span class="dt">intgroup=</span><span class="st">&quot;ZT&quot;</span>, <span class="dt">returnData=</span>T)</span>
<span id="cb328-5"><a href="rna-seq-analysis.html#cb328-5"></a>  e=e<span class="op">+</span><span class="dv">1</span> <span class="co">#this makes e greater each time</span></span>
<span id="cb328-6"><a href="rna-seq-analysis.html#cb328-6"></a>  n=<span class="kw">paste0</span>(<span class="st">&quot;p&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>,e)</span>
<span id="cb328-7"><a href="rna-seq-analysis.html#cb328-7"></a> t=<span class="kw">paste0</span>(annotation_DN1<span class="op">$</span>gene_name[annotation_DN1<span class="op">$</span>transcript_name<span class="op">==</span>i],<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>,i) <span class="co">#title for the plot</span></span>
<span id="cb328-8"><a href="rna-seq-analysis.html#cb328-8"></a> <span class="kw">assign</span>(n,<span class="kw">ggplot</span>(d, <span class="kw">aes</span>(<span class="dt">x=</span>ZT, <span class="dt">y=</span>count)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w=</span><span class="fl">0.1</span>,<span class="dt">h=</span><span class="dv">0</span>)) <span class="op">+</span><span class="st"> </span><span class="kw">labs</span> (<span class="dt">title=</span>t) )</span>
<span id="cb328-9"><a href="rna-seq-analysis.html#cb328-9"></a>   }</span>
<span id="cb328-10"><a href="rna-seq-analysis.html#cb328-10"></a></span>
<span id="cb328-11"><a href="rna-seq-analysis.html#cb328-11"></a><span class="kw">grid.arrange</span>(p_<span class="dv">1</span>,p_<span class="dv">2</span>,p_<span class="dv">3</span>, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-374"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-37-4.png" alt="Plot gene expression using ggplot." width="100%" />
<p class="caption">
Figure 6.7: Plot gene expression using ggplot.
</p>
</div>
<div class="sourceCode" id="cb329"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb329-1"><a href="rna-seq-analysis.html#cb329-1"></a><span class="co">#Notice that will overwrite everything if you want to preserve the plots you need to actually change the object names</span></span>
<span id="cb329-2"><a href="rna-seq-analysis.html#cb329-2"></a>e=<span class="dv">0</span></span>
<span id="cb329-3"><a href="rna-seq-analysis.html#cb329-3"></a><span class="cf">for</span> (i <span class="cf">in</span> per_trans){</span>
<span id="cb329-4"><a href="rna-seq-analysis.html#cb329-4"></a>  d &lt;-<span class="st"> </span><span class="kw">plotCounts</span>(dds.DN1, <span class="dt">gene=</span>i, <span class="dt">intgroup=</span><span class="st">&quot;ZT&quot;</span>, <span class="dt">returnData=</span>T)</span>
<span id="cb329-5"><a href="rna-seq-analysis.html#cb329-5"></a>  e=e<span class="op">+</span><span class="dv">1</span> <span class="co">#this makes e greater each time</span></span>
<span id="cb329-6"><a href="rna-seq-analysis.html#cb329-6"></a>  n=<span class="kw">paste0</span>(<span class="st">&quot;p&quot;</span>,<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>,e)</span>
<span id="cb329-7"><a href="rna-seq-analysis.html#cb329-7"></a> t=<span class="kw">paste0</span>(annotation_DN1<span class="op">$</span>gene_name[annotation_DN1<span class="op">$</span>transcript_name<span class="op">==</span>i],<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>,i) <span class="co">#title for the plot</span></span>
<span id="cb329-8"><a href="rna-seq-analysis.html#cb329-8"></a> <span class="kw">assign</span>(n,(<span class="kw">ggplot</span>(d, <span class="kw">aes</span>(<span class="dt">x=</span>ZT, <span class="dt">y=</span>count, <span class="dt">group=</span><span class="dv">1</span>) ) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w=</span><span class="fl">0.1</span>,<span class="dt">h=</span><span class="dv">0</span>))  <span class="op">+</span><span class="st"> </span><span class="kw">labs</span> (<span class="dt">title=</span>t) ))</span>
<span id="cb329-9"><a href="rna-seq-analysis.html#cb329-9"></a>}</span>
<span id="cb329-10"><a href="rna-seq-analysis.html#cb329-10"></a></span>
<span id="cb329-11"><a href="rna-seq-analysis.html#cb329-11"></a><span class="kw">grid.arrange</span>(p_<span class="dv">1</span>,p_<span class="dv">2</span>, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-375"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-37-5.png" alt="Plot gene expression using ggplot." width="100%" />
<p class="caption">
Figure 6.8: Plot gene expression using ggplot.
</p>
</div>
<pre><code>## [1] &quot;FBtr0077567&quot; &quot;FBtr0333252&quot; &quot;FBtr0333253&quot; &quot;FBtr0333254&quot; &quot;FBtr0333255&quot;
## [6] &quot;FBtr0333256&quot; &quot;FBtr0333258&quot; &quot;FBtr0333259&quot;</code></pre>
<p>Let’s discuss what we see in the graphics.</p>
<p>We have some of the figures that seems to have the exactly same values for each replicate..this is at least suspicious. This might be just the same quantification so it might not be correct to say that they are different isoforms and maybe it is better to collapse them or change the strategy to analyze different isoforms.</p>
</div>
<div id="plot-counts-using-ggplot" class="section level3" number="6.3.6">
<h3><span class="header-section-number">6.3.6</span> Plot counts using ggplot</h3>
<p>Doing a few manipulations on the table we can have even better plots.</p>
<div class="sourceCode" id="cb331"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb331-1"><a href="rna-seq-analysis.html#cb331-1"></a><span class="co">#big plot</span></span>
<span id="cb331-2"><a href="rna-seq-analysis.html#cb331-2"></a><span class="co">#I take the normalized values and do the plots</span></span>
<span id="cb331-3"><a href="rna-seq-analysis.html#cb331-3"></a>toplot&lt;-<span class="kw">as.data.frame</span>(<span class="kw">as.data.frame</span>(<span class="kw">counts</span>(dds.DN1, <span class="dt">normalized=</span>T)))</span>
<span id="cb331-4"><a href="rna-seq-analysis.html#cb331-4"></a>toplot<span class="op">$</span>transcript_name&lt;-<span class="kw">rownames</span>(toplot)</span>
<span id="cb331-5"><a href="rna-seq-analysis.html#cb331-5"></a></span>
<span id="cb331-6"><a href="rna-seq-analysis.html#cb331-6"></a>toplot&lt;-toplot <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">gather</span>(zt_rep, value, <span class="op">-</span>transcript_name) <span class="co">#This is a more complex line of code, lets explore the results.</span></span>
<span id="cb331-7"><a href="rna-seq-analysis.html#cb331-7"></a><span class="kw">head</span>(toplot)</span></code></pre></div>
<pre><code>##   transcript_name    zt_rep value
## 1     FBtr0070202 DN1_ZT3_1     0
## 2     FBtr0070207 DN1_ZT3_1     0
## 3     FBtr0070238 DN1_ZT3_1     0
## 4     FBtr0070251 DN1_ZT3_1     0
## 5     FBtr0070484 DN1_ZT3_1     0
## 6     FBtr0070489 DN1_ZT3_1     0</code></pre>
<div class="sourceCode" id="cb333"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb333-1"><a href="rna-seq-analysis.html#cb333-1"></a>toplot=<span class="kw">merge</span>(toplot,annotation_DN1,<span class="dt">by=</span><span class="st">&quot;transcript_name&quot;</span>)</span>
<span id="cb333-2"><a href="rna-seq-analysis.html#cb333-2"></a><span class="kw">head</span>(toplot)</span></code></pre></div>
<pre><code>##   transcript_name     zt_rep    value gene_name
## 1     FBtr0005088 DN1_ZT11_2 333.0941  Pp2A-29B
## 2     FBtr0005088 DN1_ZT19_2 332.2765  Pp2A-29B
## 3     FBtr0005088  DN1_ZT7_2 520.4211  Pp2A-29B
## 4     FBtr0005088  DN1_ZT3_2 235.0216  Pp2A-29B
## 5     FBtr0005088 DN1_ZT23_2 392.4796  Pp2A-29B
## 6     FBtr0005088 DN1_ZT19_1 406.2523  Pp2A-29B</code></pre>
<blockquote>
<p>What changed here? How this allow for better plotting?</p>
</blockquote>
<div class="sourceCode" id="cb335"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb335-1"><a href="rna-seq-analysis.html#cb335-1"></a><span class="co">#Lets create now more meta data </span></span>
<span id="cb335-2"><a href="rna-seq-analysis.html#cb335-2"></a>toplot<span class="op">$</span>zt&lt;-<span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">as.character</span>(toplot<span class="op">$</span>zt_rep),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">2</span>)</span>
<span id="cb335-3"><a href="rna-seq-analysis.html#cb335-3"></a>toplot<span class="op">$</span>ztime&lt;-<span class="kw">gsub</span>(toplot<span class="op">$</span>zt,<span class="dt">pattern =</span> <span class="st">&quot;ZT&quot;</span>,<span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb335-4"><a href="rna-seq-analysis.html#cb335-4"></a>toplot<span class="op">$</span>rep&lt;-<span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">as.character</span>(toplot<span class="op">$</span>zt_rep),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">3</span>)</span>
<span id="cb335-5"><a href="rna-seq-analysis.html#cb335-5"></a>toplot<span class="op">$</span>ztime_<span class="dv">2</span>&lt;-<span class="kw">as.numeric</span>(toplot<span class="op">$</span>ztime)</span>
<span id="cb335-6"><a href="rna-seq-analysis.html#cb335-6"></a>toplot<span class="op">$</span>ztime_<span class="dv">2</span>[toplot<span class="op">$</span>rep<span class="op">==</span><span class="st">&quot;2&quot;</span>]&lt;-toplot<span class="op">$</span>ztime_<span class="dv">2</span>[toplot<span class="op">$</span>rep<span class="op">==</span><span class="st">&quot;2&quot;</span>]<span class="op">+</span><span class="dv">24</span></span>
<span id="cb335-7"><a href="rna-seq-analysis.html#cb335-7"></a></span>
<span id="cb335-8"><a href="rna-seq-analysis.html#cb335-8"></a><span class="co">#And plot</span></span>
<span id="cb335-9"><a href="rna-seq-analysis.html#cb335-9"></a>genes.cyc =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;tim&quot;</span>,<span class="st">&quot;Clk&quot;</span>,<span class="st">&quot;per&quot;</span>,<span class="st">&quot;cwo&quot;</span>) <span class="co">#you can put here as many genes as you want</span></span>
<span id="cb335-10"><a href="rna-seq-analysis.html#cb335-10"></a></span>
<span id="cb335-11"><a href="rna-seq-analysis.html#cb335-11"></a></span>
<span id="cb335-12"><a href="rna-seq-analysis.html#cb335-12"></a><span class="kw">ggplot</span>(toplot[toplot<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span>genes.cyc ,],<span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw">as.numeric</span>(ztime),<span class="dt">y =</span> value,<span class="dt">color=</span>gene_name,<span class="dt">shape=</span>rep)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>() </span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-401"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-40-1.png" alt="Count plots with ggplot" width="100%" />
<p class="caption">
Figure 6.9: Count plots with ggplot
</p>
</div>
<div class="sourceCode" id="cb336"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb336-1"><a href="rna-seq-analysis.html#cb336-1"></a><span class="kw">ggplot</span>(toplot[toplot<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span>genes.cyc ,],<span class="kw">aes</span>(<span class="dt">x =</span>ztime_<span class="dv">2</span>,<span class="dt">y =</span> value,<span class="dt">color=</span>gene_name,<span class="dt">shape=</span>rep)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>()</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-402"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-40-2.png" alt="Count plots with ggplot" width="100%" />
<p class="caption">
Figure 6.10: Count plots with ggplot
</p>
</div>
<div class="sourceCode" id="cb337"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb337-1"><a href="rna-seq-analysis.html#cb337-1"></a><span class="kw">ggplot</span>(toplot[toplot<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span>genes.cyc ,],<span class="kw">aes</span>(<span class="dt">x =</span>ztime_<span class="dv">2</span>,<span class="dt">y =</span> value,<span class="dt">color=</span>gene_name,<span class="dt">shape=</span>rep)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span><span class="kw">facet_wrap</span>(<span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>,<span class="dt">facets =</span> <span class="op">~</span>gene_name )</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-403"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-40-3.png" alt="Count plots with ggplot" width="100%" />
<p class="caption">
Figure 6.11: Count plots with ggplot
</p>
</div>
</div>
</div>
<div id="cycling-genes-analysis" class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Cycling genes analysis</h2>
<p>As stated before, if we want to see cycling we have to look at repetitive patterns in the data cross the day. To determine cycling genes, different approaches can be done. But basically the idea is to assess for each gene:</p>
<p>1.Identify cycling elements in the data. (ie which genes are changing between the timepoints)
2.If so, in which time point that particular gene has its peak.
3.Which is the amplitude and the period of the cycle: ie. how pronounced is the peak and how often it is cycling.</p>
<p>One popular algorithm used is <a href="https://openwetware.org/wiki/HughesLab:JTK_Cycle">JTK</a>
And its description says: “Its purpose is to identify rhythmic components in large, genome-scale data sets and estimate their period length, phase, and amplitude”</p>
<p>We will use the library: <a href="https://academic.oup.com/bioinformatics/article/32/21/3351/2415176">MetaCycle</a>
It is the successor of JTK according to JTK authors and it incorporates different approaches:</p>
<p>“Using the same input file, MetaCycle::meta2d implements ARSER (ARS), JTK_CYCLE (JTK) and Lomb-Scargle (LS), from dramatically different disciplines, computer science, statistics and physics, respectively.
Three independent algorithms (ARS, JTK and LS) were then selected from best of breed methods (Deckard et al., 2013; Wu et al., 2014). If set analysisStrategy as”auto“(default), it will automatically select proper method from cycMethod for each input dataset”</p>
<p>There is a nice tutorial only from the <a href="https://cran.r-project.org/web/packages/MetaCycle/vignettes/implementation.html">CRAN repository</a></p>
<p>And of course you can always use the internal R help <code>?meta2d</code></p>
<div class="sourceCode" id="cb338"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb338-1"><a href="rna-seq-analysis.html#cb338-1"></a><span class="co">#We already have the tables prepared</span></span>
<span id="cb338-2"><a href="rna-seq-analysis.html#cb338-2"></a><span class="kw">head</span>(DN1_gene_expression)<span class="co">#</span></span></code></pre></div>
<pre><code>##        Symbol DN1_ZT3_1 DN1_ZT7_1 DN1_ZT11_1 DN1_ZT15_1 DN1_ZT19_1 DN1_ZT23_1
## 1 FBtr0070202         0         0          0          0          0          0
## 2 FBtr0070207         0         0          0          0         14          0
## 3 FBtr0070238         0         0          0          0         21          0
## 4 FBtr0070251         0         0          0          0          0          0
## 5 FBtr0070484         0         0          0          0          0          0
## 6 FBtr0070489         0         0          0          0          0          0
##   DN1_ZT3_2 DN1_ZT7_2 DN1_ZT11_2 DN1_ZT15_2 DN1_ZT19_2 DN1_ZT23_2
## 1         0         0          0          0          0          0
## 2         0         0          0          0         16          7
## 3         0         0          0          0         15          9
## 4         0         0          1          0          2          0
## 5         0         0          0          0          0          1
## 6         0         0          0          0          0          0</code></pre>
<p>This is the raw data, but we should use the normalized one</p>
<div class="sourceCode" id="cb340"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb340-1"><a href="rna-seq-analysis.html#cb340-1"></a><span class="kw">head</span>(<span class="kw">counts</span>(dds.DN1, <span class="dt">normalized=</span>T))</span></code></pre></div>
<pre><code>##             DN1_ZT3_1 DN1_ZT7_1 DN1_ZT11_1 DN1_ZT15_1 DN1_ZT19_1 DN1_ZT23_1
## FBtr0070202         0         0          0          0   0.000000          0
## FBtr0070207         0         0          0          0   7.039025          0
## FBtr0070238         0         0          0          0  10.558537          0
## FBtr0070251         0         0          0          0   0.000000          0
## FBtr0070484         0         0          0          0   0.000000          0
## FBtr0070489         0         0          0          0   0.000000          0
##             DN1_ZT3_2 DN1_ZT7_2 DN1_ZT11_2 DN1_ZT15_2 DN1_ZT19_2 DN1_ZT23_2
## FBtr0070202         0         0   0.000000          0  0.0000000   0.000000
## FBtr0070207         0         0   0.000000          0  5.6437621  12.719245
## FBtr0070238         0         0   0.000000          0  5.2910269  16.353315
## FBtr0070251         0         0   2.947736          0  0.7054703   0.000000
## FBtr0070484         0         0   0.000000          0  0.0000000   1.817035
## FBtr0070489         0         0   0.000000          0  0.0000000   0.000000</code></pre>
<p>If you read the specifications of meta2d, you need the first column to be the gene name. Lets do it. We can do it wiht base R or with specific packages.</p>
<div class="sourceCode" id="cb342"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb342-1"><a href="rna-seq-analysis.html#cb342-1"></a>DN1_gene_expression_norm&lt;-<span class="kw">as.data.frame</span>(<span class="kw">counts</span>(dds.DN1, <span class="dt">normalized=</span>T))</span>
<span id="cb342-2"><a href="rna-seq-analysis.html#cb342-2"></a></span>
<span id="cb342-3"><a href="rna-seq-analysis.html#cb342-3"></a>DN1_gene_expression_norm &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dt">Genes =</span> <span class="kw">rownames</span>(DN1_gene_expression_norm), DN1_gene_expression_norm)</span>
<span id="cb342-4"><a href="rna-seq-analysis.html#cb342-4"></a></span>
<span id="cb342-5"><a href="rna-seq-analysis.html#cb342-5"></a><span class="co">#We need now to write them in an outside file and run the command. </span></span>
<span id="cb342-6"><a href="rna-seq-analysis.html#cb342-6"></a><span class="co">#Note that you might be able to use the original data, this is just to keep on learning how to manage the data</span></span>
<span id="cb342-7"><a href="rna-seq-analysis.html#cb342-7"></a><span class="kw">write.csv</span>(DN1_gene_expression_norm, <span class="dt">file=</span><span class="st">&quot;cycDN1.csv&quot;</span>, <span class="dt">row.names=</span><span class="ot">FALSE</span>)</span>
<span id="cb342-8"><a href="rna-seq-analysis.html#cb342-8"></a></span>
<span id="cb342-9"><a href="rna-seq-analysis.html#cb342-9"></a>DN1_cyc &lt;-<span class="st"> </span><span class="kw">meta2d</span>(<span class="dt">infile=</span><span class="st">&quot;cycDN1.csv&quot;</span>,<span class="dt">filestyle=</span><span class="st">&quot;csv&quot;</span>, <span class="dt">timepoints=</span>(<span class="kw">rep</span>(<span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">7</span>,<span class="dv">11</span>,<span class="dv">15</span>,<span class="dv">19</span>,<span class="dv">23</span>),<span class="dv">2</span>)),<span class="dt">outputFile=</span><span class="ot">FALSE</span>,<span class="dt">outRawData=</span>F)</span></code></pre></div>
<pre><code>## The JTK is in process from  11:29:24 07-25-2020 
## Warning: the input &#39;maxper&#39; is not suitable for JTK, it was reset as  24 
## The analysis by JTK is finished at  11:29:56 07-25-2020 
## The LS is in process from  11:29:56 07-25-2020 
## The analysis by LS is finished at  11:36:54 07-25-2020 
## DONE! The analysis about &#39; cycDN1.csv &#39;  has been finished.
##                 user.self     sys.self      elapsed   user.child    sys.child 
## &quot;Time used:&quot;    &quot;437.646&quot;     &quot;20.844&quot;    &quot;480.543&quot;          &quot;0&quot;          &quot;0&quot;</code></pre>
<p>As we can read in the help page <code>?meta2d</code> the output of this function is:</p>
<p>meta2d will write analysis results in different files under outdir if set outputFile = TRUE. Files named with “ARSresult”, “JTKresult” and “LSreult” store analysis results from ARS, JTK and LS respectively. The file named with “meta2d” is the integration file, and it stores integrated values in columns with a common name tag-“meta2d”. The integration file also contains p-value, FDR value, period, phase(adjusted phase if adjustedPhase = “predictedPer”) and amplitude values calculated by each method. If outputFile = FALSE is selected, meta2d will return a list containing the following components:</p>
<p>ARS analysis results from ARS method
JTK analysis results from JTK method
LS analysis results from LS method
meta the integrated analysis results as mentioned above</p>
<p>Let’s explore the results</p>
<div class="sourceCode" id="cb344"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb344-1"><a href="rna-seq-analysis.html#cb344-1"></a><span class="kw">head</span>(DN1_cyc<span class="op">$</span>JTK)</span></code></pre></div>
<pre><code>##         CycID BH.Q ADJ.P PER LAG AMP
## 1 FBtr0070202    1     1   0   0   0
## 2 FBtr0070207    1     1   0   0   0
## 3 FBtr0070238    1     1   0   0   0
## 4 FBtr0070251    1     1   0   0   0
## 5 FBtr0070484    1     1   0   0   0
## 6 FBtr0070489    1     1   0   0   0</code></pre>
<div class="sourceCode" id="cb346"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb346-1"><a href="rna-seq-analysis.html#cb346-1"></a><span class="kw">head</span>(DN1_cyc<span class="op">$</span>LS)</span></code></pre></div>
<pre><code>##         CycID PhaseShift PhaseShiftHeight PeakIndex   PeakSPD Period         p
## 1 FBtr0070202  22.999926         0.000000        NA        NA     NA 1.0000000
## 2 FBtr0070207  22.999922         6.795629         1 2.5714094     28 0.4706206
## 3 FBtr0070238  22.999926         8.713640         1 2.4520804     28 0.5134413
## 4 FBtr0070251  10.999996         1.473868        48 0.9032149     20 0.9843473
## 5 FBtr0070484   7.721415         0.000000         1 1.2731542     28 0.9277372
## 6 FBtr0070489  22.999926         0.000000        NA        NA     NA 1.0000000
##    N Nindependent Nyquist BH.Q
## 1 12            8      NA    1
## 2 12            8     0.3    1
## 3 12            8     0.3    1
## 4 12            8     0.3    1
## 5 12            8     0.3    1
## 6 12            8      NA    1</code></pre>
<div class="sourceCode" id="cb348"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb348-1"><a href="rna-seq-analysis.html#cb348-1"></a><span class="kw">head</span>(DN1_cyc<span class="op">$</span>meta)</span></code></pre></div>
<pre><code>##         CycID JTK_pvalue JTK_BH.Q JTK_period JTK_adjphase JTK_amplitude
## 1 FBtr0070202          1        1          0          NaN             0
## 2 FBtr0070207          1        1          0          NaN             0
## 3 FBtr0070238          1        1          0          NaN             0
## 4 FBtr0070251          1        1          0          NaN             0
## 5 FBtr0070484          1        1          0          NaN             0
## 6 FBtr0070489          1        1          0          NaN             0
##   LS_pvalue LS_BH.Q LS_period LS_adjphase LS_amplitude meta2d_pvalue
## 1 1.0000000       1        NA          NA     0.000000     1.0000000
## 2 0.4706206       1        28   22.999922     6.795629     0.8253288
## 3 0.5134413       1        28   22.999926     8.713640     0.8557113
## 4 0.9843473       1        20   10.999996     1.473868     0.9998769
## 5 0.9277372       1        28    7.721415     0.000000     0.9973238
## 6 1.0000000       1        NA          NA     0.000000     1.0000000
##   meta2d_BH.Q meta2d_period meta2d_phase meta2d_Base meta2d_AMP meta2d_rAMP
## 1           1            NA           NA   0.0000000         NA          NA
## 2           1            28    22.999922   2.4093672 2.81516339  1.16842439
## 3           1            28    22.999926   3.0549360 3.57379221  1.16984191
## 4           1            20    10.999996   0.3720886 0.50175568  0.50175568
## 5           1            28     7.721415   0.1560293 0.07341249  0.07341249
## 6           1            NA           NA   0.0000000         NA          NA</code></pre>
<p>Clearly the “meta” is what we want. We can store it in other object to explore it in R or just write a table</p>
<div class="sourceCode" id="cb350"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb350-1"><a href="rna-seq-analysis.html#cb350-1"></a>DN1_cyc_meta&lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(DN1_cyc<span class="op">$</span>meta)</span>
<span id="cb350-2"><a href="rna-seq-analysis.html#cb350-2"></a></span>
<span id="cb350-3"><a href="rna-seq-analysis.html#cb350-3"></a><span class="co">#We will add the gene-names and then export it</span></span>
<span id="cb350-4"><a href="rna-seq-analysis.html#cb350-4"></a>DN1_cyc_meta<span class="op">$</span>transcript_name=DN1_cyc_meta<span class="op">$</span>CycID</span>
<span id="cb350-5"><a href="rna-seq-analysis.html#cb350-5"></a>DN1_cyc_meta=<span class="kw">merge</span>(DN1_cyc_meta,annotation_DN1,<span class="dt">by=</span><span class="st">&quot;transcript_name&quot;</span>)</span>
<span id="cb350-6"><a href="rna-seq-analysis.html#cb350-6"></a></span>
<span id="cb350-7"><a href="rna-seq-analysis.html#cb350-7"></a><span class="co">#the tim, clk and per:</span></span>
<span id="cb350-8"><a href="rna-seq-analysis.html#cb350-8"></a>DN1_cyc_meta[<span class="kw">which</span>(DN1_cyc_meta<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;tim&quot;</span>,<span class="st">&quot;Clk&quot;</span>,<span class="st">&quot;per&quot;</span>)),]</span></code></pre></div>
<pre><code>##       transcript_name       CycID   JTK_pvalue  JTK_BH.Q JTK_period
## 368       FBtr0070477 FBtr0070477 0.0080482069 0.5867038         24
## 5185      FBtr0076785 FBtr0076785 0.0285920047 0.8536843         20
## 5743      FBtr0077567 FBtr0077567 0.0007025279 0.3232319         20
## 15490     FBtr0100132 FBtr0100132 0.0285920047 0.8536843         20
## 15491     FBtr0100134 FBtr0100134 0.0285920047 0.8536843         20
## 26382     FBtr0332311 FBtr0332311 0.0285920047 0.8536843         24
## 27307     FBtr0333252 FBtr0333252 0.0007025279 0.3232319         20
## 27308     FBtr0333253 FBtr0333253 0.0007025279 0.3232319         20
## 27309     FBtr0333254 FBtr0333254 0.0285920047 0.8536843         24
## 27310     FBtr0333255 FBtr0333255 0.0285920047 0.8536843         24
## 27311     FBtr0333256 FBtr0333256 0.0007025279 0.3232319         20
## 27312     FBtr0333258 FBtr0333258 0.0007025279 0.3232319         20
## 27313     FBtr0333259 FBtr0333259 0.0007025279 0.3232319         20
##       JTK_adjphase JTK_amplitude LS_pvalue LS_BH.Q LS_period LS_adjphase
## 368             13      305.3228 0.2045246       1  23.25088    12.95559
## 5185             3      128.3993 0.4297392       1  27.76371    22.99993
## 5743            13     1648.7247 0.0751332       1  23.08772    14.72816
## 15490            3      128.3993 0.4297392       1  27.76371    22.99993
## 15491            3      128.3993 0.4297392       1  27.76371    22.99993
## 26382           13      318.0454 0.3542849       1  23.25088    14.29129
## 27307           13     1648.7247 0.0751332       1  23.08772    14.72816
## 27308           13     1648.7247 0.0751332       1  23.08772    14.72816
## 27309           15      139.8126 0.1773643       1  22.15488    15.32424
## 27310           15      139.8126 0.1773643       1  22.15488    15.32424
## 27311           13     1648.7247 0.0751332       1  23.08772    14.72816
## 27312           13     1648.7247 0.0751332       1  23.08772    14.72816
## 27313           13     1648.7247 0.0751332       1  23.08772    14.72816
##       LS_amplitude meta2d_pvalue meta2d_BH.Q meta2d_period meta2d_phase
## 368       724.5210  0.0121962439   1.0000000      23.62544     12.98070
## 5185      488.6181  0.0663405962   1.0000000      23.88186     23.62413
## 5743     4366.7125  0.0005726614   0.4069818      21.54386     13.87341
## 15490     488.6181  0.0663405962   1.0000000      23.88186     23.62413
## 15491     488.6181  0.0663405962   1.0000000      23.88186     23.62413
## 26382     790.6228  0.0566482210   1.0000000      23.62544     13.65932
## 27307    4366.7125  0.0005726614   0.4069818      21.54386     13.87341
## 27308    4366.7125  0.0005726614   0.4069818      21.54386     13.87341
## 27309     320.8403  0.0318683270   1.0000000      23.07744     15.19288
## 27310     320.8403  0.0318683270   1.0000000      23.07744     15.19288
## 27311    4366.7125  0.0005726614   0.4069818      21.54386     13.87341
## 27312    4366.7125  0.0005726614   0.4069818      21.54386     13.87341
## 27313    4366.7125  0.0005726614   0.4069818      21.54386     13.87341
##       meta2d_Base meta2d_AMP meta2d_rAMP gene_name
## 368      404.2411   310.1832   0.7673221       per
## 5185     428.9183   455.0223   1.0608602       Clk
## 5743    1976.0053  1791.7205   0.9067387       tim
## 15490    428.9183   455.0223   1.0608602       Clk
## 15491    428.9183   455.0223   1.0608602       Clk
## 26382    420.1319   314.4002   0.7483370       per
## 27307   1976.0053  1791.7205   0.9067387       tim
## 27308   1976.0053  1791.7205   0.9067387       tim
## 27309    150.3703   156.7295   1.0422898       tim
## 27310    150.3703   156.7295   1.0422898       tim
## 27311   1976.0053  1791.7205   0.9067387       tim
## 27312   1976.0053  1791.7205   0.9067387       tim
## 27313   1976.0053  1791.7205   0.9067387       tim</code></pre>
<p>Save the data</p>
<div class="sourceCode" id="cb352"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb352-1"><a href="rna-seq-analysis.html#cb352-1"></a><span class="kw">write.table</span>(<span class="dt">x =</span> DN1_cyc_meta, <span class="dt">file=</span><span class="st">&quot;cycDN1_reults.txt&quot;</span>,<span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>,<span class="dt">row.names =</span> F,<span class="dt">col.names =</span> T)</span></code></pre></div>
<p>Plot the data: this it really usefull to make plots. We will use the object created before.</p>
<div class="sourceCode" id="cb353"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb353-1"><a href="rna-seq-analysis.html#cb353-1"></a><span class="co">#And plot</span></span>
<span id="cb353-2"><a href="rna-seq-analysis.html#cb353-2"></a>genes.cyc =<span class="st"> </span>DN1_cyc_meta<span class="op">$</span>gene_name[<span class="kw">c</span>(DN1_cyc_meta<span class="op">$</span>JTK_pvalue<span class="op">&lt;</span><span class="fl">0.05</span> <span class="op">&amp;</span><span class="st"> </span>DN1_cyc_meta<span class="op">$</span>JTK_amplitude<span class="op">&gt;</span><span class="fl">128.3</span>)]<span class="co">#you can put here as many genes as you want, I am seleceting genes with amplitud bigger than clocl (Clk)</span></span>
<span id="cb353-3"><a href="rna-seq-analysis.html#cb353-3"></a></span>
<span id="cb353-4"><a href="rna-seq-analysis.html#cb353-4"></a><span class="co">#I pick the first 8 genes, but you can look at all of them.</span></span>
<span id="cb353-5"><a href="rna-seq-analysis.html#cb353-5"></a></span>
<span id="cb353-6"><a href="rna-seq-analysis.html#cb353-6"></a><span class="kw">ggplot</span>(toplot[toplot<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span>genes.cyc[<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>] ,],<span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw">as.numeric</span>(ztime),<span class="dt">y =</span> value,<span class="dt">color=</span>gene_name,<span class="dt">shape=</span>rep)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>() </span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-491"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-49-1.png" alt="Cycling genes." width="100%" />
<p class="caption">
Figure 6.12: Cycling genes.
</p>
</div>
<div class="sourceCode" id="cb354"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb354-1"><a href="rna-seq-analysis.html#cb354-1"></a><span class="kw">ggplot</span>(toplot[toplot<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span>genes.cyc[<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>] ,],<span class="kw">aes</span>(<span class="dt">x =</span>ztime_<span class="dv">2</span>,<span class="dt">y =</span> value,<span class="dt">color=</span>gene_name,<span class="dt">shape=</span>rep)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>()</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-492"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-49-2.png" alt="Cycling genes." width="100%" />
<p class="caption">
Figure 6.13: Cycling genes.
</p>
</div>
<div class="sourceCode" id="cb355"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb355-1"><a href="rna-seq-analysis.html#cb355-1"></a><span class="kw">ggplot</span>(toplot[toplot<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span>genes.cyc[<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>] ,],<span class="kw">aes</span>(<span class="dt">x =</span>ztime_<span class="dv">2</span>,<span class="dt">y =</span> value,<span class="dt">color=</span>gene_name,<span class="dt">shape=</span>rep)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span><span class="kw">facet_wrap</span>(<span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>,<span class="dt">facets =</span> <span class="op">~</span>gene_name )</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-493"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-49-3.png" alt="Cycling genes." width="100%" />
<p class="caption">
Figure 6.14: Cycling genes.
</p>
</div>
</div>
<div id="activity-4" class="section level2" number="6.5">
<h2><span class="header-section-number">6.5</span> Activity</h2>
<blockquote>
<p>Finish the analyziz for the other celltypes.
Find a way to compare the results (hint: loog at dendograms)</p>
</blockquote>
</div>
<div id="extra-heatmap-and-pca-in-the-other-celltypes." class="section level2" number="6.6">
<h2><span class="header-section-number">6.6</span> Extra: Heatmap and PCA in the other celltypes.</h2>
<p><strong>LNv</strong></p>
<div class="sourceCode" id="cb356"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb356-1"><a href="rna-seq-analysis.html#cb356-1"></a><span class="co">#This is actually the main function of the package and prepares everything to run the differential expression analysis.</span></span>
<span id="cb356-2"><a href="rna-seq-analysis.html#cb356-2"></a>dds.LNv &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> <span class="kw">as.matrix</span>(ds2_LNv_gene_expression), <span class="dt">colData =</span> colData.LNv,<span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>ZT)</span>
<span id="cb356-3"><a href="rna-seq-analysis.html#cb356-3"></a></span>
<span id="cb356-4"><a href="rna-seq-analysis.html#cb356-4"></a><span class="co">#We will create the heat map with the normalized counts</span></span>
<span id="cb356-5"><a href="rna-seq-analysis.html#cb356-5"></a><span class="co">#First we need the distance matrix </span></span>
<span id="cb356-6"><a href="rna-seq-analysis.html#cb356-6"></a>sampleDists &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">t</span>(<span class="kw">assay</span>(dds.LNv)))</span>
<span id="cb356-7"><a href="rna-seq-analysis.html#cb356-7"></a></span>
<span id="cb356-8"><a href="rna-seq-analysis.html#cb356-8"></a><span class="co">#Then we take this in a matrix and create the heatmap</span></span>
<span id="cb356-9"><a href="rna-seq-analysis.html#cb356-9"></a>sampleDistMatrix &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(sampleDists)</span>
<span id="cb356-10"><a href="rna-seq-analysis.html#cb356-10"></a><span class="kw">rownames</span>(sampleDistMatrix) &lt;-<span class="st"> </span><span class="kw">paste</span>(dds.LNv<span class="op">$</span>ZT,dds.LNv<span class="op">$</span>rep,<span class="dt">sep=</span><span class="st">&quot;_rep_&quot;</span>)</span>
<span id="cb356-11"><a href="rna-seq-analysis.html#cb356-11"></a><span class="kw">colnames</span>(sampleDistMatrix) &lt;-<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb356-12"><a href="rna-seq-analysis.html#cb356-12"></a>colors &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>( <span class="kw">rev</span>(<span class="kw">brewer.pal</span>(<span class="dv">9</span>, <span class="st">&quot;Blues&quot;</span>)) )(<span class="dv">255</span>)</span>
<span id="cb356-13"><a href="rna-seq-analysis.html#cb356-13"></a><span class="kw">pheatmap</span>(sampleDistMatrix,</span>
<span id="cb356-14"><a href="rna-seq-analysis.html#cb356-14"></a>         <span class="dt">clustering_distance_rows=</span>sampleDists,</span>
<span id="cb356-15"><a href="rna-seq-analysis.html#cb356-15"></a>         <span class="dt">clustering_distance_cols=</span>sampleDists,</span>
<span id="cb356-16"><a href="rna-seq-analysis.html#cb356-16"></a>         <span class="dt">col=</span>colors)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-501"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-50-1.png" alt="Heatmap and PCA in LNv cells" width="100%" />
<p class="caption">
Figure 6.15: Heatmap and PCA in LNv cells
</p>
</div>
<div class="sourceCode" id="cb357"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb357-1"><a href="rna-seq-analysis.html#cb357-1"></a><span class="co">#Now we do some transformation to see how the data looks like in terms of differences and if we should take any sample out.</span></span>
<span id="cb357-2"><a href="rna-seq-analysis.html#cb357-2"></a><span class="co">#please read the help doing ???rlog??</span></span>
<span id="cb357-3"><a href="rna-seq-analysis.html#cb357-3"></a>rdl.LNv &lt;-<span class="st"> </span><span class="kw">rlog</span>(dds.LNv) </span>
<span id="cb357-4"><a href="rna-seq-analysis.html#cb357-4"></a></span>
<span id="cb357-5"><a href="rna-seq-analysis.html#cb357-5"></a><span class="co">#now we can plot the PCA analysis</span></span>
<span id="cb357-6"><a href="rna-seq-analysis.html#cb357-6"></a><span class="kw">plotPCA</span>(rdl.LNv, <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-502"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-50-2.png" alt="Heatmap and PCA in LNv cells" width="100%" />
<p class="caption">
Figure 6.16: Heatmap and PCA in LNv cells
</p>
</div>
<div class="sourceCode" id="cb358"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb358-1"><a href="rna-seq-analysis.html#cb358-1"></a><span class="co">#and also we can work with the plot by taking the data out and using ggplot2</span></span>
<span id="cb358-2"><a href="rna-seq-analysis.html#cb358-2"></a>data &lt;-<span class="st"> </span><span class="kw">plotPCA</span>(rdl.LNv, <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>), <span class="dt">returnData=</span><span class="ot">TRUE</span>)</span>
<span id="cb358-3"><a href="rna-seq-analysis.html#cb358-3"></a>percentVar &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="dv">100</span> <span class="op">*</span><span class="st"> </span><span class="kw">attr</span>(data, <span class="st">&quot;percentVar&quot;</span>))</span>
<span id="cb358-4"><a href="rna-seq-analysis.html#cb358-4"></a></span>
<span id="cb358-5"><a href="rna-seq-analysis.html#cb358-5"></a><span class="kw">ggplot</span>(data, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>name)) <span class="op">+</span></span>
<span id="cb358-6"><a href="rna-seq-analysis.html#cb358-6"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb358-7"><a href="rna-seq-analysis.html#cb358-7"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> name),</span>
<span id="cb358-8"><a href="rna-seq-analysis.html#cb358-8"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb358-9"><a href="rna-seq-analysis.html#cb358-9"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb358-10"><a href="rna-seq-analysis.html#cb358-10"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb358-11"><a href="rna-seq-analysis.html#cb358-11"></a><span class="st">  </span><span class="kw">theme_minimal</span>()<span class="op">+</span></span>
<span id="cb358-12"><a href="rna-seq-analysis.html#cb358-12"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC1: &quot;</span>,percentVar[<span class="dv">1</span>],<span class="st">&quot;% variance&quot;</span>)) <span class="op">+</span></span>
<span id="cb358-13"><a href="rna-seq-analysis.html#cb358-13"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC2: &quot;</span>,percentVar[<span class="dv">2</span>],<span class="st">&quot;% variance&quot;</span>)) <span class="op">+</span></span>
<span id="cb358-14"><a href="rna-seq-analysis.html#cb358-14"></a><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title=</span><span class="st">&quot;PC1 vs PC2&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-503"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-50-3.png" alt="Heatmap and PCA in LNv cells" width="100%" />
<p class="caption">
Figure 6.17: Heatmap and PCA in LNv cells
</p>
</div>
<div class="sourceCode" id="cb359"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb359-1"><a href="rna-seq-analysis.html#cb359-1"></a><span class="co">#We can even create 3D plots for all the PC we want</span></span>
<span id="cb359-2"><a href="rna-seq-analysis.html#cb359-2"></a>ntop &lt;-<span class="st"> </span><span class="dv">1000</span></span>
<span id="cb359-3"><a href="rna-seq-analysis.html#cb359-3"></a>rv &lt;-<span class="st"> </span><span class="kw">rowVars</span>(<span class="kw">assay</span>(rdl.LNv))</span>
<span id="cb359-4"><a href="rna-seq-analysis.html#cb359-4"></a>select &lt;-<span class="st"> </span><span class="kw">order</span>(rv, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="kw">seq_len</span>(<span class="kw">min</span>(ntop, <span class="kw">length</span>(rv)))]</span>
<span id="cb359-5"><a href="rna-seq-analysis.html#cb359-5"></a>mat &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">assay</span>(rdl.LNv)[select, ])</span>
<span id="cb359-6"><a href="rna-seq-analysis.html#cb359-6"></a>pc &lt;-<span class="st"> </span><span class="kw">prcomp</span>(mat)</span>
<span id="cb359-7"><a href="rna-seq-analysis.html#cb359-7"></a>pc_df=<span class="kw">as.data.frame</span>(pc<span class="op">$</span>x)</span>
<span id="cb359-8"><a href="rna-seq-analysis.html#cb359-8"></a>pc_df=<span class="kw">merge</span>(pc_df,colData.LNv,<span class="dt">by=</span><span class="st">&quot;row.names&quot;</span>,<span class="dt">all=</span>T)</span>
<span id="cb359-9"><a href="rna-seq-analysis.html#cb359-9"></a><span class="kw">rownames</span>(pc_df)=pc_df<span class="op">$</span>Row.names</span>
<span id="cb359-10"><a href="rna-seq-analysis.html#cb359-10"></a></span>
<span id="cb359-11"><a href="rna-seq-analysis.html#cb359-11"></a><span class="co">#This analyze how much variation are we seeing in each PC</span></span>
<span id="cb359-12"><a href="rna-seq-analysis.html#cb359-12"></a><span class="kw">fviz_eig</span>(pc, <span class="dt">addlabels =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-504"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-50-4.png" alt="Heatmap and PCA in LNv cells" width="100%" />
<p class="caption">
Figure 6.18: Heatmap and PCA in LNv cells
</p>
</div>
<div class="sourceCode" id="cb360"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb360-1"><a href="rna-seq-analysis.html#cb360-1"></a>p1=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>Row.names)) <span class="op">+</span></span>
<span id="cb360-2"><a href="rna-seq-analysis.html#cb360-2"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb360-3"><a href="rna-seq-analysis.html#cb360-3"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Row.names),</span>
<span id="cb360-4"><a href="rna-seq-analysis.html#cb360-4"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb360-5"><a href="rna-seq-analysis.html#cb360-5"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb360-6"><a href="rna-seq-analysis.html#cb360-6"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb360-7"><a href="rna-seq-analysis.html#cb360-7"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb360-8"><a href="rna-seq-analysis.html#cb360-8"></a>p1</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-505"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-50-5.png" alt="Heatmap and PCA in LNv cells" width="100%" />
<p class="caption">
Figure 6.19: Heatmap and PCA in LNv cells
</p>
</div>
<div class="sourceCode" id="cb361"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb361-1"><a href="rna-seq-analysis.html#cb361-1"></a>p2=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC3, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>Row.names)) <span class="op">+</span></span>
<span id="cb361-2"><a href="rna-seq-analysis.html#cb361-2"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb361-3"><a href="rna-seq-analysis.html#cb361-3"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Row.names),</span>
<span id="cb361-4"><a href="rna-seq-analysis.html#cb361-4"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb361-5"><a href="rna-seq-analysis.html#cb361-5"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb361-6"><a href="rna-seq-analysis.html#cb361-6"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb361-7"><a href="rna-seq-analysis.html#cb361-7"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb361-8"><a href="rna-seq-analysis.html#cb361-8"></a>p2</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-506"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-50-6.png" alt="Heatmap and PCA in LNv cells" width="100%" />
<p class="caption">
Figure 6.20: Heatmap and PCA in LNv cells
</p>
</div>
<div class="sourceCode" id="cb362"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb362-1"><a href="rna-seq-analysis.html#cb362-1"></a><span class="co">#and both together</span></span>
<span id="cb362-2"><a href="rna-seq-analysis.html#cb362-2"></a><span class="kw">grid.arrange</span>(p1,p2, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-507"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-50-7.png" alt="Heatmap and PCA in LNv cells" width="100%" />
<p class="caption">
Figure 6.21: Heatmap and PCA in LNv cells
</p>
</div>
<pre><code>#We can do some fancy 3D plots also
plot_ly(pc_df, x = ~PC1, y = ~PC2, z = ~PC3, color = ~ZT,colorscale = c(&#39;#FFE1A1&#39;, &#39;#683531&#39;))

plot_ly(pc_df, x = ~PC2, y = ~PC3, z = ~PC4, color = ~ZT,colorscale = c(&#39;#FFE1A1&#39;, &#39;#683531&#39;))
</code></pre>
<p><strong>LNd</strong></p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb364-1"><a href="rna-seq-analysis.html#cb364-1"></a><span class="co">#This is actually the main function of the package and prepares everything to run the differential expression analysis.</span></span>
<span id="cb364-2"><a href="rna-seq-analysis.html#cb364-2"></a>dds.LNd &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> <span class="kw">as.matrix</span>(ds2_LNd_gene_expression), <span class="dt">colData =</span> colData.LNd,<span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>ZT)</span>
<span id="cb364-3"><a href="rna-seq-analysis.html#cb364-3"></a></span>
<span id="cb364-4"><a href="rna-seq-analysis.html#cb364-4"></a><span class="co">#We will create the heat map with the normalized counts</span></span>
<span id="cb364-5"><a href="rna-seq-analysis.html#cb364-5"></a><span class="co">#First we need the distance matrix </span></span>
<span id="cb364-6"><a href="rna-seq-analysis.html#cb364-6"></a>sampleDists &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">t</span>(<span class="kw">assay</span>(dds.LNd)))</span>
<span id="cb364-7"><a href="rna-seq-analysis.html#cb364-7"></a></span>
<span id="cb364-8"><a href="rna-seq-analysis.html#cb364-8"></a><span class="co">#Then we take this in a matrix and create the heatmap</span></span>
<span id="cb364-9"><a href="rna-seq-analysis.html#cb364-9"></a>sampleDistMatrix &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(sampleDists)</span>
<span id="cb364-10"><a href="rna-seq-analysis.html#cb364-10"></a><span class="kw">rownames</span>(sampleDistMatrix) &lt;-<span class="st"> </span><span class="kw">paste</span>(dds.LNd<span class="op">$</span>ZT,dds.LNd<span class="op">$</span>rep,<span class="dt">sep=</span><span class="st">&quot;_rep_&quot;</span>)</span>
<span id="cb364-11"><a href="rna-seq-analysis.html#cb364-11"></a><span class="kw">colnames</span>(sampleDistMatrix) &lt;-<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb364-12"><a href="rna-seq-analysis.html#cb364-12"></a>colors &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>( <span class="kw">rev</span>(<span class="kw">brewer.pal</span>(<span class="dv">9</span>, <span class="st">&quot;Blues&quot;</span>)) )(<span class="dv">255</span>)</span>
<span id="cb364-13"><a href="rna-seq-analysis.html#cb364-13"></a><span class="kw">pheatmap</span>(sampleDistMatrix,</span>
<span id="cb364-14"><a href="rna-seq-analysis.html#cb364-14"></a>         <span class="dt">clustering_distance_rows=</span>sampleDists,</span>
<span id="cb364-15"><a href="rna-seq-analysis.html#cb364-15"></a>         <span class="dt">clustering_distance_cols=</span>sampleDists,</span>
<span id="cb364-16"><a href="rna-seq-analysis.html#cb364-16"></a>         <span class="dt">col=</span>colors)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-511"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-51-1.png" alt="Heatmap and PCA in the LNd cells" width="100%" />
<p class="caption">
Figure 6.22: Heatmap and PCA in the LNd cells
</p>
</div>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="rna-seq-analysis.html#cb365-1"></a><span class="co">#Now we do some transformation to see how the data looks like in terms of differences and if we should take any sample out.</span></span>
<span id="cb365-2"><a href="rna-seq-analysis.html#cb365-2"></a><span class="co">#please read the help doing ???rlog??</span></span>
<span id="cb365-3"><a href="rna-seq-analysis.html#cb365-3"></a>rdl.LNd &lt;-<span class="st"> </span><span class="kw">rlog</span>(dds.LNd) </span>
<span id="cb365-4"><a href="rna-seq-analysis.html#cb365-4"></a></span>
<span id="cb365-5"><a href="rna-seq-analysis.html#cb365-5"></a><span class="co">#now we can plot the PCA analysis</span></span>
<span id="cb365-6"><a href="rna-seq-analysis.html#cb365-6"></a><span class="kw">plotPCA</span>(rdl.LNd, <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-512"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-51-2.png" alt="Heatmap and PCA in the LNd cells" width="100%" />
<p class="caption">
Figure 6.23: Heatmap and PCA in the LNd cells
</p>
</div>
<div class="sourceCode" id="cb366"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb366-1"><a href="rna-seq-analysis.html#cb366-1"></a><span class="co">#and also we can work with the plot by taking the data out and using ggplot2</span></span>
<span id="cb366-2"><a href="rna-seq-analysis.html#cb366-2"></a>data &lt;-<span class="st"> </span><span class="kw">plotPCA</span>(rdl.LNd, <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>), <span class="dt">returnData=</span><span class="ot">TRUE</span>)</span>
<span id="cb366-3"><a href="rna-seq-analysis.html#cb366-3"></a>percentVar &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="dv">100</span> <span class="op">*</span><span class="st"> </span><span class="kw">attr</span>(data, <span class="st">&quot;percentVar&quot;</span>))</span>
<span id="cb366-4"><a href="rna-seq-analysis.html#cb366-4"></a></span>
<span id="cb366-5"><a href="rna-seq-analysis.html#cb366-5"></a><span class="kw">ggplot</span>(data, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>name)) <span class="op">+</span></span>
<span id="cb366-6"><a href="rna-seq-analysis.html#cb366-6"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb366-7"><a href="rna-seq-analysis.html#cb366-7"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> name),</span>
<span id="cb366-8"><a href="rna-seq-analysis.html#cb366-8"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb366-9"><a href="rna-seq-analysis.html#cb366-9"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb366-10"><a href="rna-seq-analysis.html#cb366-10"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb366-11"><a href="rna-seq-analysis.html#cb366-11"></a><span class="st">  </span><span class="kw">theme_minimal</span>()<span class="op">+</span></span>
<span id="cb366-12"><a href="rna-seq-analysis.html#cb366-12"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC1: &quot;</span>,percentVar[<span class="dv">1</span>],<span class="st">&quot;% variance&quot;</span>)) <span class="op">+</span></span>
<span id="cb366-13"><a href="rna-seq-analysis.html#cb366-13"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC2: &quot;</span>,percentVar[<span class="dv">2</span>],<span class="st">&quot;% variance&quot;</span>)) <span class="op">+</span></span>
<span id="cb366-14"><a href="rna-seq-analysis.html#cb366-14"></a><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title=</span><span class="st">&quot;PC1 vs PC2&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-513"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-51-3.png" alt="Heatmap and PCA in the LNd cells" width="100%" />
<p class="caption">
Figure 6.24: Heatmap and PCA in the LNd cells
</p>
</div>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb367-1"><a href="rna-seq-analysis.html#cb367-1"></a><span class="co">#We can even create 3D plots for all the PC we want</span></span>
<span id="cb367-2"><a href="rna-seq-analysis.html#cb367-2"></a>ntop &lt;-<span class="st"> </span><span class="dv">1000</span></span>
<span id="cb367-3"><a href="rna-seq-analysis.html#cb367-3"></a>rv &lt;-<span class="st"> </span><span class="kw">rowVars</span>(<span class="kw">assay</span>(rdl.LNd))</span>
<span id="cb367-4"><a href="rna-seq-analysis.html#cb367-4"></a>select &lt;-<span class="st"> </span><span class="kw">order</span>(rv, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="kw">seq_len</span>(<span class="kw">min</span>(ntop, <span class="kw">length</span>(rv)))]</span>
<span id="cb367-5"><a href="rna-seq-analysis.html#cb367-5"></a>mat &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">assay</span>(rdl.LNd)[select, ])</span>
<span id="cb367-6"><a href="rna-seq-analysis.html#cb367-6"></a>pc &lt;-<span class="st"> </span><span class="kw">prcomp</span>(mat)</span>
<span id="cb367-7"><a href="rna-seq-analysis.html#cb367-7"></a>pc_df=<span class="kw">as.data.frame</span>(pc<span class="op">$</span>x)</span>
<span id="cb367-8"><a href="rna-seq-analysis.html#cb367-8"></a>pc_df=<span class="kw">merge</span>(pc_df,colData.LNd,<span class="dt">by=</span><span class="st">&quot;row.names&quot;</span>,<span class="dt">all=</span>T)</span>
<span id="cb367-9"><a href="rna-seq-analysis.html#cb367-9"></a><span class="kw">rownames</span>(pc_df)=pc_df<span class="op">$</span>Row.names</span>
<span id="cb367-10"><a href="rna-seq-analysis.html#cb367-10"></a></span>
<span id="cb367-11"><a href="rna-seq-analysis.html#cb367-11"></a><span class="co">#This analyze how much variation are we seeing in each PC</span></span>
<span id="cb367-12"><a href="rna-seq-analysis.html#cb367-12"></a><span class="kw">fviz_eig</span>(pc, <span class="dt">addlabels =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-514"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-51-4.png" alt="Heatmap and PCA in the LNd cells" width="100%" />
<p class="caption">
Figure 6.25: Heatmap and PCA in the LNd cells
</p>
</div>
<div class="sourceCode" id="cb368"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb368-1"><a href="rna-seq-analysis.html#cb368-1"></a>p1=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>Row.names)) <span class="op">+</span></span>
<span id="cb368-2"><a href="rna-seq-analysis.html#cb368-2"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb368-3"><a href="rna-seq-analysis.html#cb368-3"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Row.names),</span>
<span id="cb368-4"><a href="rna-seq-analysis.html#cb368-4"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb368-5"><a href="rna-seq-analysis.html#cb368-5"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb368-6"><a href="rna-seq-analysis.html#cb368-6"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb368-7"><a href="rna-seq-analysis.html#cb368-7"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb368-8"><a href="rna-seq-analysis.html#cb368-8"></a>p1</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-515"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-51-5.png" alt="Heatmap and PCA in the LNd cells" width="100%" />
<p class="caption">
Figure 6.26: Heatmap and PCA in the LNd cells
</p>
</div>
<div class="sourceCode" id="cb369"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb369-1"><a href="rna-seq-analysis.html#cb369-1"></a>p2=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC3, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>Row.names)) <span class="op">+</span></span>
<span id="cb369-2"><a href="rna-seq-analysis.html#cb369-2"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb369-3"><a href="rna-seq-analysis.html#cb369-3"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Row.names),</span>
<span id="cb369-4"><a href="rna-seq-analysis.html#cb369-4"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb369-5"><a href="rna-seq-analysis.html#cb369-5"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb369-6"><a href="rna-seq-analysis.html#cb369-6"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb369-7"><a href="rna-seq-analysis.html#cb369-7"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb369-8"><a href="rna-seq-analysis.html#cb369-8"></a>p2</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-516"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-51-6.png" alt="Heatmap and PCA in the LNd cells" width="100%" />
<p class="caption">
Figure 6.27: Heatmap and PCA in the LNd cells
</p>
</div>
<div class="sourceCode" id="cb370"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb370-1"><a href="rna-seq-analysis.html#cb370-1"></a><span class="co">#and both together</span></span>
<span id="cb370-2"><a href="rna-seq-analysis.html#cb370-2"></a><span class="kw">grid.arrange</span>(p1,p2, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-517"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-51-7.png" alt="Heatmap and PCA in the LNd cells" width="100%" />
<p class="caption">
Figure 6.28: Heatmap and PCA in the LNd cells
</p>
</div>
<pre><code>#We can do some fancy 3D plots also
plot_ly(pc_df, x = ~PC1, y = ~PC2, z = ~PC3, color = ~ZT,colorscale = c(&#39;#FFE1A1&#39;, &#39;#683531&#39;))

plot_ly(pc_df, x = ~PC2, y = ~PC3, z = ~PC4, color = ~ZT,colorscale = c(&#39;#FFE1A1&#39;, &#39;#683531&#39;))
</code></pre>
<p><strong>TH</strong></p>
<div class="sourceCode" id="cb372"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb372-1"><a href="rna-seq-analysis.html#cb372-1"></a><span class="co">#This is actually the main function of the package and prepares everything to run the differential expression analysis.</span></span>
<span id="cb372-2"><a href="rna-seq-analysis.html#cb372-2"></a>dds.TH &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> <span class="kw">as.matrix</span>(ds2_TH_gene_expression), <span class="dt">colData =</span> colData.TH,<span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>ZT)</span>
<span id="cb372-3"><a href="rna-seq-analysis.html#cb372-3"></a></span>
<span id="cb372-4"><a href="rna-seq-analysis.html#cb372-4"></a><span class="co">#We will create the heat map with the normalized counts</span></span>
<span id="cb372-5"><a href="rna-seq-analysis.html#cb372-5"></a><span class="co">#First we need the distnace matrix </span></span>
<span id="cb372-6"><a href="rna-seq-analysis.html#cb372-6"></a>sampleDists &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">t</span>(<span class="kw">assay</span>(dds.TH)))</span>
<span id="cb372-7"><a href="rna-seq-analysis.html#cb372-7"></a></span>
<span id="cb372-8"><a href="rna-seq-analysis.html#cb372-8"></a><span class="co">#Then we take this in a matrix and create the heatmap</span></span>
<span id="cb372-9"><a href="rna-seq-analysis.html#cb372-9"></a>sampleDistMatrix &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(sampleDists)</span>
<span id="cb372-10"><a href="rna-seq-analysis.html#cb372-10"></a><span class="kw">rownames</span>(sampleDistMatrix) &lt;-<span class="st"> </span><span class="kw">paste</span>(dds.TH<span class="op">$</span>ZT,dds.TH<span class="op">$</span>rep,<span class="dt">sep=</span><span class="st">&quot;_rep_&quot;</span>)</span>
<span id="cb372-11"><a href="rna-seq-analysis.html#cb372-11"></a><span class="kw">colnames</span>(sampleDistMatrix) &lt;-<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb372-12"><a href="rna-seq-analysis.html#cb372-12"></a>colors &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>( <span class="kw">rev</span>(<span class="kw">brewer.pal</span>(<span class="dv">9</span>, <span class="st">&quot;Blues&quot;</span>)) )(<span class="dv">255</span>)</span>
<span id="cb372-13"><a href="rna-seq-analysis.html#cb372-13"></a><span class="kw">pheatmap</span>(sampleDistMatrix,</span>
<span id="cb372-14"><a href="rna-seq-analysis.html#cb372-14"></a>         <span class="dt">clustering_distance_rows=</span>sampleDists,</span>
<span id="cb372-15"><a href="rna-seq-analysis.html#cb372-15"></a>         <span class="dt">clustering_distance_cols=</span>sampleDists,</span>
<span id="cb372-16"><a href="rna-seq-analysis.html#cb372-16"></a>         <span class="dt">col=</span>colors)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-521"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-52-1.png" alt="Heatmap and PCA in TH cells" width="100%" />
<p class="caption">
Figure 6.29: Heatmap and PCA in TH cells
</p>
</div>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb373-1"><a href="rna-seq-analysis.html#cb373-1"></a><span class="co">#This is another way to see this distances. </span></span>
<span id="cb373-2"><a href="rna-seq-analysis.html#cb373-2"></a><span class="co">#We are doing a correlation between samples and then a clustering and a phylogram plot</span></span>
<span id="cb373-3"><a href="rna-seq-analysis.html#cb373-3"></a><span class="co">#d &lt;- cor(assay(dds.TH), method=&quot;spearman&quot;)</span></span>
<span id="cb373-4"><a href="rna-seq-analysis.html#cb373-4"></a><span class="co">#hc &lt;- hclust(dist(1-d))</span></span>
<span id="cb373-5"><a href="rna-seq-analysis.html#cb373-5"></a><span class="co">#plot.phylo(as.phylo(hc), type=&quot;p&quot;, edge.col=4, edge.width=3, show.node.label=TRUE, no.margin=TRUE)</span></span>
<span id="cb373-6"><a href="rna-seq-analysis.html#cb373-6"></a></span>
<span id="cb373-7"><a href="rna-seq-analysis.html#cb373-7"></a><span class="co">#Now we do some transformation to see how the data looks like in terms of differences and if we should take any sample out.</span></span>
<span id="cb373-8"><a href="rna-seq-analysis.html#cb373-8"></a><span class="co">#please read the help doing ???rlog??</span></span>
<span id="cb373-9"><a href="rna-seq-analysis.html#cb373-9"></a>rdl.TH &lt;-<span class="st"> </span><span class="kw">rlog</span>(dds.TH) </span>
<span id="cb373-10"><a href="rna-seq-analysis.html#cb373-10"></a></span>
<span id="cb373-11"><a href="rna-seq-analysis.html#cb373-11"></a><span class="co">#now we can plot the PCA analysis</span></span>
<span id="cb373-12"><a href="rna-seq-analysis.html#cb373-12"></a><span class="kw">plotPCA</span>(rdl.TH, <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-522"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-52-2.png" alt="Heatmap and PCA in TH cells" width="100%" />
<p class="caption">
Figure 6.30: Heatmap and PCA in TH cells
</p>
</div>
<div class="sourceCode" id="cb374"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb374-1"><a href="rna-seq-analysis.html#cb374-1"></a><span class="co">#and also we can work with the plot by taking the data out and using ggplot2</span></span>
<span id="cb374-2"><a href="rna-seq-analysis.html#cb374-2"></a>data &lt;-<span class="st"> </span><span class="kw">plotPCA</span>(rdl.TH, <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>), <span class="dt">returnData=</span><span class="ot">TRUE</span>)</span>
<span id="cb374-3"><a href="rna-seq-analysis.html#cb374-3"></a>percentVar &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="dv">100</span> <span class="op">*</span><span class="st"> </span><span class="kw">attr</span>(data, <span class="st">&quot;percentVar&quot;</span>))</span>
<span id="cb374-4"><a href="rna-seq-analysis.html#cb374-4"></a></span>
<span id="cb374-5"><a href="rna-seq-analysis.html#cb374-5"></a><span class="kw">ggplot</span>(data, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>name)) <span class="op">+</span></span>
<span id="cb374-6"><a href="rna-seq-analysis.html#cb374-6"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb374-7"><a href="rna-seq-analysis.html#cb374-7"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> name),</span>
<span id="cb374-8"><a href="rna-seq-analysis.html#cb374-8"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb374-9"><a href="rna-seq-analysis.html#cb374-9"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb374-10"><a href="rna-seq-analysis.html#cb374-10"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb374-11"><a href="rna-seq-analysis.html#cb374-11"></a><span class="st">  </span><span class="kw">theme_minimal</span>()<span class="op">+</span></span>
<span id="cb374-12"><a href="rna-seq-analysis.html#cb374-12"></a><span class="st">  </span><span class="kw">xlab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC1: &quot;</span>,percentVar[<span class="dv">1</span>],<span class="st">&quot;% variance&quot;</span>)) <span class="op">+</span></span>
<span id="cb374-13"><a href="rna-seq-analysis.html#cb374-13"></a><span class="st">  </span><span class="kw">ylab</span>(<span class="kw">paste0</span>(<span class="st">&quot;PC2: &quot;</span>,percentVar[<span class="dv">2</span>],<span class="st">&quot;% variance&quot;</span>)) <span class="op">+</span></span>
<span id="cb374-14"><a href="rna-seq-analysis.html#cb374-14"></a><span class="st">  </span><span class="kw">labs</span>(<span class="dt">title=</span><span class="st">&quot;PC1 vs PC2&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-523"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-52-3.png" alt="Heatmap and PCA in TH cells" width="100%" />
<p class="caption">
Figure 6.31: Heatmap and PCA in TH cells
</p>
</div>
<div class="sourceCode" id="cb375"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb375-1"><a href="rna-seq-analysis.html#cb375-1"></a><span class="co">#We can even create 3D plots for all the PC we want</span></span>
<span id="cb375-2"><a href="rna-seq-analysis.html#cb375-2"></a>ntop &lt;-<span class="st"> </span><span class="dv">1000</span></span>
<span id="cb375-3"><a href="rna-seq-analysis.html#cb375-3"></a>rv &lt;-<span class="st"> </span><span class="kw">rowVars</span>(<span class="kw">assay</span>(rdl.TH))</span>
<span id="cb375-4"><a href="rna-seq-analysis.html#cb375-4"></a>select &lt;-<span class="st"> </span><span class="kw">order</span>(rv, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="kw">seq_len</span>(<span class="kw">min</span>(ntop, <span class="kw">length</span>(rv)))]</span>
<span id="cb375-5"><a href="rna-seq-analysis.html#cb375-5"></a>mat &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">assay</span>(rdl.TH)[select, ])</span>
<span id="cb375-6"><a href="rna-seq-analysis.html#cb375-6"></a>pc &lt;-<span class="st"> </span><span class="kw">prcomp</span>(mat)</span>
<span id="cb375-7"><a href="rna-seq-analysis.html#cb375-7"></a>pc_df=<span class="kw">as.data.frame</span>(pc<span class="op">$</span>x)</span>
<span id="cb375-8"><a href="rna-seq-analysis.html#cb375-8"></a>pc_df=<span class="kw">merge</span>(pc_df,colData.TH,<span class="dt">by=</span><span class="st">&quot;row.names&quot;</span>,<span class="dt">all=</span>T)</span>
<span id="cb375-9"><a href="rna-seq-analysis.html#cb375-9"></a><span class="kw">rownames</span>(pc_df)=pc_df<span class="op">$</span>Row.names</span>
<span id="cb375-10"><a href="rna-seq-analysis.html#cb375-10"></a></span>
<span id="cb375-11"><a href="rna-seq-analysis.html#cb375-11"></a><span class="co">#This analyze how much variation are we seeing in each PC</span></span>
<span id="cb375-12"><a href="rna-seq-analysis.html#cb375-12"></a><span class="kw">fviz_eig</span>(pc, <span class="dt">addlabels =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-524"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-52-4.png" alt="Heatmap and PCA in TH cells" width="100%" />
<p class="caption">
Figure 6.32: Heatmap and PCA in TH cells
</p>
</div>
<div class="sourceCode" id="cb376"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb376-1"><a href="rna-seq-analysis.html#cb376-1"></a>p1=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>Row.names)) <span class="op">+</span></span>
<span id="cb376-2"><a href="rna-seq-analysis.html#cb376-2"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb376-3"><a href="rna-seq-analysis.html#cb376-3"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Row.names),</span>
<span id="cb376-4"><a href="rna-seq-analysis.html#cb376-4"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb376-5"><a href="rna-seq-analysis.html#cb376-5"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb376-6"><a href="rna-seq-analysis.html#cb376-6"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb376-7"><a href="rna-seq-analysis.html#cb376-7"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb376-8"><a href="rna-seq-analysis.html#cb376-8"></a>p1</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-525"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-52-5.png" alt="Heatmap and PCA in TH cells" width="100%" />
<p class="caption">
Figure 6.33: Heatmap and PCA in TH cells
</p>
</div>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb377-1"><a href="rna-seq-analysis.html#cb377-1"></a>p2=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC3, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>Row.names)) <span class="op">+</span></span>
<span id="cb377-2"><a href="rna-seq-analysis.html#cb377-2"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb377-3"><a href="rna-seq-analysis.html#cb377-3"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Row.names),</span>
<span id="cb377-4"><a href="rna-seq-analysis.html#cb377-4"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb377-5"><a href="rna-seq-analysis.html#cb377-5"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb377-6"><a href="rna-seq-analysis.html#cb377-6"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb377-7"><a href="rna-seq-analysis.html#cb377-7"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb377-8"><a href="rna-seq-analysis.html#cb377-8"></a>p2</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-526"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-52-6.png" alt="Heatmap and PCA in TH cells" width="100%" />
<p class="caption">
Figure 6.34: Heatmap and PCA in TH cells
</p>
</div>
<div class="sourceCode" id="cb378"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb378-1"><a href="rna-seq-analysis.html#cb378-1"></a><span class="co">#and both together</span></span>
<span id="cb378-2"><a href="rna-seq-analysis.html#cb378-2"></a><span class="kw">grid.arrange</span>(p1,p2, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-527"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-52-7.png" alt="Heatmap and PCA in TH cells" width="100%" />
<p class="caption">
Figure 6.35: Heatmap and PCA in TH cells
</p>
</div>
<pre><code>#We can do some fancy 3D plots also
plot_ly(pc_df, x = ~PC1, y = ~PC2, z = ~PC3, color = ~ZT,colorscale = c(&#39;#FFE1A1&#39;, &#39;#683531&#39;))

plot_ly(pc_df, x = ~PC2, y = ~PC3, z = ~PC4, color = ~ZT,colorscale = c(&#39;#FFE1A1&#39;, &#39;#683531&#39;))
</code></pre>
</div>
<div id="extra-advanced-cycling-genes-analysis" class="section level2" number="6.7">
<h2><span class="header-section-number">6.7</span> Extra: Advanced cycling genes analysis</h2>
<p>This is a more complex set-up I arrived after many trials this is way I found to get more reliable results.</p>
<ol style="list-style-type: decimal">
<li>Before doing the analysis I normalize the expression by the maximum level in each replicate. This mean that for all the replicates the levels will be between 0 and 1.</li>
<li>I calculate the amplitud manually</li>
</ol>
<p>I share here some of the loops to do this</p>
<pre><code>for (i in ls(pattern = &quot;^only&quot;)[-2]){
  
  if (i==&quot;onlyCS.ZT.ane&quot;){
    print(i)
    
    x&lt;-get(i)
    ind=apply(x[-1], 1, function(x)(sum(x)&gt;0))
    x=x[ind,]
    assign(i,x)
    print(names(x))
    
    #I separate them by replicate
    x1&lt;-x[,c(1,grep(pattern = &quot;_1&quot;,names(x)))]
    x2&lt;-x[,c(1,grep(&quot;_2&quot;,names(x)))]
    x3&lt;-x[,c(1,grep(&quot;_3&quot;,names(x)))]
    #I apply a funciton to see that they are all non-cero
    #ind1=apply(x1[-1], 1, function(x)(sum(x)&gt;0))
    #ind2=apply(x2[,-1], 1, function(x)(sum(x)&gt;0))
    #ind2=apply(x3[,-1], 1, function(x)(sum(x)&gt;0))
    
    #I then do the normalization by max
    x1&lt;-t(apply((x1[,-1]), 1, function(x)(x/max(x))))
    x1&lt;-cbind(as.data.frame(x$Gene),x1)
    names(x1)[1]=&quot;Gene&quot;
    x2&lt;-t(apply((x2[,-1]), 1, function(x)(x/max(x))))
    x3&lt;-t(apply((x3[,-1]), 1, function(x)(x/max(x)))) 
    #I bind them
    x&lt;-cbind(x1,x2)
    x&lt;-cbind(x,x3)
    x&lt;-x[complete.cases(x),]
    n&lt;-paste0(&quot;normtomax_&quot;,i)
    print(n)
    assign(n,x)
  }
  
  else{
    print(i)
    x&lt;-get(i)
    
    #I convert them regular reads, not log2
    x[,-1]&lt;-(2^ x[,-1])
    #then I apply the sum &gt; 0
    ind=apply(x[-1], 1, function(x)(sum(x)&gt;0))
    x=x[ind,]
    assign(i,x)
    print(names(x))
    
    #I separate them by replicate
    x1&lt;-x[,c(1,grep(pattern = &quot;Rep1&quot;,names(x)))]
    x2&lt;-x[,c(1,grep(&quot;Rep2&quot;,names(x)))]
    x3&lt;-x[,c(1,grep(&quot;Rep3&quot;,names(x)))]
    #I apply a funciton to see that they are all non-cero
    #ind1=apply(x1[-1], 1, function(x)(sum(x)&gt;0))
    #ind2=apply(x2[,-1], 1, function(x)(sum(x)&gt;0))
    #ind2=apply(x3[,-1], 1, function(x)(sum(x)&gt;0))
    
    #I then do the normalization by max
    x1&lt;-t(apply((x1[,-1]), 1, function(x)(x/max(x))))
    x1&lt;-cbind(as.data.frame(x$Gene),x1)
    names(x1)[1]=&quot;Gene&quot;
    x2&lt;-t(apply((x2[,-1]), 1, function(x)(x/max(x))))
    x3&lt;-t(apply((x3[,-1]), 1, function(x)(x/max(x)))) 
    #I bind them
    x&lt;-cbind(x1,x2)
    x&lt;-cbind(x,x3)
    x&lt;-x[complete.cases(x),]
    n&lt;-paste0(&quot;normtomax_&quot;,i)
    print(n)
    assign(n,x)
  }

}

#####circadian analysis####

for (i in ls(pattern = &quot;^normtomax&quot;)){
  print(i)
  #create names I need
  n=paste0(i,&quot;.csv&quot;)
  nn=paste0(&quot;circ.ana_&quot;,i)
  #write the table to do analysis
  write.table(get(i), file=n, row.names=FALSE,sep = &quot;,&quot;)
  #run analysis for anes names and deplanke names
  head(get(i))
  if (i==&quot;normtomax_onlyCS.ZT.ane&quot;){
    assign(nn,meta2d(infile=n,filestyle=&quot;csv&quot;, timepoints=as.numeric(gsub(pattern = &quot;ZT&quot;,replacement = &quot;&quot;,(sapply(strsplit((names(get(i)))[-1],&quot;_&quot;),&quot;[[&quot;,1)))),outputFile=FALSE,outRawData=F))
  }
  else if (i==&quot;normtomax_onlyw1118.ZT&quot;) {
    assign(nn,meta2d(infile=n,filestyle=&quot;csv&quot;, timepoints=as.numeric(gsub(pattern = &quot;ZT&quot;,replacement = &quot;&quot;,(sapply(strsplit((names(get(i)))[-1],&quot;_&quot;),&quot;[[&quot;,3)))),outputFile=FALSE,outRawData=F))
  }
  else {
    assign(nn,meta2d(infile=n,filestyle=&quot;csv&quot;, timepoints=as.numeric(gsub(pattern = &quot;CT&quot;,replacement = &quot;&quot;,(sapply(strsplit((names(get(i)))[-1],&quot;_&quot;),&quot;[[&quot;,3)))),outputFile=FALSE,outRawData=F))
  }
}

rm(list = ls(pattern = &quot;meta2d&quot;))
for (i in ls(pattern = &quot;^circ.ana&quot;)){
  print(i)
  n=paste0(&quot;meta2d&quot;,sep=&quot;_&quot;,i)
  assign(n,get(i)[[4]])
}

mean_nona=function(x){
  mean(x,na.rm =TRUE)
}


merge_counts=function(cyc,cs){
  cs$mean.exp=apply(cs[,-c(1)],1,mean_nona)
  #cyc$name=cyc$CycID
  cs=cs[order(cs$Gene),]
  cyc=cyc[order(cyc$CycID),]
  print(unique(cyc$CycID==cs$Gene))
  cyc&lt;-cbind(cyc, cs)
  
  toMatch &lt;- c(&quot;_1_&quot;, &quot;_2_&quot;, &quot;_3_&quot;)
  cyc$zero=(apply((cyc)[,grep(paste(toMatch,collapse=&quot;|&quot;),names(cyc))],1,function(x)(length(which(x %in% 0))&gt;1)))
  cyc=cyc[cyc$zero==F,]
  cyc$AMP_1=(apply((cyc)[,grep(&quot;_1_&quot;,names(cyc))],1,function(x)(1/min(x))))
  cyc$AMP_2=(apply((cyc)[,grep(&quot;_2_&quot;,names(cyc))],1,function(x)(1/min(x))))
  #cyc$AMP_3=(apply((cyc)[,grep(&quot;_3_&quot;,names(cyc))],1,function(x)(1/min(x))))
  cyc$AMP_min=(apply((cyc)[,grep(&quot;^AMP&quot;,names(cyc))],1,function(x)(min(x))))
  cyc$AMP_min[cyc$AMP_min==&quot;Inf&quot;]=10
  
  return(cyc)
}

</code></pre>
</div>
<div id="extra-negative-binomial-model" class="section level2" number="6.8">
<h2><span class="header-section-number">6.8</span> Extra: negative binomial model</h2>
<p>The model would be then: counts Kij for gene i, sample j are modeled using a negative binomial distribution with fitted mean μij and a gene-specific dispersion parameter αi. The fitted mean is composed of a sample-specific size factor sj and a parameter qij proportional to the expected true concentration of fragments for sample j.</p>
<p><span class="math display">\[
K_{ij}∼NB(μ_{ij},α_i)
\]</span>
<span class="math display">\[
μij=s_jq_{ij}
\]</span></p>
</div>
<div id="resources-and-bibliography-2" class="section level2" number="6.9">
<h2><span class="header-section-number">6.9</span> Resources and Bibliography</h2>
<p>Dubowy C, Sehgal A. Circadian Rhythms and Sleep in Drosophila melanogaster. Genetics. 2017;205(4):1373-1397. <a href="doi:10.1534/genetics.115.185157" class="uri">doi:10.1534/genetics.115.185157</a></p>
<p>MEIRELES-FILHO, Antonio Carlos Alves and KYRIACOU, Charalambos Panayiotis. Circadian rhythms in insect disease vectors. Mem. Inst. Oswaldo Cruz [online]. 2013, vol.108, suppl.1 [cited 2020-07-08], pp.48-58.</p>
<p>Abruzzi KC, Zadina A, Luo W, Wiyanto E, Rahman R, Guo F, et al. (2017) RNA-seq analysis of Drosophila clock and non-clock neurons reveals neuron-specific cycling and novel candidate neuropeptides. PLoS Genet 13(2): e1006613. <a href="https://doi.org/10.1371/journal.pgen.1006613" class="uri">https://doi.org/10.1371/journal.pgen.1006613</a></p>
<p>Schubert FK, Hagedorn N, Yoshii T, Helfrich-Förster C, Rieger D. Neuroanatomical details of the lateral neurons of Drosophila melanogaster support their functional role in the circadian system. J Comp Neurol. 2018;526(7):1209-1231. <a href="doi:10.1002/cne.24406" class="uri">doi:10.1002/cne.24406</a></p>
<p><a href="http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" class="uri">http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></p>
<p>Introduction to implementation steps of MetaCycle
Gang Wu, Ron Anafi, Michael Hughes, Karl Kornacker, and John Hogenesch
2015-12-04
<a href="https://cran.r-project.org/web/packages/MetaCycle/vignettes/implementation.html" class="uri">https://cran.r-project.org/web/packages/MetaCycle/vignettes/implementation.html</a></p>

</div>
</div>



</div>
<div class="psyteachr_footer">
  
</div>
<script>

/* update total correct if #total_correct exists */
update_total_correct = function() {
  if (t = document.getElementById("total_correct")) {
    t.innerHTML =
      document.getElementsByClassName("correct").length + " of " +
      document.getElementsByClassName("solveme").length + " correct";
  }
}

/* solution button toggling function */
b_func = function() {
  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "");
  }
  
  if (my_answer !== "" & real_answers.includes(my_answer)) {
    cl.add("correct");
  } else {
    cl.remove("correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol){
    var tol = JSON.parse(this.dataset.tol);  
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("correct");
    } else {
      cl.remove("correct");
    }  
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("correct");
    }  
  }
  
  update_total_correct();
}

window.onload = function() {
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('solution')) {
      buttons[i].onclick = b_func;
    }
  }
  
  /* set up solveme inputs */
  var solveme = document.getElementsByClassName("solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off"); 
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";
    
    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;
    
    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;
  }
  
  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="chip-sequencing-analysis.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"],
"google": false
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": {},
"toc": {
"collapse": "section",
"scroll_highlight": true
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
