<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Lab 6 RNA-seq analysis | Genomics with R for biologists</title>
  <meta name="description" content="This is the index for the coursebook" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="Lab 6 RNA-seq analysis | Genomics with R for biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is the index for the coursebook" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Lab 6 RNA-seq analysis | Genomics with R for biologists" />
  
  <meta name="twitter:description" content="This is the index for the coursebook" />
  

<meta name="author" content="Ines Lucia Patop inespatop@brandeis.edu" />


<meta name="date" content="2020-08-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="chip-sequencing-analysis.html"/>
<link rel="next" href="circadian-analysis.html"/>
<script src="libs/header-attrs-2.3/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="include/psyteachr.css" type="text/css" />
<link rel="stylesheet" href="include/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Genomics with R for biologists</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Overview</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#course-aims"><i class="fa fa-check"></i><b>0.1</b> Course Aims</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#intended-learning-outcomes"><i class="fa fa-check"></i><b>0.2</b> Intended Learning Outcomes</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#next-steps"><i class="fa fa-check"></i><b>0.3</b> Next steps</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#data-availability"><i class="fa fa-check"></i><b>0.4</b> Data availability</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#acknowledgments"><i class="fa fa-check"></i><b>0.5</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html"><i class="fa fa-check"></i><b>1</b> Introduction to R and RStudio</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>1.1</b> Objectives</a></li>
<li class="chapter" data-level="1.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>1.2</b> Introduction</a></li>
<li class="chapter" data-level="1.3" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#rstudio-world"><i class="fa fa-check"></i><b>1.3</b> RStudio world</a></li>
<li class="chapter" data-level="1.4" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#create-a-project"><i class="fa fa-check"></i><b>1.4</b> Create a project</a></li>
<li class="chapter" data-level="1.5" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#useful-shortcuts-in-rstudio"><i class="fa fa-check"></i><b>1.5</b> Useful shortcuts in RStudio:</a></li>
<li class="chapter" data-level="1.6" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#installing-and-loading-packages"><i class="fa fa-check"></i><b>1.6</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.7" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#creating-objects"><i class="fa fa-check"></i><b>1.7</b> Creating objects</a></li>
<li class="chapter" data-level="1.8" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#r-as-a-calculator"><i class="fa fa-check"></i><b>1.8</b> R as a calculator</a></li>
<li class="chapter" data-level="1.9" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#functions"><i class="fa fa-check"></i><b>1.9</b> Functions</a></li>
<li class="chapter" data-level="1.10" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#loops"><i class="fa fa-check"></i><b>1.10</b> Loops</a></li>
<li class="chapter" data-level="1.11" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#conditions"><i class="fa fa-check"></i><b>1.11</b> Conditions</a></li>
<li class="chapter" data-level="1.12" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#character-objects"><i class="fa fa-check"></i><b>1.12</b> Character objects</a></li>
<li class="chapter" data-level="1.13" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#identify"><i class="fa fa-check"></i><b>1.13</b> Identify</a></li>
<li class="chapter" data-level="1.14" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#modify"><i class="fa fa-check"></i><b>1.14</b> Modify</a></li>
<li class="chapter" data-level="1.15" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#useful-resources"><i class="fa fa-check"></i><b>1.15</b> Useful resources</a></li>
<li class="chapter" data-level="1.16" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#activity"><i class="fa fa-check"></i><b>1.16</b> Activity:</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-manipulation.html"><a href="data-manipulation.html"><i class="fa fa-check"></i><b>2</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="2.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>2.1</b> Objectives</a></li>
<li class="chapter" data-level="2.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>2.2</b> Introduction</a></li>
<li class="chapter" data-level="2.3" data-path="data-manipulation.html"><a href="data-manipulation.html#load-data"><i class="fa fa-check"></i><b>2.3</b> Load data</a></li>
<li class="chapter" data-level="2.4" data-path="data-manipulation.html"><a href="data-manipulation.html#data-exploration"><i class="fa fa-check"></i><b>2.4</b> Data exploration</a></li>
<li class="chapter" data-level="2.5" data-path="data-manipulation.html"><a href="data-manipulation.html#subsetting"><i class="fa fa-check"></i><b>2.5</b> Subsetting</a></li>
<li class="chapter" data-level="2.6" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#activity"><i class="fa fa-check"></i><b>2.6</b> Activity:</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html"><i class="fa fa-check"></i><b>3</b> Visualization and Statistical test</a>
<ul>
<li class="chapter" data-level="3.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#objectives-1"><i class="fa fa-check"></i><b>3.1</b> Objectives</a></li>
<li class="chapter" data-level="3.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#introduction-1"><i class="fa fa-check"></i><b>3.2</b> Introduction</a></li>
<li class="chapter" data-level="3.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#plots"><i class="fa fa-check"></i><b>3.3</b> Plots</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#line-and-points"><i class="fa fa-check"></i><b>3.3.1</b> Line and points</a></li>
<li class="chapter" data-level="3.3.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#historgam-and-density-plots"><i class="fa fa-check"></i><b>3.3.2</b> Historgam and density plots</a></li>
<li class="chapter" data-level="3.3.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#boxplot"><i class="fa fa-check"></i><b>3.3.3</b> Boxplot</a></li>
<li class="chapter" data-level="3.3.4" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#saving-plots"><i class="fa fa-check"></i><b>3.3.4</b> Saving plots</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#statistical-test"><i class="fa fa-check"></i><b>3.4</b> Statistical test</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#descriptive-statistics"><i class="fa fa-check"></i><b>3.4.1</b> Descriptive statistics</a></li>
<li class="chapter" data-level="3.4.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#t-testwilcoxon"><i class="fa fa-check"></i><b>3.4.2</b> T-test/Wilcoxon</a></li>
<li class="chapter" data-level="3.4.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#anovakruskalwallis"><i class="fa fa-check"></i><b>3.4.3</b> Anova/Kruskal–Wallis</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#activity-1"><i class="fa fa-check"></i><b>3.5</b> Activity:</a></li>
<li class="chapter" data-level="3.6" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#resources"><i class="fa fa-check"></i><b>3.6</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html"><i class="fa fa-check"></i><b>4</b> Sequencing techniques and preprocessing</a>
<ul>
<li class="chapter" data-level="4.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>4.1</b> Objectives</a></li>
<li class="chapter" data-level="4.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#different-type-of-data-dna-rna-splicing-single-end-paired-end"><i class="fa fa-check"></i><b>4.3</b> Different type of data (DNA, RNA, splicing, single-end paired-end)</a></li>
<li class="chapter" data-level="4.4" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#from-the-sequencer-machine-to-the-sequence"><i class="fa fa-check"></i><b>4.4</b> From the sequencer machine to the sequence</a></li>
<li class="chapter" data-level="4.5" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#general-processing"><i class="fa fa-check"></i><b>4.5</b> General processing</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#visualization-of-alignment-igv"><i class="fa fa-check"></i><b>4.5.1</b> Visualization of alignment (IGV)</a></li>
<li class="chapter" data-level="4.5.2" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#gene-counts-read-esat-feature-counts-peak-calling"><i class="fa fa-check"></i><b>4.5.2</b> Gene counts / read (esat feature counts) /peak calling</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#using-the-terminal"><i class="fa fa-check"></i><b>4.6</b> Using the terminal</a></li>
<li class="chapter" data-level="4.7" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#a-bit-of-shell-commands"><i class="fa fa-check"></i><b>4.7</b> A bit of shell commands</a></li>
<li class="chapter" data-level="4.8" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#resources-and-bibliography"><i class="fa fa-check"></i><b>4.8</b> Resources and Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html"><i class="fa fa-check"></i><b>5</b> Chip Sequencing Analysis</a>
<ul>
<li class="chapter" data-level="5.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>5.1</b> Objectives</a></li>
<li class="chapter" data-level="5.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>5.2</b> Introduction</a></li>
<li class="chapter" data-level="5.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#data-pre-processing-peak-calling"><i class="fa fa-check"></i><b>5.3</b> Data pre-processing (peak-calling):</a></li>
<li class="chapter" data-level="5.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#exploring-the-results"><i class="fa fa-check"></i><b>5.4</b> Exploring the results:</a></li>
<li class="chapter" data-level="5.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#annotating-the-peaks."><i class="fa fa-check"></i><b>5.5</b> Annotating the peaks.</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#read-the-annotation-file"><i class="fa fa-check"></i><b>5.5.1</b> Read the annotation file</a></li>
<li class="chapter" data-level="5.5.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#get-the-genes-from-this-annotation-data-base"><i class="fa fa-check"></i><b>5.5.2</b> Get the genes from this annotation data base</a></li>
<li class="chapter" data-level="5.5.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#get-the-genomic-ranges-of-the-peaks-from-macs-and-find-which-gene-is-closer-to-them-ie.-annotate-them"><i class="fa fa-check"></i><b>5.5.3</b> Get the genomic ranges of the peaks from MACS and find which gene is closer to them (ie. Annotate them)</a></li>
<li class="chapter" data-level="5.5.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#change-the-gene-format."><i class="fa fa-check"></i><b>5.5.4</b> Change the gene format.</a></li>
<li class="chapter" data-level="5.5.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-and-export-the-table"><i class="fa fa-check"></i><b>5.5.5</b> Explore and export the table</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#annotate-with-chipseeker"><i class="fa fa-check"></i><b>5.6</b> Annotate with <code>ChIPseeker</code></a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-and-export-the-output"><i class="fa fa-check"></i><b>5.6.1</b> Explore and Export the output</a></li>
<li class="chapter" data-level="5.6.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#positive-controls."><i class="fa fa-check"></i><b>5.6.2</b> Positive controls.</a></li>
<li class="chapter" data-level="5.6.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#plots"><i class="fa fa-check"></i><b>5.6.3</b> Plots</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#sequence-motif-analysis"><i class="fa fa-check"></i><b>5.7</b> Sequence Motif analysis</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#put-the-peak-sequences-in-fasta-format"><i class="fa fa-check"></i><b>5.7.1</b> Put the peak sequences in fasta format</a></li>
<li class="chapter" data-level="5.7.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#find-the-motifs"><i class="fa fa-check"></i><b>5.7.2</b> Find the MOTIFS</a></li>
<li class="chapter" data-level="5.7.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-the-results"><i class="fa fa-check"></i><b>5.7.3</b> Explore the results</a></li>
<li class="chapter" data-level="5.7.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#plot-the-results"><i class="fa fa-check"></i><b>5.7.4</b> Plot the results</a></li>
<li class="chapter" data-level="5.7.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>5.7.5</b> Plotting with Ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#extra-how-to-solve-some-common-annotation-issues."><i class="fa fa-check"></i><b>5.8</b> Extra: How to solve some common annotation issues.</a></li>
<li class="chapter" data-level="5.9" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#extra-detailed-explanation-of-the-macs-output-files"><i class="fa fa-check"></i><b>5.9</b> Extra: Detailed explanation of the MACS output files</a></li>
<li class="chapter" data-level="5.10" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#activity"><i class="fa fa-check"></i><b>5.10</b> Activity</a></li>
<li class="chapter" data-level="5.11" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#resources-and-bibliography"><i class="fa fa-check"></i><b>5.11</b> Resources and Bibliography</a></li>
<li class="chapter" data-level="5.12" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#session-info-all-the-packages-installed."><i class="fa fa-check"></i><b>5.12</b> Session info: all the packages installed.</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html"><i class="fa fa-check"></i><b>6</b> RNA-seq analysis</a>
<ul>
<li class="chapter" data-level="6.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>6.1</b> Objectives</a></li>
<li class="chapter" data-level="6.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>6.2</b> Introduction</a></li>
<li class="chapter" data-level="6.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#differential-gene-expression"><i class="fa fa-check"></i><b>6.3</b> Differential gene expression</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#loading-the-tables"><i class="fa fa-check"></i><b>6.3.1</b> Loading the tables</a></li>
<li class="chapter" data-level="6.3.2" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#create-the-deseq2-count-matrix"><i class="fa fa-check"></i><b>6.3.2</b> Create the DeSeq2 count matrix</a></li>
<li class="chapter" data-level="6.3.3" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#meta-data-coldata-preparation"><i class="fa fa-check"></i><b>6.3.3</b> Meta-data (colData) preparation</a></li>
<li class="chapter" data-level="6.3.4" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#quality-filtering-distance-matrix-and-principal-component-analysis-pca"><i class="fa fa-check"></i><b>6.3.4</b> Quality filtering: Distance matrix and principal component analysis (PCA)</a></li>
<li class="chapter" data-level="6.3.5" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#differential-gene-expression-analysis-between-zt."><i class="fa fa-check"></i><b>6.3.5</b> Differential gene expression analysis between ZT.</a></li>
<li class="chapter" data-level="6.3.6" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#explore-results"><i class="fa fa-check"></i><b>6.3.6</b> Explore results</a></li>
<li class="chapter" data-level="6.3.7" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#plot-counts-using-ggplot"><i class="fa fa-check"></i><b>6.3.7</b> Plot counts using ggplot</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#gene-ontology-go-term-analysis"><i class="fa fa-check"></i><b>6.4</b> Gene Ontology (GO) term analysis</a></li>
<li class="chapter" data-level="6.5" data-path="rna-seq-analysis.html"><a href="rna-seq-analysis.html#extra-negative-binomial-model"><i class="fa fa-check"></i><b>6.5</b> Extra: negative binomial model</a></li>
<li class="chapter" data-level="6.6" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#activity"><i class="fa fa-check"></i><b>6.6</b> Activity</a></li>
<li class="chapter" data-level="6.7" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#resources-and-bibliography"><i class="fa fa-check"></i><b>6.7</b> Resources and Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="circadian-analysis.html"><a href="circadian-analysis.html"><i class="fa fa-check"></i><b>7</b> Circadian analysis</a>
<ul>
<li class="chapter" data-level="7.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#objectives-1"><i class="fa fa-check"></i><b>7.1</b> Objectives</a></li>
<li class="chapter" data-level="7.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#introduction-1"><i class="fa fa-check"></i><b>7.2</b> Introduction</a></li>
<li class="chapter" data-level="7.3" data-path="circadian-analysis.html"><a href="circadian-analysis.html#cycling-genes-analysis"><i class="fa fa-check"></i><b>7.3</b> Cycling genes analysis</a>
<ul>
<li class="chapter" data-level="7.3.1" data-path="circadian-analysis.html"><a href="circadian-analysis.html#prepare-the-tables"><i class="fa fa-check"></i><b>7.3.1</b> Prepare the tables</a></li>
<li class="chapter" data-level="7.3.2" data-path="circadian-analysis.html"><a href="circadian-analysis.html#meta2d"><i class="fa fa-check"></i><b>7.3.2</b> Meta2d</a></li>
<li class="chapter" data-level="7.3.3" data-path="circadian-analysis.html"><a href="circadian-analysis.html#plot-the-data."><i class="fa fa-check"></i><b>7.3.3</b> Plot the data.</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="circadian-analysis.html"><a href="circadian-analysis.html#cycling-genes-analysis-normalizing-by-maximum"><i class="fa fa-check"></i><b>7.4</b> Cycling genes analysis normalizing by maximum</a></li>
<li class="chapter" data-level="7.5" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#activity-1"><i class="fa fa-check"></i><b>7.5</b> Activity</a></li>
<li class="chapter" data-level="7.6" data-path="circadian-analysis.html"><a href="circadian-analysis.html#resources-and-bibliography-1"><i class="fa fa-check"></i><b>7.6</b> Resources and Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html"><i class="fa fa-check"></i><b>8</b> Single cell RNA-seq</a>
<ul>
<li class="chapter" data-level="8.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>8.1</b> Objectives</a></li>
<li class="chapter" data-level="8.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>8.2</b> Introduction</a></li>
<li class="chapter" data-level="8.3" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#packages-used"><i class="fa fa-check"></i><b>8.3</b> Packages used</a></li>
<li class="chapter" data-level="8.4" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#initialize-the-data"><i class="fa fa-check"></i><b>8.4</b> Initialize the data</a></li>
<li class="chapter" data-level="8.5" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#quality-filtering"><i class="fa fa-check"></i><b>8.5</b> Quality filtering</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#get-number-of-genes-captured-and-number-of-umi-remember-the-difference-between-reads-and-umi"><i class="fa fa-check"></i><b>8.5.1</b> Get number of genes captured and number of UMI (remember the difference between reads and UMI)</a></li>
<li class="chapter" data-level="8.5.2" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#calculate-percentage-of-mitochondrial-genes"><i class="fa fa-check"></i><b>8.5.2</b> Calculate percentage of mitochondrial genes:</a></li>
<li class="chapter" data-level="8.5.3" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#removing-low-quality-cells"><i class="fa fa-check"></i><b>8.5.3</b> Removing low quality cells</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#normalizing-the-data"><i class="fa fa-check"></i><b>8.6</b> Normalizing the data</a></li>
<li class="chapter" data-level="8.7" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>8.7</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="8.8" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#integrating-all-the-samples."><i class="fa fa-check"></i><b>8.8</b> Integrating ALL the samples.</a></li>
<li class="chapter" data-level="8.9" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#scaling-the-data-and-removing-unwanted-sources-of-variation"><i class="fa fa-check"></i><b>8.9</b> Scaling the data and removing unwanted sources of variation</a></li>
<li class="chapter" data-level="8.10" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#perform-linear-dimensional-reduction"><i class="fa fa-check"></i><b>8.10</b> Perform linear dimensional reduction</a></li>
<li class="chapter" data-level="8.11" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#clustering"><i class="fa fa-check"></i><b>8.11</b> Clustering</a></li>
<li class="chapter" data-level="8.12" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#visualization"><i class="fa fa-check"></i><b>8.12</b> Visualization</a></li>
<li class="chapter" data-level="8.13" data-path="single-cell-rna-seq.html"><a href="single-cell-rna-seq.html#finding-differentially-expressed-features-cluster-biomarkers"><i class="fa fa-check"></i><b>8.13</b> Finding differentially expressed features (cluster biomarkers)</a></li>
</ul></li>
<li class="appendix"><span><b>Appendices</b></span></li>
<li class="divider"></li>
<li><a rel="license" href="https://creativecommons.org/licenses/by-sa/4.0/" 
    target="blank"><img alt="Creative Commons License" 
    style="border-width:0" 
    src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><li>
<li><a href="https://psyteachr.github.io/" target="blank">PsyTeachR</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Genomics with R for biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="rna-seq-analysis" class="section level1" number="6">
<h1><span class="header-section-number">Lab 6</span> RNA-seq analysis</h1>
<div id="objectives" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Objectives</h2>
<p>After this section you should be able to:</p>
<ol style="list-style-type: decimal">
<li>Understand RNA sequencing counts.</li>
<li>Perform distance matrix and PCA.</li>
<li>Do differential gene expression using DeSeq2.</li>
</ol>
</div>
<div id="introduction" class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Introduction</h2>
<p>We will re-analyze some data from <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006613">Abruzzi et al, 2017</a>. These are data that explore gene expression in different neuronal cell types in a circadian layout.</p>
<p>This time we will just take the process data from the paper, and no directly the Raw data like in the previous chapter (Chip-seq). The data was downloaded from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE77451">GEO</a>. The only pre-processing to the data was done to change column names.</p>
<p>As you can see in the methods section of the paper they have different cell-types and timepoints. Let’s try to understand what type of system and data we have in hands.</p>
<p><strong>Circadian behavior</strong></p>
<p>Most living organisms change their behavior and metabolism over the day. Some animals like us have more activity over the day while others like mice have more activity over night. Flies in particular have a peak of activity the first hours of the morning and the last hours of the afternoon with a “siesta” (nap in Spanish) at the middle of the day.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-2"></span>
<img src="RNAseq/circadianact.png" alt="Fly activity over the day. Adapted from Nicholas R. J Glossop et al 2011. To meassure fly activity each individual fly is monitored either by counting when they cross a red-light bim (regular activity monitors) of by software tracking (flyboxes)." width="500px" />
<p class="caption">
Figure 4.1: Fly activity over the day. Adapted from Nicholas R. J Glossop et al 2011. To meassure fly activity each individual fly is monitored either by counting when they cross a red-light bim (regular activity monitors) of by software tracking (flyboxes).
</p>
</div>
<p>These changes in behavior are govern by changes in gene expression in particular cell-types in the brain called the pace-makers. If we look at gene expression in these cells over the day we will notice that some genes oscillate. Like <strong>Clk</strong> and <strong>tim</strong>, the proteins we analyzed in the previous Lab.</p>
<p>In the circadian field time is reported as “hours after the light is ON”. This is zeitgeber time (ZT) and it is useful to have a unified measurement of time. Most animals in labs are kept in 12 hours of light followed by 12 hours of dark. Therefore, ZT0 is the “sunrise” or lights on and ZT12 is lights off in a light/dark (LD)12:12 cycle time.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-3"></span>
<img src="RNAseq/cycwithzt.png" alt="Fly activity over the day with ZT scale. Figure adapted from Dubowy et al 2017" width="500px" />
<p class="caption">
Figure 4.2: Fly activity over the day with ZT scale. Figure adapted from Dubowy et al 2017
</p>
</div>
<p>As stated before, circadian behaviors is generated by a small subset of neurons.</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-4"></span>
<img src="RNAseq/circadiancells.png" alt="Schematic representation of the circadian neural network. Four small ventrolateral neurons (s‐LNvs, red), the 5th s‐LNv (dark violet), four large ventrolateral neurons (l‐LNvs, brown), six dorsolateral neurons (LNds, orange), three lateral posterior neurons (LPN, green), and ca. 60 neurons per hemisphere in three dorsal groups (DN1–3, lilac, cyan, blue, respectively). Adapted from Schubert et al 2018" width="500px" />
<p class="caption">
Figure 4.3: Schematic representation of the circadian neural network. Four small ventrolateral neurons (s‐LNvs, red), the 5th s‐LNv (dark violet), four large ventrolateral neurons (l‐LNvs, brown), six dorsolateral neurons (LNds, orange), three lateral posterior neurons (LPN, green), and ca. 60 neurons per hemisphere in three dorsal groups (DN1–3, lilac, cyan, blue, respectively). Adapted from Schubert et al 2018
</p>
</div>
<p>In the data analyzed in this lab, the authors collected RNA samples over 6 times of the day (ZT3, 7, 11, 15, 19, 23) from 4 cell-types: LNv, LNd, DN1 and Dopaminergic neurons (TH cells). This last group of cells are not part of the circadian core clock but in this study they see oscillating genes!</p>
<p><strong>Packages:</strong></p>
<p>We will use the following packages. Do not forget to executing them every time you open a new R session. Before running them for the first time, you will have to install them (go back to Lab1 if you have any doubt).</p>
<div class="sourceCode" id="cb262"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb262-1"><a href="rna-seq-analysis.html#cb262-1"></a><span class="kw">library</span>(DESeq2) <span class="co">#for differential expression analysis</span></span>
<span id="cb262-2"><a href="rna-seq-analysis.html#cb262-2"></a><span class="kw">library</span>(ggplot2) <span class="co">#for 2D graph</span></span>
<span id="cb262-3"><a href="rna-seq-analysis.html#cb262-3"></a><span class="kw">library</span>(ggrepel) <span class="co">#to get the names in ggplot graph </span></span>
<span id="cb262-4"><a href="rna-seq-analysis.html#cb262-4"></a><span class="kw">library</span>(gridExtra) <span class="co">#to arrange the plots</span></span>
<span id="cb262-5"><a href="rna-seq-analysis.html#cb262-5"></a><span class="kw">library</span>(factoextra) <span class="co">#extra plots</span></span>
<span id="cb262-6"><a href="rna-seq-analysis.html#cb262-6"></a><span class="kw">library</span>(plotly) <span class="co">#for 3D graphics</span></span>
<span id="cb262-7"><a href="rna-seq-analysis.html#cb262-7"></a><span class="kw">library</span>(plyr) <span class="co">#table manipulation</span></span>
<span id="cb262-8"><a href="rna-seq-analysis.html#cb262-8"></a><span class="kw">library</span>(dplyr)<span class="co">#table manipulation</span></span>
<span id="cb262-9"><a href="rna-seq-analysis.html#cb262-9"></a><span class="kw">library</span>(tidyr)<span class="co">#table manipulation</span></span>
<span id="cb262-10"><a href="rna-seq-analysis.html#cb262-10"></a><span class="kw">library</span>(<span class="st">&quot;RColorBrewer&quot;</span>) <span class="co">#extra color palettes</span></span>
<span id="cb262-11"><a href="rna-seq-analysis.html#cb262-11"></a><span class="kw">library</span>(<span class="st">&quot;pheatmap&quot;</span>) <span class="co">#nice heatmaps</span></span>
<span id="cb262-12"><a href="rna-seq-analysis.html#cb262-12"></a><span class="kw">library</span>(org.Dm.eg.db) <span class="co">#get annotation</span></span>
<span id="cb262-13"><a href="rna-seq-analysis.html#cb262-13"></a><span class="kw">library</span>(MetaCycle) <span class="co"># to identify cycling genes, the successor of JTK according to JTK authors</span></span>
<span id="cb262-14"><a href="rna-seq-analysis.html#cb262-14"></a><span class="kw">library</span>(pcaExplorer) <span class="co">#more PCA analysis</span></span></code></pre></div>
</div>
<div id="differential-gene-expression" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Differential gene expression</h2>
<p>Gene expression data is made of integer numbers by genes. Each number represents exactly how many reads aligned to the DNA sequence that is assigned to each gene.</p>
<table style="width:100%;">
<caption>Gene counts table</caption>
<colgroup>
<col width="9%" />
<col width="21%" />
<col width="23%" />
<col width="21%" />
<col width="23%" />
</colgroup>
<thead>
<tr class="header">
<th align="right">Gene</th>
<th align="right">Condition 1 replicate 1</th>
<th align="right">Condition 1 replicate 2</th>
<th align="right">Condition 2 replicate 1</th>
<th align="right">Condition 2 replicate 2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">A</td>
<td align="right">20</td>
<td align="right">400</td>
<td align="right">60</td>
<td align="right">40</td>
</tr>
<tr class="even">
<td align="right">B</td>
<td align="right">2</td>
<td align="right">50</td>
<td align="right">2000</td>
<td align="right">2500</td>
</tr>
<tr class="odd">
<td align="right">C</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">D</td>
<td align="right">5</td>
<td align="right">2050</td>
<td align="right">150</td>
<td align="right">144</td>
</tr>
</tbody>
</table>
<p>Ultimately, what we want to do is to compare gene expression between conditions. To do so we have to make some considerations. One is the fact that we want to make conclusions about the whole transcriptome of the organism we are working with. This means around 12000 comparisons (assuming 12000 genes). This implies a huge multiple-comparison correction. Another important consideration is the difference in sequencing depth, to overcame this issue the reads are normalized using an estimate that includes the total amount of reads in that sample. Finally, genes that are too lowly expressed or have a huge dispersion have to be removed from the analysis as they are too noisy to get a trustable conclusion from them.</p>
<p>To perform differential gene expression, we have to build a model that explains the differences in gene expression using the condition we are testing, fit the data to that model and see if it can explain the differences. This is done using generalized linear models (GLM) modeled using a negative binomial distribution (NB). Why is this? Because the data are counts, ie. integers (there are not 12.5 reads, you either have 12 or 13), to model discrete counts, we use either poison or NB distribution. NB is used because it allows for expression and variance to be unlinked.</p>
<p>Luckily for us, there are many packages dedicated to solve this problem. We will use <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">DeSeq2</a></p>
<p>The steps performed by the DESeq function are documented in the manual page of DESeq2; briefly, they are:</p>
<p>1 estimation of size factors by estimateSizeFactors (this is for normalization)
2 estimation of dispersion by estimateDispersions (calculate the gene expression dispersion)
3 negative binomial GLM fitting and Wald statistics by nbinomWaldTest (this is the actual differential gene expression step)</p>
<p>The only requirement for Deseq2 is to have the tables of data in a proper format and a meta-data object that indicates what is each sample.</p>
<p>Imagine we have a wildtype and a mutant, DeSeq2 requires
Table: <strong>1: Data matrix with Gene name as ROWNAMES</strong></p>
<table>
<caption><strong>2: MetaData (colData) in which ROWNAMES are the COLNAMES of the Data matrix and the conditions are column names</strong></caption>
<thead>
<tr class="header">
<th align="right">Row Names</th>
<th align="right">WildType.1</th>
<th align="right">WildType.2</th>
<th align="right">Mutant.1</th>
<th align="right">Mutant.2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">geneA</td>
<td align="right">20</td>
<td align="right">400</td>
<td align="right">60</td>
<td align="right">40</td>
</tr>
<tr class="even">
<td align="right">geneB</td>
<td align="right">2</td>
<td align="right">50</td>
<td align="right">2000</td>
<td align="right">2500</td>
</tr>
<tr class="odd">
<td align="right">geneC</td>
<td align="right">0</td>
<td align="right">5</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="right">geneD</td>
<td align="right">5</td>
<td align="right">2050</td>
<td align="right">150</td>
<td align="right">144</td>
</tr>
</tbody>
</table>
<table>
<thead>
<tr class="header">
<th align="right">Row Names</th>
<th align="right">Genotype</th>
<th align="right">replicate</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">WildType.1</td>
<td align="right">WT</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="right">WildType.2</td>
<td align="right">WT</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="right">Mutant.1</td>
<td align="right">actinKD</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="right">Mutant.2</td>
<td align="right">actinKD</td>
<td align="right">2</td>
</tr>
</tbody>
</table>
<p>The we can use the Genotype as a condition to compare gene expression for example.</p>
<p>At this time, we will start by comparing gene expression at two timepoints in one cell-type. This is good to have a first impression of the data but it is not the way to go if your idea is to analyze circadian cycling behavior.</p>
<div id="loading-the-tables" class="section level3" number="6.3.1">
<h3><span class="header-section-number">6.3.1</span> Loading the tables</h3>
<p>This time we have 4 gene expression tables: one for each cell type we are analyzing. Therefore, some of the processing will be done using for loops. This might be confusing at the beginning but is a really useful tool.</p>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="rna-seq-analysis.html#cb263-1"></a><span class="co">#We can list the files that are present in our working directory, or any path we specify. Remember the path will change depending where you have your data.</span></span>
<span id="cb263-2"><a href="rna-seq-analysis.html#cb263-2"></a><span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;../RNAseq/&quot;</span>)</span>
<span id="cb263-3"><a href="rna-seq-analysis.html#cb263-3"></a><span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;../RNAseq/&quot;</span>,<span class="dt">pattern =</span> <span class="st">&quot;expression&quot;</span>) <span class="co">#what do you think patter is doing? remember to go to the help page if you need to.</span></span>
<span id="cb263-4"><a href="rna-seq-analysis.html#cb263-4"></a><span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;../RNAseq/&quot;</span>,<span class="dt">pattern =</span> <span class="st">&quot;expression&quot;</span>,<span class="dt">full.names =</span> T) <span class="co">#what happened now?</span></span>
<span id="cb263-5"><a href="rna-seq-analysis.html#cb263-5"></a></span>
<span id="cb263-6"><a href="rna-seq-analysis.html#cb263-6"></a>paper_tables&lt;-<span class="kw">list.files</span>(<span class="dt">path =</span> <span class="st">&quot;../RNAseq/&quot;</span>,<span class="dt">pattern =</span> <span class="st">&quot;xpression&quot;</span>,<span class="dt">full.names =</span> T) <span class="co">#what we are doing here?, check your environment, do you see something new?</span></span>
<span id="cb263-7"><a href="rna-seq-analysis.html#cb263-7"></a></span>
<span id="cb263-8"><a href="rna-seq-analysis.html#cb263-8"></a><span class="co">#If we try with one</span></span>
<span id="cb263-9"><a href="rna-seq-analysis.html#cb263-9"></a>TH_gene_expression&lt;-<span class="kw">read.table</span>(<span class="st">&quot;../RNAseq/TH_gene_expression.txt&quot;</span>,<span class="dt">header =</span> T)</span></code></pre></div>
<pre><code>##  [1] &quot;all_tables.rdata&quot;        &quot;circadianact.png&quot;       
##  [3] &quot;circadianbrain.png&quot;      &quot;circadiancells.png&quot;     
##  [5] &quot;cycDN1_reults.txt&quot;       &quot;cycDN1.csv&quot;             
##  [7] &quot;cycwithzt.png&quot;           &quot;DN1_gene_expression.txt&quot;
##  [9] &quot;F1.large.jpg&quot;            &quot;howcircadian.png&quot;       
## [11] &quot;LNd_gene_expression.txt&quot; &quot;LNv_gene_expression.txt&quot;
## [13] &quot;RNASeq_DeSeq2.html&quot;      &quot;RNASeq_DeSeq2.rmd&quot;      
## [15] &quot;RNAseqDataPrep.html&quot;     &quot;TableMod.rmd&quot;           
## [17] &quot;TH_gene_expression.txt&quot; 
## [1] &quot;DN1_gene_expression.txt&quot; &quot;LNd_gene_expression.txt&quot;
## [3] &quot;LNv_gene_expression.txt&quot; &quot;TH_gene_expression.txt&quot; 
## [1] &quot;../RNAseq//DN1_gene_expression.txt&quot; &quot;../RNAseq//LNd_gene_expression.txt&quot;
## [3] &quot;../RNAseq//LNv_gene_expression.txt&quot; &quot;../RNAseq//TH_gene_expression.txt&quot;</code></pre>
<p>How can we make it work for all the tables? We want to change the name, we basically want to remove the “.txt” ending. That can be done using gsub. We want to remove “.txt”, so we replace it for an empty string.</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="rna-seq-analysis.html#cb265-1"></a><span class="kw">gsub</span>(<span class="dt">x =</span> <span class="st">&quot;../RNAseq/TH_gene_expression.txt&quot;</span>,<span class="dt">pattern =</span><span class="st">&quot;.txt&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;../RNAseq/TH_gene_expression&quot;</code></pre>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="rna-seq-analysis.html#cb267-1"></a><span class="co">#We also need to remove the path file</span></span>
<span id="cb267-2"><a href="rna-seq-analysis.html#cb267-2"></a><span class="kw">gsub</span>(<span class="dt">x =</span> <span class="st">&quot;../RNAseq/TH_gene_expression.txt&quot;</span>,<span class="dt">pattern =</span><span class="st">&quot;../RNAseq/&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;TH_gene_expression.txt&quot;</code></pre>
<p>Now we can try to do it for all of them, just trying to see if changing the names is working but NOT applying it to anything until we know it works.</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="rna-seq-analysis.html#cb269-1"></a><span class="cf">for</span> (e <span class="cf">in</span> (paper_tables)){</span>
<span id="cb269-2"><a href="rna-seq-analysis.html#cb269-2"></a>  <span class="kw">print</span>(<span class="st">&quot;original name:          &quot;</span>)</span>
<span id="cb269-3"><a href="rna-seq-analysis.html#cb269-3"></a>  <span class="kw">print</span>(e) <span class="co"># this is printing e</span></span>
<span id="cb269-4"><a href="rna-seq-analysis.html#cb269-4"></a>  n&lt;-<span class="kw">gsub</span>(<span class="dt">pattern =</span><span class="st">&quot;.txt&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>,<span class="dt">x =</span> e)</span>
<span id="cb269-5"><a href="rna-seq-analysis.html#cb269-5"></a>  n&lt;-<span class="kw">gsub</span>(<span class="dt">pattern =</span><span class="st">&quot;../RNAseq//&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>,<span class="dt">x =</span> n)</span>
<span id="cb269-6"><a href="rna-seq-analysis.html#cb269-6"></a>  <span class="kw">print</span>(<span class="st">&quot;final name:        &quot;</span>)</span>
<span id="cb269-7"><a href="rna-seq-analysis.html#cb269-7"></a>  <span class="kw">print</span>(n)</span>
<span id="cb269-8"><a href="rna-seq-analysis.html#cb269-8"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;original name:          &quot;
## [1] &quot;../RNAseq//DN1_gene_expression.txt&quot;
## [1] &quot;final name:        &quot;
## [1] &quot;DN1_gene_expression&quot;
## [1] &quot;original name:          &quot;
## [1] &quot;../RNAseq//LNd_gene_expression.txt&quot;
## [1] &quot;final name:        &quot;
## [1] &quot;LNd_gene_expression&quot;
## [1] &quot;original name:          &quot;
## [1] &quot;../RNAseq//LNv_gene_expression.txt&quot;
## [1] &quot;final name:        &quot;
## [1] &quot;LNv_gene_expression&quot;
## [1] &quot;original name:          &quot;
## [1] &quot;../RNAseq//TH_gene_expression.txt&quot;
## [1] &quot;final name:        &quot;
## [1] &quot;TH_gene_expression&quot;</code></pre>
<blockquote>
<p>What is this for loop doing? Do you think it is working? Why?</p>
</blockquote>
<p>Now we will read the tables and put them in the names we created. To do this we will use the function assign, that actually assign anything to a given name. The name is stored in “n”. What we want to assign is the “read.delim” of the file name stored in e.</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="rna-seq-analysis.html#cb271-1"></a><span class="cf">for</span> (e <span class="cf">in</span> (paper_tables)){</span>
<span id="cb271-2"><a href="rna-seq-analysis.html#cb271-2"></a>  n&lt;-<span class="kw">gsub</span>(<span class="dt">pattern =</span><span class="st">&quot;.txt&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>,<span class="dt">x =</span> e)</span>
<span id="cb271-3"><a href="rna-seq-analysis.html#cb271-3"></a>  n&lt;-<span class="kw">gsub</span>(<span class="dt">pattern =</span><span class="st">&quot;../RNAseq//&quot;</span>,  <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>,<span class="dt">x =</span> n)</span>
<span id="cb271-4"><a href="rna-seq-analysis.html#cb271-4"></a>  <span class="kw">print</span>(<span class="kw">paste0</span>(<span class="st">&quot;I am reading: &quot;</span>,n))</span>
<span id="cb271-5"><a href="rna-seq-analysis.html#cb271-5"></a>  <span class="kw">assign</span>(n,<span class="kw">read.delim</span>(e))</span>
<span id="cb271-6"><a href="rna-seq-analysis.html#cb271-6"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;I am reading: DN1_gene_expression&quot;
## [1] &quot;I am reading: LNd_gene_expression&quot;
## [1] &quot;I am reading: LNv_gene_expression&quot;
## [1] &quot;I am reading: TH_gene_expression&quot;</code></pre>
</div>
<div id="create-the-deseq2-count-matrix" class="section level3" number="6.3.2">
<h3><span class="header-section-number">6.3.2</span> Create the DeSeq2 count matrix</h3>
<p>DeSeq2 expects to have a data.frame with the gene-names as <code>row.names</code> and then just the counts.</p>
<p>I will first show it for one of the tables and then do a for loop to do it for all of them.</p>
<div class="sourceCode" id="cb273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="rna-seq-analysis.html#cb273-1"></a><span class="co">#We do not want to overwrite the raw tables. So, we will create new objects with the same data and we will modify these new objects and preserve the original data. Is like doing &quot;save as&quot; in word.</span></span>
<span id="cb273-2"><a href="rna-seq-analysis.html#cb273-2"></a></span>
<span id="cb273-3"><a href="rna-seq-analysis.html#cb273-3"></a>ds2_DN1_gene_expression_try&lt;-DN1_gene_expression <span class="co">#what is this doing?</span></span>
<span id="cb273-4"><a href="rna-seq-analysis.html#cb273-4"></a></span>
<span id="cb273-5"><a href="rna-seq-analysis.html#cb273-5"></a><span class="co">#We do now the loop for all of them to create the `ds2_` objects</span></span>
<span id="cb273-6"><a href="rna-seq-analysis.html#cb273-6"></a><span class="kw">ls</span>(<span class="dt">pattern =</span> <span class="st">&quot;gene_expression&quot;</span>) <span class="co">#this list the objects in the environment that has the selected pattern in its name</span></span></code></pre></div>
<pre><code>## [1] &quot;DN1_gene_expression&quot;         &quot;ds2_DN1_gene_expression_try&quot;
## [3] &quot;LNd_gene_expression&quot;         &quot;LNv_gene_expression&quot;        
## [5] &quot;TH_gene_expression&quot;</code></pre>
<div class="sourceCode" id="cb275"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb275-1"><a href="rna-seq-analysis.html#cb275-1"></a><span class="co">#The new name we want is just adding &quot;ds2_&quot; to the beginning of the object name. We can do this with the paste0 function.</span></span>
<span id="cb275-2"><a href="rna-seq-analysis.html#cb275-2"></a><span class="kw">paste0</span>(<span class="st">&quot;ds2_&quot;</span>,<span class="st">&quot;DN1_gene_expression&quot;</span>)</span>
<span id="cb275-3"><a href="rna-seq-analysis.html#cb275-3"></a></span>
<span id="cb275-4"><a href="rna-seq-analysis.html#cb275-4"></a><span class="cf">for</span> (e <span class="cf">in</span> <span class="kw">ls</span>(<span class="dt">pattern =</span> <span class="st">&quot;gene_expression&quot;</span>)){ </span>
<span id="cb275-5"><a href="rna-seq-analysis.html#cb275-5"></a>  x =<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;ds2&quot;</span>, <span class="dt">sep=</span><span class="st">&quot;_&quot;</span>,e) <span class="co">#this is the new name</span></span>
<span id="cb275-6"><a href="rna-seq-analysis.html#cb275-6"></a>  <span class="kw">print</span>(<span class="st">&quot;new name is:&quot;</span>)</span>
<span id="cb275-7"><a href="rna-seq-analysis.html#cb275-7"></a>  <span class="kw">print</span>(x) </span>
<span id="cb275-8"><a href="rna-seq-analysis.html#cb275-8"></a>  <span class="kw">assign</span>(x, <span class="kw">get</span>(e)) <span class="co">#get is a function that extracts the data from an object name</span></span>
<span id="cb275-9"><a href="rna-seq-analysis.html#cb275-9"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;ds2_DN1_gene_expression&quot;
## [1] &quot;new name is:&quot;
## [1] &quot;ds2_DN1_gene_expression&quot;
## [1] &quot;new name is:&quot;
## [1] &quot;ds2_ds2_DN1_gene_expression_try&quot;
## [1] &quot;new name is:&quot;
## [1] &quot;ds2_LNd_gene_expression&quot;
## [1] &quot;new name is:&quot;
## [1] &quot;ds2_LNv_gene_expression&quot;
## [1] &quot;new name is:&quot;
## [1] &quot;ds2_TH_gene_expression&quot;</code></pre>
<p>Now we have to do some manipulation in the tables for them to work with Deseq2.</p>
<div class="sourceCode" id="cb277"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb277-1"><a href="rna-seq-analysis.html#cb277-1"></a>ds2_DN1_gene_expression_try &lt;-<span class="st"> </span>ds2_DN1_gene_expression</span>
<span id="cb277-2"><a href="rna-seq-analysis.html#cb277-2"></a><span class="kw">row.names</span>(ds2_DN1_gene_expression_try) =<span class="st"> </span>ds2_DN1_gene_expression<span class="op">$</span>Symbol <span class="co"># we put the symbols as row.names</span></span>
<span id="cb277-3"><a href="rna-seq-analysis.html#cb277-3"></a>ds2_DN1_gene_expression_try&lt;-ds2_DN1_gene_expression_try[,<span class="op">-</span><span class="dv">1</span>] <span class="co"># we take the symbols column out</span></span>
<span id="cb277-4"><a href="rna-seq-analysis.html#cb277-4"></a><span class="kw">head</span>(ds2_DN1_gene_expression_try) <span class="co"># we can see that now the table is what we need</span></span>
<span id="cb277-5"><a href="rna-seq-analysis.html#cb277-5"></a><span class="kw">rm</span>(ds2_DN1_gene_expression_try) <span class="co">#remove it because you do not need it anymore and because it can create problems</span></span></code></pre></div>
<pre><code>##             DN1_ZT3_1 DN1_ZT7_1 DN1_ZT11_1 DN1_ZT15_1 DN1_ZT19_1 DN1_ZT23_1
## FBtr0070202         0         0          0          0          0          0
## FBtr0070207         0         0          0          0         14          0
## FBtr0070238         0         0          0          0         21          0
## FBtr0070251         0         0          0          0          0          0
## FBtr0070484         0         0          0          0          0          0
## FBtr0070489         0         0          0          0          0          0
##             DN1_ZT3_2 DN1_ZT7_2 DN1_ZT11_2 DN1_ZT15_2 DN1_ZT19_2 DN1_ZT23_2
## FBtr0070202         0         0          0          0          0          0
## FBtr0070207         0         0          0          0         16          7
## FBtr0070238         0         0          0          0         15          9
## FBtr0070251         0         0          1          0          2          0
## FBtr0070484         0         0          0          0          0          1
## FBtr0070489         0         0          0          0          0          0</code></pre>
<p>We can add this step to the for loop, or even better we can create a function and practice something new in R.</p>
<div class="sourceCode" id="cb279"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb279-1"><a href="rna-seq-analysis.html#cb279-1"></a><span class="co">#We create a small function to remember</span></span>
<span id="cb279-2"><a href="rna-seq-analysis.html#cb279-2"></a>sum2&lt;-<span class="cf">function</span>(x){</span>
<span id="cb279-3"><a href="rna-seq-analysis.html#cb279-3"></a> y=x<span class="op">+</span><span class="dv">2</span> </span>
<span id="cb279-4"><a href="rna-seq-analysis.html#cb279-4"></a> <span class="kw">return</span>(y)</span>
<span id="cb279-5"><a href="rna-seq-analysis.html#cb279-5"></a>}</span>
<span id="cb279-6"><a href="rna-seq-analysis.html#cb279-6"></a></span>
<span id="cb279-7"><a href="rna-seq-analysis.html#cb279-7"></a><span class="kw">sum2</span>(<span class="dv">4</span>) </span></code></pre></div>
<pre><code>## [1] 6</code></pre>
<blockquote>
<p>what is this doing? why?</p>
</blockquote>
<div class="sourceCode" id="cb281"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb281-1"><a href="rna-seq-analysis.html#cb281-1"></a><span class="co">#Now we create a function that takes the data frame and does all the transformation we did before</span></span>
<span id="cb281-2"><a href="rna-seq-analysis.html#cb281-2"></a>create_ds2=<span class="st"> </span><span class="cf">function</span>(df_name){</span>
<span id="cb281-3"><a href="rna-seq-analysis.html#cb281-3"></a>  df =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">get</span>(df_name))</span>
<span id="cb281-4"><a href="rna-seq-analysis.html#cb281-4"></a>  <span class="kw">row.names</span>(df) =<span class="st"> </span><span class="kw">as.character</span>(df[,<span class="dv">1</span>])</span>
<span id="cb281-5"><a href="rna-seq-analysis.html#cb281-5"></a>  df=df[,<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb281-6"><a href="rna-seq-analysis.html#cb281-6"></a>  <span class="kw">return</span>(df) <span class="co"># this actually tells the function to return the transformed data.frame</span></span>
<span id="cb281-7"><a href="rna-seq-analysis.html#cb281-7"></a>}</span>
<span id="cb281-8"><a href="rna-seq-analysis.html#cb281-8"></a></span>
<span id="cb281-9"><a href="rna-seq-analysis.html#cb281-9"></a><span class="co">#Let&#39;s try the function in one of the ds2 objects</span></span>
<span id="cb281-10"><a href="rna-seq-analysis.html#cb281-10"></a><span class="co">#head(create_ds2(ds2_LNd_gene_expression))[,c(1:3)]</span></span>
<span id="cb281-11"><a href="rna-seq-analysis.html#cb281-11"></a><span class="cf">for</span> (e <span class="cf">in</span> <span class="kw">ls</span>(<span class="dt">pattern =</span> <span class="st">&quot;ds2_&quot;</span>)){</span>
<span id="cb281-12"><a href="rna-seq-analysis.html#cb281-12"></a>  <span class="kw">print</span>(e)</span>
<span id="cb281-13"><a href="rna-seq-analysis.html#cb281-13"></a>  <span class="kw">assign</span>(e,<span class="kw">create_ds2</span>(e)) <span class="co">#this should actually run the function and apply it to the data frame</span></span>
<span id="cb281-14"><a href="rna-seq-analysis.html#cb281-14"></a>}</span></code></pre></div>
<pre><code>## [1] &quot;ds2_DN1_gene_expression&quot;
## [1] &quot;ds2_ds2_DN1_gene_expression_try&quot;
## [1] &quot;ds2_LNd_gene_expression&quot;
## [1] &quot;ds2_LNv_gene_expression&quot;
## [1] &quot;ds2_TH_gene_expression&quot;</code></pre>
</div>
<div id="meta-data-coldata-preparation" class="section level3 tabset tabset-fade tabset-pills" number="6.3.3">
<h3><span class="header-section-number">6.3.3</span> Meta-data (colData) preparation</h3>
<p>We now need to have a data.frame with the mapping between the columns and the ZT, the program call this colData. Is what is usually called “meta data”. Is a file that explains what each object/column is.</p>
<p>We will do it by hand just changing the names. No more loops for now. After you execute each step I recommend you to inspect the object to see what is happening.</p>
<p><strong>DN1</strong></p>
<div class="sourceCode" id="cb283"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb283-1"><a href="rna-seq-analysis.html#cb283-1"></a><span class="co">#takes the names in the data frame and put them in the new objects. We want to get the information for each </span></span>
<span id="cb283-2"><a href="rna-seq-analysis.html#cb283-2"></a>colData.DN1 =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">names</span>(ds2_DN1_gene_expression)) </span>
<span id="cb283-3"><a href="rna-seq-analysis.html#cb283-3"></a><span class="co">#now we put them as row.names</span></span>
<span id="cb283-4"><a href="rna-seq-analysis.html#cb283-4"></a><span class="kw">rownames</span>(colData.DN1) =<span class="st"> </span>colData.DN1[,<span class="dv">1</span>]</span>
<span id="cb283-5"><a href="rna-seq-analysis.html#cb283-5"></a><span class="co">#now we take the columns out </span></span>
<span id="cb283-6"><a href="rna-seq-analysis.html#cb283-6"></a>colData.DN1 =colData.DN1[,<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb283-7"><a href="rna-seq-analysis.html#cb283-7"></a><span class="co">#this is taking the last part of the rownames and put them in a column named `ZT`</span></span>
<span id="cb283-8"><a href="rna-seq-analysis.html#cb283-8"></a>colData.DN1<span class="op">$</span>ZT =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.DN1),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">2</span>)</span>
<span id="cb283-9"><a href="rna-seq-analysis.html#cb283-9"></a>colData.DN1<span class="op">$</span>rep =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.DN1),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">3</span>)</span>
<span id="cb283-10"><a href="rna-seq-analysis.html#cb283-10"></a></span>
<span id="cb283-11"><a href="rna-seq-analysis.html#cb283-11"></a><span class="kw">head</span>(colData.DN1)</span></code></pre></div>
<pre><code>##              ZT rep
## DN1_ZT3_1   ZT3   1
## DN1_ZT7_1   ZT7   1
## DN1_ZT11_1 ZT11   1
## DN1_ZT15_1 ZT15   1
## DN1_ZT19_1 ZT19   1
## DN1_ZT23_1 ZT23   1</code></pre>
<p><strong>LNv</strong></p>
<div class="sourceCode" id="cb285"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb285-1"><a href="rna-seq-analysis.html#cb285-1"></a>colData.LNv =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">names</span>(ds2_LNv_gene_expression)) </span>
<span id="cb285-2"><a href="rna-seq-analysis.html#cb285-2"></a><span class="co">#now we put them as row.names</span></span>
<span id="cb285-3"><a href="rna-seq-analysis.html#cb285-3"></a><span class="kw">rownames</span>(colData.LNv) =<span class="st"> </span>colData.LNv[,<span class="dv">1</span>]</span>
<span id="cb285-4"><a href="rna-seq-analysis.html#cb285-4"></a><span class="co">#now we take the columns out </span></span>
<span id="cb285-5"><a href="rna-seq-analysis.html#cb285-5"></a>colData.LNv =colData.LNv[,<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb285-6"><a href="rna-seq-analysis.html#cb285-6"></a><span class="co">#this is taking the last part of the rownames and put them in a column named `ZT`</span></span>
<span id="cb285-7"><a href="rna-seq-analysis.html#cb285-7"></a>colData.LNv<span class="op">$</span>ZT =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.LNv),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">2</span>)</span>
<span id="cb285-8"><a href="rna-seq-analysis.html#cb285-8"></a>colData.LNv<span class="op">$</span>rep =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.LNv),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">3</span>)</span></code></pre></div>
<p><strong>LNd</strong></p>
<div class="sourceCode" id="cb286"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb286-1"><a href="rna-seq-analysis.html#cb286-1"></a>colData.LNd =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">names</span>(ds2_LNd_gene_expression)) </span>
<span id="cb286-2"><a href="rna-seq-analysis.html#cb286-2"></a><span class="co">#now we put them as row.names</span></span>
<span id="cb286-3"><a href="rna-seq-analysis.html#cb286-3"></a><span class="kw">rownames</span>(colData.LNd) =<span class="st"> </span>colData.LNd[,<span class="dv">1</span>]</span>
<span id="cb286-4"><a href="rna-seq-analysis.html#cb286-4"></a><span class="co">#now we take the columns out </span></span>
<span id="cb286-5"><a href="rna-seq-analysis.html#cb286-5"></a>colData.LNd =colData.LNd[,<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb286-6"><a href="rna-seq-analysis.html#cb286-6"></a><span class="co">#this is taking the last part of the rownames and put them in a column named `ZT`</span></span>
<span id="cb286-7"><a href="rna-seq-analysis.html#cb286-7"></a>colData.LNd<span class="op">$</span>ZT =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.LNd),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">2</span>)</span>
<span id="cb286-8"><a href="rna-seq-analysis.html#cb286-8"></a>colData.LNd<span class="op">$</span>rep =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.LNd),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">3</span>)</span></code></pre></div>
<p><strong>TH</strong></p>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="rna-seq-analysis.html#cb287-1"></a>colData.TH =<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">names</span>(ds2_TH_gene_expression)) </span>
<span id="cb287-2"><a href="rna-seq-analysis.html#cb287-2"></a><span class="co">#now we put them as row.names</span></span>
<span id="cb287-3"><a href="rna-seq-analysis.html#cb287-3"></a><span class="kw">rownames</span>(colData.TH) =<span class="st"> </span>colData.TH[,<span class="dv">1</span>]</span>
<span id="cb287-4"><a href="rna-seq-analysis.html#cb287-4"></a><span class="co">#now we take the columns out and </span></span>
<span id="cb287-5"><a href="rna-seq-analysis.html#cb287-5"></a>colData.TH =colData.TH[,<span class="op">-</span><span class="dv">1</span>]</span>
<span id="cb287-6"><a href="rna-seq-analysis.html#cb287-6"></a>colData.TH<span class="op">$</span>ZT =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.TH),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">2</span>)</span>
<span id="cb287-7"><a href="rna-seq-analysis.html#cb287-7"></a>colData.TH<span class="op">$</span>rep =<span class="st"> </span><span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">rownames</span>(colData.TH),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">3</span>)</span></code></pre></div>
</div>
<div id="quality-filtering-distance-matrix-and-principal-component-analysis-pca" class="section level3 tabset tabset-fade tabset-pills" number="6.3.4">
<h3><span class="header-section-number">6.3.4</span> Quality filtering: Distance matrix and principal component analysis (PCA)</h3>
<p>If we want to evaluate the difference between two conditions we have to make sure that the difference between replicates of the same conditions are smaller. In other words, we expect the replicates of the same condition to be similar to each other. Checking this is a regular procedure in differential gene expression analysis as it helps us to identify samples that might be outliers.</p>
<p>To evaluate this, we will use two different approaches. Firstly, we will calculate a distance matrix, this is basically calculating the difference between each sample. To visualize this, we will use a heatmap. Secondly, we will do a principal component analysis (PCA) to get an idea on how similar are the samples between each other. To make it simple (but not completely correct) PCA is a way to reduce the dimensions (in this case many many genes) to a set of vectors (principal components, PC) that summarize the combination of the gene information. Then, if two samples have similar values in this PCs then they are similar to each other.</p>
<p>We first have to initiate the DeSeq2 object. This imply telling the package which is the matrix of read counts, the colData (meta data) and the design. The design is what the program uses to do the differential expression. Is simple in this case: only the ZT.</p>
<p><strong>Initialize the data</strong></p>
<div class="sourceCode" id="cb288"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb288-1"><a href="rna-seq-analysis.html#cb288-1"></a><span class="co">#This is actually the main function of the package and prepares everything to run the differential expression analysis.</span></span>
<span id="cb288-2"><a href="rna-seq-analysis.html#cb288-2"></a>dds.DN1 &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> <span class="kw">as.matrix</span>(ds2_DN1_gene_expression), <span class="dt">colData =</span> colData.DN1,<span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>ZT)</span>
<span id="cb288-3"><a href="rna-seq-analysis.html#cb288-3"></a></span>
<span id="cb288-4"><a href="rna-seq-analysis.html#cb288-4"></a><span class="co">#class of dds.DN1</span></span>
<span id="cb288-5"><a href="rna-seq-analysis.html#cb288-5"></a><span class="kw">class</span>(dds.DN1)</span></code></pre></div>
<pre><code>## [1] &quot;DESeqDataSet&quot;
## attr(,&quot;package&quot;)
## [1] &quot;DESeq2&quot;</code></pre>
<p><strong>Distance Matrix</strong></p>
<p>As stated before, a distance matrix is a way to quantify differences. In this case each column of the data matrix will be compared with all the others.
You can read more <a href="https://www.displayr.com/what-is-a-distance-matrix/">here</a> or <a href="http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/120-correspondence-analysis-theory-and-practice/">here</a>. <code>dist()</code> is the function that calculates the Euclidean distance, you can select which type of distance you want to calculate, read the help page for more information.</p>
<div class="sourceCode" id="cb290"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb290-1"><a href="rna-seq-analysis.html#cb290-1"></a><span class="co"># Calculate the distance matrix</span></span>
<span id="cb290-2"><a href="rna-seq-analysis.html#cb290-2"></a>sampleDists &lt;-<span class="st"> </span><span class="kw">dist</span>(<span class="kw">t</span>(<span class="kw">assay</span>(dds.DN1))) </span>
<span id="cb290-3"><a href="rna-seq-analysis.html#cb290-3"></a></span>
<span id="cb290-4"><a href="rna-seq-analysis.html#cb290-4"></a><span class="co"># Lets look at this step by step:</span></span>
<span id="cb290-5"><a href="rna-seq-analysis.html#cb290-5"></a><span class="co"># assay(dds.DN1) # extract the reads</span></span>
<span id="cb290-6"><a href="rna-seq-analysis.html#cb290-6"></a><span class="co"># t(assay(dds.DN1)) #transpose the matrix, this has to do with the fact that we want to calculate the distance between samples, if we do not transpose we will literally calculate the difference between genes across samples</span></span>
<span id="cb290-7"><a href="rna-seq-analysis.html#cb290-7"></a></span>
<span id="cb290-8"><a href="rna-seq-analysis.html#cb290-8"></a><span class="co">#Then we take this in a matrix and create the heatmap</span></span>
<span id="cb290-9"><a href="rna-seq-analysis.html#cb290-9"></a>sampleDistMatrix &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(sampleDists) <span class="co">#convert to matrix</span></span>
<span id="cb290-10"><a href="rna-seq-analysis.html#cb290-10"></a></span>
<span id="cb290-11"><a href="rna-seq-analysis.html#cb290-11"></a><span class="kw">rownames</span>(sampleDistMatrix) &lt;-<span class="st"> </span><span class="kw">paste</span>(dds.DN1<span class="op">$</span>ZT,dds.DN1<span class="op">$</span>rep,<span class="dt">sep=</span><span class="st">&quot;_rep_&quot;</span>) <span class="co"># put the rownames</span></span>
<span id="cb290-12"><a href="rna-seq-analysis.html#cb290-12"></a><span class="kw">colnames</span>(sampleDistMatrix) &lt;-<span class="st"> </span><span class="ot">NULL</span></span>
<span id="cb290-13"><a href="rna-seq-analysis.html#cb290-13"></a>colors &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>( <span class="kw">rev</span>(<span class="kw">brewer.pal</span>(<span class="dv">9</span>, <span class="st">&quot;Blues&quot;</span>)) )(<span class="dv">255</span>)<span class="co">#select color </span></span>
<span id="cb290-14"><a href="rna-seq-analysis.html#cb290-14"></a></span>
<span id="cb290-15"><a href="rna-seq-analysis.html#cb290-15"></a><span class="co">#Heatmap function</span></span>
<span id="cb290-16"><a href="rna-seq-analysis.html#cb290-16"></a><span class="kw">pheatmap</span>(sampleDistMatrix,</span>
<span id="cb290-17"><a href="rna-seq-analysis.html#cb290-17"></a>         <span class="dt">clustering_distance_rows=</span>sampleDists,</span>
<span id="cb290-18"><a href="rna-seq-analysis.html#cb290-18"></a>         <span class="dt">clustering_distance_cols=</span>sampleDists,</span>
<span id="cb290-19"><a href="rna-seq-analysis.html#cb290-19"></a>                  <span class="dt">annotation_row =</span> <span class="kw">rownames</span>(sampleDists),</span>
<span id="cb290-20"><a href="rna-seq-analysis.html#cb290-20"></a>         <span class="dt">annotation_col =</span> <span class="kw">rownames</span>(sampleDists),</span>
<span id="cb290-21"><a href="rna-seq-analysis.html#cb290-21"></a>         <span class="dt">cutree_rows =</span> <span class="dv">2</span>,</span>
<span id="cb290-22"><a href="rna-seq-analysis.html#cb290-22"></a>         <span class="dt">cutree_cols =</span> <span class="dv">2</span>,</span>
<span id="cb290-23"><a href="rna-seq-analysis.html#cb290-23"></a>         <span class="dt">col=</span>colors)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-21"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-21-1.png" alt="Heatmap representing distance between samples" width="100%" />
<p class="caption">
Figure 6.1: Heatmap representing distance between samples
</p>
</div>
<p><strong>PCA using DeSeq2</strong></p>
<p>We have to do a logarithmic transformation to see how the data looks like after minimizing differences between samples for rows with small counts, and which normalizes with respect to library size. You can read the help executing <code>?rlog</code>.</p>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="rna-seq-analysis.html#cb291-1"></a>rdl.DN1 &lt;-<span class="st"> </span><span class="kw">rlog</span>(dds.DN1) </span></code></pre></div>
<p>Now we can plot the PCA analysis already integrated in the DeSeq2 package.</p>
<div class="sourceCode" id="cb292"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb292-1"><a href="rna-seq-analysis.html#cb292-1"></a><span class="kw">plotPCA</span>(rdl.DN1, <span class="dt">intgroup=</span><span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-23"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-23-1.png" alt="PCA coloring by ZT" width="100%" />
<p class="caption">
Figure 6.2: PCA coloring by ZT
</p>
</div>
<p><strong>PCA using basic R</strong></p>
<p>We will use the function <code>prcomp()</code> to calculate the PCs.
But before we do that we will select the most variable genes across samples. If a gene has the same value across all the samples will not be explicative or informative.</p>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="rna-seq-analysis.html#cb293-1"></a>rv &lt;-<span class="st"> </span><span class="kw">rowVars</span>(<span class="kw">assay</span>(rdl.DN1)) <span class="co"># we calcualte the total variation in each gene (in this case rows), rowwVars literally calculates the variance in the rows</span></span>
<span id="cb293-2"><a href="rna-seq-analysis.html#cb293-2"></a>ntop &lt;-<span class="st"> </span><span class="dv">10000</span> <span class="co"># we select how many gene we want to use </span></span>
<span id="cb293-3"><a href="rna-seq-analysis.html#cb293-3"></a>sg &lt;-<span class="st"> </span><span class="kw">order</span>(rv, <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)[<span class="kw">seq_len</span>(<span class="kw">min</span>(ntop, <span class="kw">length</span>(rv)))] <span class="co"># here we select the most variable genes acording to the ntop value we set up before</span></span>
<span id="cb293-4"><a href="rna-seq-analysis.html#cb293-4"></a>mat &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">assay</span>(rdl.DN1)[sg, ]) <span class="co"># here we extract then the counts using the function assay from DeSeq2</span></span>
<span id="cb293-5"><a href="rna-seq-analysis.html#cb293-5"></a>pc &lt;-<span class="st"> </span><span class="kw">prcomp</span>(mat) <span class="co"># calcualte PCs</span></span>
<span id="cb293-6"><a href="rna-seq-analysis.html#cb293-6"></a><span class="co">#head(pc,n=3L)</span></span></code></pre></div>
<p>To select how many PCs are “important” we will use the percentage of variance explained by each PC. This takes us to the next step of understanding PCA. Each PC might be seen as a vector that explain some part of the variation of the data. Then if a PC explains only 2% of the data it makes no sense to pay attention to if we compared it to other PCs that might explain more.</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="rna-seq-analysis.html#cb294-1"></a><span class="kw">fviz_eig</span>(pc, <span class="dt">addlabels =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-25"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-25-1.png" alt="Variance explained by each PC" width="100%" />
<p class="caption">
Figure 5.3: Variance explained by each PC
</p>
</div>
<p>Now we can plot more than 2 PCs</p>
<div class="sourceCode" id="cb295"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb295-1"><a href="rna-seq-analysis.html#cb295-1"></a>pc_df=<span class="kw">as.data.frame</span>(pc<span class="op">$</span>x) <span class="co">#transform to data frame</span></span>
<span id="cb295-2"><a href="rna-seq-analysis.html#cb295-2"></a>pc_df=<span class="kw">merge</span>(pc_df,colData.DN1,<span class="dt">by=</span><span class="st">&quot;row.names&quot;</span>,<span class="dt">all=</span>T) <span class="co"># merge with the colData to have the meta data of each saple</span></span>
<span id="cb295-3"><a href="rna-seq-analysis.html#cb295-3"></a><span class="kw">rownames</span>(pc_df)=pc_df<span class="op">$</span>Row.names</span>
<span id="cb295-4"><a href="rna-seq-analysis.html#cb295-4"></a></span>
<span id="cb295-5"><a href="rna-seq-analysis.html#cb295-5"></a></span>
<span id="cb295-6"><a href="rna-seq-analysis.html#cb295-6"></a>p1=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT)) <span class="op">+</span></span>
<span id="cb295-7"><a href="rna-seq-analysis.html#cb295-7"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>)  <span class="op">+</span></span>
<span id="cb295-8"><a href="rna-seq-analysis.html#cb295-8"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb295-9"><a href="rna-seq-analysis.html#cb295-9"></a></span>
<span id="cb295-10"><a href="rna-seq-analysis.html#cb295-10"></a></span>
<span id="cb295-11"><a href="rna-seq-analysis.html#cb295-11"></a>p2=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC3, PC2, <span class="dt">color=</span>ZT)) <span class="op">+</span></span>
<span id="cb295-12"><a href="rna-seq-analysis.html#cb295-12"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb295-13"><a href="rna-seq-analysis.html#cb295-13"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb295-14"><a href="rna-seq-analysis.html#cb295-14"></a></span>
<span id="cb295-15"><a href="rna-seq-analysis.html#cb295-15"></a><span class="co">#and both together</span></span>
<span id="cb295-16"><a href="rna-seq-analysis.html#cb295-16"></a><span class="kw">grid.arrange</span>(p1,p2, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-26"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-26-1.png" alt="Ploting many principal components" width="100%" />
<p class="caption">
Figure 5.4: Ploting many principal components
</p>
</div>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="rna-seq-analysis.html#cb296-1"></a>p1=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC1, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>Row.names)) <span class="op">+</span></span>
<span id="cb296-2"><a href="rna-seq-analysis.html#cb296-2"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb296-3"><a href="rna-seq-analysis.html#cb296-3"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Row.names),</span>
<span id="cb296-4"><a href="rna-seq-analysis.html#cb296-4"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb296-5"><a href="rna-seq-analysis.html#cb296-5"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb296-6"><a href="rna-seq-analysis.html#cb296-6"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb296-7"><a href="rna-seq-analysis.html#cb296-7"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb296-8"><a href="rna-seq-analysis.html#cb296-8"></a></span>
<span id="cb296-9"><a href="rna-seq-analysis.html#cb296-9"></a></span>
<span id="cb296-10"><a href="rna-seq-analysis.html#cb296-10"></a>p2=<span class="kw">ggplot</span>(pc_df, <span class="kw">aes</span>(PC3, PC2, <span class="dt">color=</span>ZT,<span class="dt">label=</span>Row.names)) <span class="op">+</span></span>
<span id="cb296-11"><a href="rna-seq-analysis.html#cb296-11"></a><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size=</span><span class="dv">3</span>) <span class="op">+</span></span>
<span id="cb296-12"><a href="rna-seq-analysis.html#cb296-12"></a><span class="st">  </span><span class="kw">geom_label_repel</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Row.names),</span>
<span id="cb296-13"><a href="rna-seq-analysis.html#cb296-13"></a>                   <span class="dt">box.padding   =</span> <span class="fl">0.35</span>, </span>
<span id="cb296-14"><a href="rna-seq-analysis.html#cb296-14"></a>                   <span class="dt">point.padding =</span> <span class="fl">0.5</span>,</span>
<span id="cb296-15"><a href="rna-seq-analysis.html#cb296-15"></a>                   <span class="dt">segment.color =</span> <span class="st">&#39;grey50&#39;</span>) <span class="op">+</span></span>
<span id="cb296-16"><a href="rna-seq-analysis.html#cb296-16"></a><span class="st">  </span><span class="kw">theme_minimal</span>()</span>
<span id="cb296-17"><a href="rna-seq-analysis.html#cb296-17"></a></span>
<span id="cb296-18"><a href="rna-seq-analysis.html#cb296-18"></a></span>
<span id="cb296-19"><a href="rna-seq-analysis.html#cb296-19"></a><span class="co">#and both together</span></span>
<span id="cb296-20"><a href="rna-seq-analysis.html#cb296-20"></a><span class="kw">grid.arrange</span>(p1,p2, <span class="dt">ncol=</span><span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-27"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-27-1.png" alt="Ploting many principal components" width="100%" />
<p class="caption">
Figure 6.3: Ploting many principal components
</p>
</div>
<pre><code>#We can do some fancy 3D plots also, try them!
plot_ly(pc_df, x = ~PC1, y = ~PC2, z = ~PC3, color = ~ZT,colorscale = c(&#39;#FFE1A1&#39;, &#39;#683531&#39;))

plot_ly(pc_df, x = ~PC2, y = ~PC3, z = ~PC4, color = ~ZT,colorscale = c(&#39;#FFE1A1&#39;, &#39;#683531&#39;))
</code></pre>
</div>
<div id="differential-gene-expression-analysis-between-zt." class="section level3 tabset tabset-fade tabset-pills" number="6.3.5">
<h3><span class="header-section-number">6.3.5</span> Differential gene expression analysis between ZT.</h3>
<p>We already saw that for DN1: ZT19 and ZT23 looks more similar between them and also ZT3, 7, 15 and 11. We will now explore particular differences. <code>DESeq</code> is the main function here it acts over the object created previously. It executes the normalization and differential gene expression analysis.</p>
<p><strong>Run differentially expression</strong></p>
<div class="sourceCode" id="cb298"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb298-1"><a href="rna-seq-analysis.html#cb298-1"></a>dds.DN1 =<span class="st"> </span><span class="kw">DESeq</span>(dds.DN1)</span></code></pre></div>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="rna-seq-analysis.html#cb305-1"></a><span class="co">#The resultNames is taking the names of the comparisons done by the DESeq function</span></span>
<span id="cb305-2"><a href="rna-seq-analysis.html#cb305-2"></a><span class="kw">resultsNames</span>(dds.DN1)</span></code></pre></div>
<pre><code>## [1] &quot;Intercept&quot;       &quot;ZT_ZT15_vs_ZT11&quot; &quot;ZT_ZT19_vs_ZT11&quot; &quot;ZT_ZT23_vs_ZT11&quot;
## [5] &quot;ZT_ZT3_vs_ZT11&quot;  &quot;ZT_ZT7_vs_ZT11&quot;</code></pre>
<p>The <code>results</code> function extract this result, we will compare ZT3 vs ZT15. Use <code>?results</code> to explore this function further.</p>
<div class="sourceCode" id="cb307"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb307-1"><a href="rna-seq-analysis.html#cb307-1"></a>res.ZT3vsZT15.DN1 &lt;-<span class="st"> </span><span class="kw">results</span>( dds.DN1, <span class="dt">contrast =</span> <span class="kw">c</span>(<span class="st">&quot;ZT&quot;</span>,<span class="st">&quot;ZT3&quot;</span>, <span class="st">&quot;ZT15&quot;</span>),<span class="dt">alpha=</span><span class="fl">0.05</span>)</span>
<span id="cb307-2"><a href="rna-seq-analysis.html#cb307-2"></a></span>
<span id="cb307-3"><a href="rna-seq-analysis.html#cb307-3"></a><span class="kw">head</span>(res.ZT3vsZT15.DN1)</span></code></pre></div>
<pre><code>## log2 fold change (MLE): ZT ZT3 vs ZT15 
## Wald test p-value: ZT ZT3 vs ZT15 
## DataFrame with 6 rows and 6 columns
##                      baseMean log2FoldChange            lfcSE      stat
##                     &lt;numeric&gt;      &lt;numeric&gt;        &lt;numeric&gt; &lt;numeric&gt;
## FBtr0070202                 0             NA               NA        NA
## FBtr0070207   2.1168359707544              0 5.24448449612991         0
## FBtr0070238  2.68357323289241              0  4.8861753047644         0
## FBtr0070251 0.304433817751997              0 5.39799191011387         0
## FBtr0070484 0.151419583982656              0 5.39799191011387         0
## FBtr0070489                 0             NA               NA        NA
##                pvalue      padj
##             &lt;numeric&gt; &lt;numeric&gt;
## FBtr0070202        NA        NA
## FBtr0070207         1         1
## FBtr0070238         1         1
## FBtr0070251         1        NA
## FBtr0070484         1        NA
## FBtr0070489        NA        NA</code></pre>
<blockquote>
<p>What do you think is each column?</p>
</blockquote>
<div class="sourceCode" id="cb309"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb309-1"><a href="rna-seq-analysis.html#cb309-1"></a><span class="kw">mcols</span>(res.ZT3vsZT15.DN1)<span class="op">$</span>description</span></code></pre></div>
<pre><code>## [1] &quot;mean of normalized counts for all samples&quot;
## [2] &quot;log2 fold change (MLE): ZT ZT3 vs ZT15&quot;   
## [3] &quot;standard error: ZT ZT3 vs ZT15&quot;           
## [4] &quot;Wald statistic: ZT ZT3 vs ZT15&quot;           
## [5] &quot;Wald test p-value: ZT ZT3 vs ZT15&quot;        
## [6] &quot;BH adjusted p-values&quot;</code></pre>
<p>We can see the distribution of each variable using the function <code>summary</code></p>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="rna-seq-analysis.html#cb311-1"></a><span class="kw">summary</span>(res.ZT3vsZT15.DN1)</span></code></pre></div>
<pre><code>## 
## out of 25337 with nonzero total read count
## adjusted p-value &lt; 0.05
## LFC &gt; 0 (up)       : 79, 0.31%
## LFC &lt; 0 (down)     : 55, 0.22%
## outliers [1]       : 0, 0%
## low counts [2]     : 3419, 13%
## (mean count &lt; 2)
## [1] see &#39;cooksCutoff&#39; argument of ?results
## [2] see &#39;independentFiltering&#39; argument of ?results</code></pre>
<p><strong>Identify differentially expressed genes: Multiple comparison correction and adjusted pvalue</strong></p>
<p>You might have noticed that apart from pvalues there are adjusted pvalues. Just to recap, when we want to make conclusion we always have some chance to make a mistake. This chance increase with each element we add to out conclusions. Image what happens when we want to make a conclusion about 10.000 genes! We then correct the pvalue. There are different methods, in this case DeSeq2 uses Benjamini-Hochberg adjusted P value.</p>
<p>We can now do an histogram to explore how is the distribution of the padj.</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="rna-seq-analysis.html#cb313-1"></a><span class="co">#histogram</span></span>
<span id="cb313-2"><a href="rna-seq-analysis.html#cb313-2"></a><span class="kw">hist</span>(res.ZT3vsZT15.DN1<span class="op">$</span>padj, <span class="dt">breaks=</span><span class="dv">100</span>, <span class="dt">col=</span><span class="st">&quot;skyblue&quot;</span>, <span class="dt">border=</span><span class="st">&quot;slateblue&quot;</span>, <span class="dt">main=</span><span class="st">&quot;&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-32"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-32-1.png" alt="Histogram of pjusted value distribution" width="100%" />
<p class="caption">
Figure 6.4: Histogram of pjusted value distribution
</p>
</div>
<p><strong>Identify differentially expressed genes: Log2(FoldChange)</strong></p>
<p>Identify differentially expressed genes is a combination of pvalue and a fold change. Fold change is literally the comparison of 2 conditions. Let’s go back to our example.</p>
<table>
<thead>
<tr class="header">
<th align="right">Row Names</th>
<th align="right">WildType.1</th>
<th align="right">WildType.2</th>
<th align="right">Mutant.1</th>
<th align="right">Mutant.2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">geneA</td>
<td align="right">230</td>
<td align="right">210</td>
<td align="right">420</td>
<td align="right">412</td>
</tr>
</tbody>
</table>
<p>Here, geneA seems to be around 2 times more in Mutant compared with the WildType. This would be an <strong>UPregulated</strong> gene.</p>
<p><span class="math display">\[
Mutant/WT=400/200=2
\]</span></p>
<p>If we imagine the opposite situation, a <strong>DOWNregulated</strong> gene, the same comparison would give a not-such intuitive result.</p>
<table>
<thead>
<tr class="header">
<th align="right">Row Names</th>
<th align="right">WildType.1</th>
<th align="right">WildType.2</th>
<th align="right">Mutant.1</th>
<th align="right">Mutant.2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="right">geneA</td>
<td align="right">430</td>
<td align="right">410</td>
<td align="right">220</td>
<td align="right">212</td>
</tr>
</tbody>
</table>
<p><span class="math display">\[
Mutant/WT=200/400=0.5
\]</span></p>
<p>To avoid this, in differential gene expression we use log2. This will give us the following:</p>
<p><span class="math display">\[
UPregulated: log2(Mutant/WT)=log2(400/200)=1
\]</span>
<span class="math display">\[
DOWNregulated: log2(Mutant/WT)=log2(200/400)=-1
\]</span></p>
<p>To know what is being used as denominator (control) we can look into the DeSeq2 manual:
“With no additional arguments to <code>results()</code>, the log2 fold change and Wald test p value will be for the last variable in the design formula, and if this is a factor, the comparison will be the last level of this variable over the reference level.”</p>
<p>In our case then ZT15 is being used as reference</p>
<p>To see significance is to use MA plots which shows log2 fold changes (on the y-axis) versus the mean of normalized counts (on the x-axis).</p>
<div class="sourceCode" id="cb314"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb314-1"><a href="rna-seq-analysis.html#cb314-1"></a><span class="kw">plotMA</span>(res.ZT3vsZT15.DN1,<span class="dt">alpha=</span><span class="fl">0.05</span>) <span class="co">#In red everything with padj &lt; 0.05</span></span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-33"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-33-1.png" alt="MA plot: log2 fold changes vs normalized counts" width="100%" />
<p class="caption">
Figure 6.5: MA plot: log2 fold changes vs normalized counts
</p>
</div>
</div>
<div id="explore-results" class="section level3" number="6.3.6">
<h3><span class="header-section-number">6.3.6</span> Explore results</h3>
<p>We can create a data frame of the results and play around.</p>
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="rna-seq-analysis.html#cb315-1"></a>res.ZT3vsZT15.DN1_df&lt;-<span class="kw">as.data.frame</span>(res.ZT3vsZT15.DN1)</span>
<span id="cb315-2"><a href="rna-seq-analysis.html#cb315-2"></a>res.ZT3vsZT15.DN1_df=res.ZT3vsZT15.DN1_df[<span class="kw">complete.cases</span>(res.ZT3vsZT15.DN1_df),]<span class="co">#remove NAs</span></span>
<span id="cb315-3"><a href="rna-seq-analysis.html#cb315-3"></a>res.ZT3vsZT15.DN1_df<span class="op">$</span>transcript_name=<span class="kw">rownames</span>(res.ZT3vsZT15.DN1_df)</span>
<span id="cb315-4"><a href="rna-seq-analysis.html#cb315-4"></a></span>
<span id="cb315-5"><a href="rna-seq-analysis.html#cb315-5"></a><span class="kw">head</span>(res.ZT3vsZT15.DN1_df)</span></code></pre></div>
<pre><code>##              baseMean log2FoldChange     lfcSE      stat     pvalue padj
## FBtr0070207  2.116836      0.0000000 5.2444845  0.000000 1.00000000    1
## FBtr0070238  2.683573      0.0000000 4.8861753  0.000000 1.00000000    1
## FBtr0070658  2.762362      0.0000000 5.3979919  0.000000 1.00000000    1
## FBtr0071170 16.483738      2.1943705 1.1682234  1.878383 0.06032885    1
## FBtr0071173  4.674766      0.0000000 5.3502754  0.000000 1.00000000    1
## FBtr0071619 49.515374     -0.8457701 0.6368886 -1.327972 0.18418743    1
##             transcript_name
## FBtr0070207     FBtr0070207
## FBtr0070238     FBtr0070238
## FBtr0070658     FBtr0070658
## FBtr0071170     FBtr0071170
## FBtr0071173     FBtr0071173
## FBtr0071619     FBtr0071619</code></pre>
<p><strong>Change the gene names</strong></p>
<p>As you can see, the names of the genes are in a nomenclature that is not intuitive. We can then use some functions that will allow us to change the names for some more informative.</p>
<p>Let’s take one example: FBtr0070207, is a transcript name (not a gene name, but one of the possible variants of the gene) <a href="http://flybase.org/reports/FBtr0070207" class="uri">http://flybase.org/reports/FBtr0070207</a>. We want for this analysis the gene name. For that we will use the package <code>org.Dm.eg.db</code> (each model organism has one of this: <a href="https://bioconductor.org/packages/release/BiocViews.html#___AnnotationData" class="uri">https://bioconductor.org/packages/release/BiocViews.html#___AnnotationData</a>)</p>
<div class="sourceCode" id="cb317"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb317-1"><a href="rna-seq-analysis.html#cb317-1"></a><span class="co">#Explore the options that the package gives us</span></span>
<span id="cb317-2"><a href="rna-seq-analysis.html#cb317-2"></a><span class="kw">keytypes</span>(org.Dm.eg.db)</span></code></pre></div>
<pre><code>##  [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot;  &quot;ENSEMBLTRANS&quot;
##  [6] &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;     &quot;EVIDENCEALL&quot;  &quot;FLYBASE&quot;     
## [11] &quot;FLYBASECG&quot;    &quot;FLYBASEPROT&quot;  &quot;GENENAME&quot;     &quot;GO&quot;           &quot;GOALL&quot;       
## [16] &quot;MAP&quot;          &quot;ONTOLOGY&quot;     &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PMID&quot;        
## [21] &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UNIGENE&quot;      &quot;UNIPROT&quot;</code></pre>
<div class="sourceCode" id="cb319"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb319-1"><a href="rna-seq-analysis.html#cb319-1"></a><span class="co">#We have here the ENSEMBLE transcript names, so we will use that keytype</span></span>
<span id="cb319-2"><a href="rna-seq-analysis.html#cb319-2"></a>genenames &lt;-<span class="st"> </span><span class="kw">mapIds</span>(org.Dm.eg.db,<span class="dt">keys =</span> <span class="kw">rownames</span>(dds.DN1),<span class="dt">column =</span> <span class="st">&quot;SYMBOL&quot;</span>,<span class="dt">keytype=</span><span class="st">&quot;ENSEMBLTRANS&quot;</span>)</span>
<span id="cb319-3"><a href="rna-seq-analysis.html#cb319-3"></a>annotation_DN1 &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">gene_name =</span> genenames, <span class="dt">transcript_name=</span><span class="kw">rownames</span>(dds.DN1), <span class="dt">row.names =</span> <span class="kw">rownames</span>(dds.DN1), <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</span>
<span id="cb319-4"><a href="rna-seq-analysis.html#cb319-4"></a><span class="kw">head</span>(annotation_DN1) </span></code></pre></div>
<pre><code>##             gene_name transcript_name
## FBtr0070202   CG14625     FBtr0070202
## FBtr0070207    fs(1)N     FBtr0070207
## FBtr0070238   CG11409     FBtr0070238
## FBtr0070251   CG11384     FBtr0070251
## FBtr0070484   fs(1)Yb     FBtr0070484
## FBtr0070489   CG12498     FBtr0070489</code></pre>
<p>Now that we have a “mapping” we can use it. We will use the function <code>merge</code>.</p>
<div class="sourceCode" id="cb321"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb321-1"><a href="rna-seq-analysis.html#cb321-1"></a>res.ZT3vsZT15.DN1_df=<span class="kw">merge</span>(res.ZT3vsZT15.DN1_df,annotation_DN1,<span class="dt">by=</span><span class="st">&quot;transcript_name&quot;</span>)</span>
<span id="cb321-2"><a href="rna-seq-analysis.html#cb321-2"></a><span class="kw">head</span>(res.ZT3vsZT15.DN1_df)</span></code></pre></div>
<pre><code>##   transcript_name  baseMean log2FoldChange     lfcSE       stat    pvalue padj
## 1     FBtr0005088 404.52656     -0.3962867 0.4939109 -0.8023445 0.4223537    1
## 2     FBtr0005673   2.35263      1.7199349 2.8262384  0.6085598 0.5428163    1
## 3     FBtr0005674   2.35263      1.7199349 2.8262384  0.6085598 0.5428163    1
## 4     FBtr0070000  44.02153     -0.7589507 1.1324434 -0.6701886 0.5027375    1
## 5     FBtr0070002   4.30420      3.7116759 2.5555706  1.4523863 0.1463942    1
## 6     FBtr0070006  23.06487     -1.1075368 1.0502903 -1.0545054 0.2916516    1
##   gene_name
## 1  Pp2A-29B
## 2      &lt;NA&gt;
## 3      &lt;NA&gt;
## 4      Nep3
## 5    CG9570
## 6    CG9572</code></pre>
<p><strong>UP regulated genes</strong>
Now let’s look for the genes that are changing with than Log2foldChange &gt; 1 (which is a fold change of 2)</p>
<div class="sourceCode" id="cb323"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb323-1"><a href="rna-seq-analysis.html#cb323-1"></a>sig_down=<span class="kw">unique</span>(res.ZT3vsZT15.DN1_df<span class="op">$</span>gene_name[(res.ZT3vsZT15.DN1_df<span class="op">$</span>padj<span class="op">&lt;</span><span class="fl">0.05</span><span class="op">&amp;</span>res.ZT3vsZT15.DN1_df<span class="op">$</span>log2FoldChange<span class="op">&lt;</span>(<span class="op">-</span><span class="dv">1</span>))])</span>
<span id="cb323-2"><a href="rna-seq-analysis.html#cb323-2"></a><span class="kw">head</span>(sig_down)</span></code></pre></div>
<pre><code>## [1] NA        &quot;CG14089&quot; &quot;CG13041&quot; &quot;CG32137&quot; &quot;Pdp1&quot;    &quot;tim&quot;</code></pre>
<p><strong>DOWN regualted genes</strong>
Now let’s look for the genes that are changing with than Log2foldChange &lt; -1 (which is a fold change of 0.5, so a downregulation of 2 times)</p>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="rna-seq-analysis.html#cb325-1"></a>sig_up=<span class="kw">unique</span>(res.ZT3vsZT15.DN1_df<span class="op">$</span>gene_name[(res.ZT3vsZT15.DN1_df<span class="op">$</span>padj<span class="op">&lt;</span><span class="fl">0.05</span><span class="op">&amp;</span>res.ZT3vsZT15.DN1_df<span class="op">$</span>log2FoldChange<span class="op">&gt;</span><span class="dv">1</span>)])</span>
<span id="cb325-2"><a href="rna-seq-analysis.html#cb325-2"></a></span>
<span id="cb325-3"><a href="rna-seq-analysis.html#cb325-3"></a><span class="kw">head</span>(sig_down)</span></code></pre></div>
<pre><code>## [1] NA        &quot;CG14089&quot; &quot;CG13041&quot; &quot;CG32137&quot; &quot;Pdp1&quot;    &quot;tim&quot;</code></pre>
<p><strong>Plot genes</strong></p>
<p>Let’s plot some genes using DeSeq2.</p>
<div class="sourceCode" id="cb327"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb327-1"><a href="rna-seq-analysis.html#cb327-1"></a><span class="kw">plotCounts</span>(dds.DN1, <span class="dt">gene=</span><span class="kw">which.min</span>(res.ZT3vsZT15.DN1<span class="op">$</span>padj), <span class="dt">intgroup=</span><span class="st">&quot;ZT&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-39-1"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-39-1.png" alt="Plot gene expression." width="100%" />
<p class="caption">
Figure 6.6: Plot gene expression.
</p>
</div>
<div class="sourceCode" id="cb328"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb328-1"><a href="rna-seq-analysis.html#cb328-1"></a><span class="kw">plotCounts</span>(dds.DN1, <span class="dt">gene=</span><span class="st">&quot;FBtr0077567&quot;</span>, <span class="dt">intgroup=</span><span class="st">&quot;ZT&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-39-2"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-39-2.png" alt="Plot gene expression." width="100%" />
<p class="caption">
Figure 6.7: Plot gene expression.
</p>
</div>
<p>We can do this in ggplot2</p>
<div class="sourceCode" id="cb329"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb329-1"><a href="rna-seq-analysis.html#cb329-1"></a>d &lt;-<span class="st"> </span><span class="kw">plotCounts</span>(dds.DN1, <span class="dt">gene=</span><span class="st">&quot;FBtr0077567&quot;</span>, <span class="dt">intgroup=</span><span class="st">&quot;ZT&quot;</span>, <span class="dt">returnData=</span><span class="ot">TRUE</span>) <span class="co">#extract eh data from the plot</span></span>
<span id="cb329-2"><a href="rna-seq-analysis.html#cb329-2"></a></span>
<span id="cb329-3"><a href="rna-seq-analysis.html#cb329-3"></a><span class="kw">ggplot</span>(d, <span class="kw">aes</span>(<span class="dt">x=</span>ZT, <span class="dt">y=</span>count)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">w=</span><span class="fl">0.1</span>,<span class="dt">h=</span><span class="dv">0</span>)) <span class="co">#use it to plot in a better way</span></span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-40"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-40-1.png" alt="Plot gene expression using ggplot." width="100%" />
<p class="caption">
Figure 3.11: Plot gene expression using ggplot.
</p>
</div>
<p>Let’s discuss what we see in the graphics.</p>
<p>We have some of the figures that seems to have the exactly same values for each replicate. This is at least suspicious. This might be just the same quantification so it might not be correct to say that they are different isoforms and maybe it is better to collapse them or change the strategy to analyze different isoforms. This also shows us the limitations we have when we use the data already aligned.</p>
</div>
<div id="plot-counts-using-ggplot" class="section level3" number="6.3.7">
<h3><span class="header-section-number">6.3.7</span> Plot counts using ggplot</h3>
<p>Let’s see how we can plot the transcript reads in a better way. Doing a few manipulations on the table we can have even better plots.</p>
<div class="sourceCode" id="cb330"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb330-1"><a href="rna-seq-analysis.html#cb330-1"></a><span class="co">#big plot</span></span>
<span id="cb330-2"><a href="rna-seq-analysis.html#cb330-2"></a><span class="co">#I take the normalized values and do the plots</span></span>
<span id="cb330-3"><a href="rna-seq-analysis.html#cb330-3"></a>toplot&lt;-<span class="kw">as.data.frame</span>(<span class="kw">as.data.frame</span>(<span class="kw">counts</span>(dds.DN1, <span class="dt">normalized=</span>T)))</span>
<span id="cb330-4"><a href="rna-seq-analysis.html#cb330-4"></a>toplot<span class="op">$</span>transcript_name&lt;-<span class="kw">rownames</span>(toplot)</span>
<span id="cb330-5"><a href="rna-seq-analysis.html#cb330-5"></a></span>
<span id="cb330-6"><a href="rna-seq-analysis.html#cb330-6"></a>toplot&lt;-toplot <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">gather</span>(zt_rep, value, <span class="op">-</span>transcript_name) <span class="co">#This is a more complex line of code, let’s explore the results.</span></span>
<span id="cb330-7"><a href="rna-seq-analysis.html#cb330-7"></a><span class="kw">head</span>(toplot)</span></code></pre></div>
<pre><code>##   transcript_name    zt_rep value
## 1     FBtr0070202 DN1_ZT3_1     0
## 2     FBtr0070207 DN1_ZT3_1     0
## 3     FBtr0070238 DN1_ZT3_1     0
## 4     FBtr0070251 DN1_ZT3_1     0
## 5     FBtr0070484 DN1_ZT3_1     0
## 6     FBtr0070489 DN1_ZT3_1     0</code></pre>
<div class="sourceCode" id="cb332"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb332-1"><a href="rna-seq-analysis.html#cb332-1"></a>toplot=<span class="kw">merge</span>(toplot,annotation_DN1,<span class="dt">by=</span><span class="st">&quot;transcript_name&quot;</span>)</span>
<span id="cb332-2"><a href="rna-seq-analysis.html#cb332-2"></a><span class="kw">head</span>(toplot)</span></code></pre></div>
<pre><code>##   transcript_name     zt_rep    value gene_name
## 1     FBtr0005088 DN1_ZT11_2 333.0941  Pp2A-29B
## 2     FBtr0005088 DN1_ZT19_2 332.2765  Pp2A-29B
## 3     FBtr0005088  DN1_ZT7_2 520.4211  Pp2A-29B
## 4     FBtr0005088  DN1_ZT3_2 235.0216  Pp2A-29B
## 5     FBtr0005088 DN1_ZT23_2 392.4796  Pp2A-29B
## 6     FBtr0005088 DN1_ZT19_1 406.2523  Pp2A-29B</code></pre>
<blockquote>
<p>What changed here? How this allow for better plotting?</p>
</blockquote>
<div class="sourceCode" id="cb334"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb334-1"><a href="rna-seq-analysis.html#cb334-1"></a><span class="co">#Lets create now more meta data </span></span>
<span id="cb334-2"><a href="rna-seq-analysis.html#cb334-2"></a>toplot<span class="op">$</span>zt&lt;-<span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">as.character</span>(toplot<span class="op">$</span>zt_rep),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">2</span>)</span>
<span id="cb334-3"><a href="rna-seq-analysis.html#cb334-3"></a>toplot<span class="op">$</span>ztime&lt;-<span class="kw">gsub</span>(toplot<span class="op">$</span>zt,<span class="dt">pattern =</span> <span class="st">&quot;ZT&quot;</span>,<span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>)</span>
<span id="cb334-4"><a href="rna-seq-analysis.html#cb334-4"></a>toplot<span class="op">$</span>rep&lt;-<span class="kw">sapply</span>(<span class="kw">strsplit</span>(<span class="kw">as.character</span>(toplot<span class="op">$</span>zt_rep),<span class="st">&quot;_&quot;</span>),<span class="st">&quot;[[&quot;</span>,<span class="dv">3</span>)</span>
<span id="cb334-5"><a href="rna-seq-analysis.html#cb334-5"></a>toplot<span class="op">$</span>ztime_<span class="dv">2</span>&lt;-<span class="kw">as.numeric</span>(toplot<span class="op">$</span>ztime)</span>
<span id="cb334-6"><a href="rna-seq-analysis.html#cb334-6"></a>toplot<span class="op">$</span>ztime_<span class="dv">2</span>[toplot<span class="op">$</span>rep<span class="op">==</span><span class="st">&quot;2&quot;</span>]&lt;-toplot<span class="op">$</span>ztime_<span class="dv">2</span>[toplot<span class="op">$</span>rep<span class="op">==</span><span class="st">&quot;2&quot;</span>]<span class="op">+</span><span class="dv">24</span></span>
<span id="cb334-7"><a href="rna-seq-analysis.html#cb334-7"></a></span>
<span id="cb334-8"><a href="rna-seq-analysis.html#cb334-8"></a><span class="co">#And plot</span></span>
<span id="cb334-9"><a href="rna-seq-analysis.html#cb334-9"></a>genes.cyc =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;tim&quot;</span>,<span class="st">&quot;Clk&quot;</span>,<span class="st">&quot;per&quot;</span>,<span class="st">&quot;cwo&quot;</span>) <span class="co">#you can put here as many genes as you want</span></span></code></pre></div>
<div class="sourceCode" id="cb335"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb335-1"><a href="rna-seq-analysis.html#cb335-1"></a><span class="kw">ggplot</span>(toplot[toplot<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span>genes.cyc ,],<span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw">as.numeric</span>(ztime),<span class="dt">y =</span> value,<span class="dt">color=</span>gene_name,<span class="dt">shape=</span>rep)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>() </span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-44"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-44-1.png" alt="Count plots with ggplot" width="100%" />
<p class="caption">
Figure 6.8: Count plots with ggplot
</p>
</div>
<div class="sourceCode" id="cb336"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb336-1"><a href="rna-seq-analysis.html#cb336-1"></a><span class="kw">ggplot</span>(toplot[toplot<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span>genes.cyc ,],<span class="kw">aes</span>(<span class="dt">x =</span>ztime_<span class="dv">2</span>,<span class="dt">y =</span> value,<span class="dt">color=</span>gene_name,<span class="dt">shape=</span>rep)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>()</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-45"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-45-1.png" alt="Count plots with ggplot using zt time" width="100%" />
<p class="caption">
Figure 6.9: Count plots with ggplot using zt time
</p>
</div>
<div class="sourceCode" id="cb337"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb337-1"><a href="rna-seq-analysis.html#cb337-1"></a><span class="kw">ggplot</span>(toplot[toplot<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span>genes.cyc ,],<span class="kw">aes</span>(<span class="dt">x =</span>ztime_<span class="dv">2</span>,<span class="dt">y =</span> value,<span class="dt">color=</span>gene_name,<span class="dt">shape=</span>rep)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span><span class="kw">facet_wrap</span>(<span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>,<span class="dt">facets =</span> <span class="op">~</span>gene_name )</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-46"></span>
<img src="05-RNASeq_DeSeq2_files/figure-html/unnamed-chunk-46-1.png" alt="Count plots with ggplot sepparating by gene name" width="100%" />
<p class="caption">
Figure 3.16: Count plots with ggplot sepparating by gene name
</p>
</div>
</div>
</div>
<div id="gene-ontology-go-term-analysis" class="section level2" number="6.4">
<h2><span class="header-section-number">6.4</span> Gene Ontology (GO) term analysis</h2>
<p>The main objective of GO term analysis is to identify common features in a list of gene shared feature that might be “enriched” when compared with the background list of of genes. There are 3 GO term sets:</p>
<ol style="list-style-type: decimal">
<li><strong>Molecular Function</strong>(MF): Molecular function or activities performed by gene product. Examples can be for example <em>catalysis</em> or <em>transport</em>.</li>
<li><strong>Biological Process</strong>(BP): The larger processes, or ‘biological programs’ that these genes takes part in. Examples can be <em>DNA repair</em> or <em>signal transduction</em>.</li>
<li><strong>Cellular Component</strong>(CC): The locations relative to cellular structures in which a gene product performs a function, either cellular compartments.</li>
</ol>
<p>For this we use the known information about each gene. Lets see one example: <em>Gad1</em>.</p>
<p>From <a href="http://flybase.org/reports/FBgn0004516">FlyBase</a>:
“Glutamic acid decarboxylase 1 (Gad1) encodes an essential, nervous system-specific glutamic acid decarboxylase, which is the synthetic enzyme for the major inhibitory neurotransmitter gamma-Aminobutyric acid (GABA). It is required for a multitude of physiological functions and adult behaviors dependent on GABA, including sleep, memory, circadian rhythms and egg hatching.”</p>
<p>Then, MF GO term are <em>glutamate decarboxylase activity</em>, <em>glutamate decarboxylase activity</em>, <em>pyridoxal phosphate binding</em>.
BP: <em>gamma-aminobutyric acid biosynthetic process</em>,<em>glutamate catabolic process</em>,<em>larval locomotor behavior</em>,<em>neuromuscular junction development</em>,<em>neurotransmitter receptor metabolic process</em>,<em>olfactory learning</em>
CC: <em>cytosol</em></p>
<p>Basically then, if a term appears more than “by chance” in a gene list, it will be then significantly enriched.</p>
<p>There are many online tools to do this type of analysis (<a href="http://cbl-gorilla.cs.technion.ac.il/" class="uri">http://cbl-gorilla.cs.technion.ac.il/</a>, ) but here we will use the R package <code>topGO</code>.</p>
<p>We will analyze the results from the RNA seq. It is important to know what to use as a background list of genes. In this case, any expressed gene in the cell-type we are analyzing.</p>
<p><code>topGO</code> requires a list of genes with an indicator of 1 for the genes that should be taken into the analysis and the rest with 0. The full list of genes will be used as background.</p>
<p>To get the list of expressed genes we will use the fact that DeSeq2 actually puts NA in the padjusted column to the lowly expressed genes.</p>
<div class="sourceCode" id="cb338"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb338-1"><a href="rna-seq-analysis.html#cb338-1"></a><span class="co">#Libraries</span></span>
<span id="cb338-2"><a href="rna-seq-analysis.html#cb338-2"></a><span class="kw">library</span>(<span class="st">&quot;topGO&quot;</span>)</span>
<span id="cb338-3"><a href="rna-seq-analysis.html#cb338-3"></a><span class="kw">library</span>(<span class="st">&quot;org.Dm.eg.db&quot;</span>)</span>
<span id="cb338-4"><a href="rna-seq-analysis.html#cb338-4"></a></span>
<span id="cb338-5"><a href="rna-seq-analysis.html#cb338-5"></a><span class="co">#Prepare list of genes that are expressed, complete cases will get rid of the genes with NAs</span></span>
<span id="cb338-6"><a href="rna-seq-analysis.html#cb338-6"></a>genes&lt;-<span class="kw">as.data.frame</span>(res.ZT3vsZT15.DN1_df<span class="op">$</span>gene_name[<span class="kw">complete.cases</span>(res.ZT3vsZT15.DN1_df)])</span>
<span id="cb338-7"><a href="rna-seq-analysis.html#cb338-7"></a><span class="kw">names</span>(genes)&lt;-<span class="st">&quot;gene_name&quot;</span></span>
<span id="cb338-8"><a href="rna-seq-analysis.html#cb338-8"></a></span>
<span id="cb338-9"><a href="rna-seq-analysis.html#cb338-9"></a><span class="co">#Add the indicator</span></span>
<span id="cb338-10"><a href="rna-seq-analysis.html#cb338-10"></a>genes<span class="op">$</span>ind=<span class="dv">0</span></span>
<span id="cb338-11"><a href="rna-seq-analysis.html#cb338-11"></a>genes<span class="op">$</span>ind[genes<span class="op">$</span>gene_name <span class="op">%in%</span><span class="st"> </span>sig_up] =<span class="st"> </span><span class="dv">1</span></span>
<span id="cb338-12"><a href="rna-seq-analysis.html#cb338-12"></a></span>
<span id="cb338-13"><a href="rna-seq-analysis.html#cb338-13"></a>alg &lt;-<span class="st"> </span><span class="kw">factor</span>( <span class="kw">as.integer</span>( genes<span class="op">$</span>ind ) )</span>
<span id="cb338-14"><a href="rna-seq-analysis.html#cb338-14"></a><span class="kw">names</span>(alg) &lt;-<span class="st"> </span>genes<span class="op">$</span>gene_name</span>
<span id="cb338-15"><a href="rna-seq-analysis.html#cb338-15"></a>onts =<span class="st"> </span><span class="kw">c</span>( <span class="st">&quot;MF&quot;</span>, <span class="st">&quot;BP&quot;</span>, <span class="st">&quot;CC&quot;</span> )</span>
<span id="cb338-16"><a href="rna-seq-analysis.html#cb338-16"></a></span>
<span id="cb338-17"><a href="rna-seq-analysis.html#cb338-17"></a>tab =<span class="st"> </span><span class="kw">as.list</span>(onts)</span>
<span id="cb338-18"><a href="rna-seq-analysis.html#cb338-18"></a>listGOtab =<span class="st">  </span><span class="kw">as.list</span>(onts)</span>
<span id="cb338-19"><a href="rna-seq-analysis.html#cb338-19"></a><span class="kw">names</span>(tab) =<span class="st"> </span>onts</span>
<span id="cb338-20"><a href="rna-seq-analysis.html#cb338-20"></a></span>
<span id="cb338-21"><a href="rna-seq-analysis.html#cb338-21"></a><span class="co">#For loop going over the 3 GO term sets</span></span>
<span id="cb338-22"><a href="rna-seq-analysis.html#cb338-22"></a><span class="cf">for</span>(e <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>){</span>
<span id="cb338-23"><a href="rna-seq-analysis.html#cb338-23"></a>  </span>
<span id="cb338-24"><a href="rna-seq-analysis.html#cb338-24"></a>  <span class="co">#start topGO data</span></span>
<span id="cb338-25"><a href="rna-seq-analysis.html#cb338-25"></a>  tgd &lt;-<span class="st"> </span><span class="kw">new</span>( <span class="st">&quot;topGOdata&quot;</span>, <span class="dt">ontology=</span>onts[e], <span class="dt">allGenes =</span> alg, <span class="dt">nodeSize=</span><span class="dv">5</span>,</span>
<span id="cb338-26"><a href="rna-seq-analysis.html#cb338-26"></a>              <span class="dt">annot=</span>annFUN.org, <span class="dt">mapping=</span><span class="st">&quot;org.Dm.eg.db&quot;</span>, <span class="dt">ID =</span> <span class="st">&quot;symbol&quot;</span> )</span>
<span id="cb338-27"><a href="rna-seq-analysis.html#cb338-27"></a>  listGO =<span class="st"> </span><span class="kw">genesInTerm</span>(tgd)</span>
<span id="cb338-28"><a href="rna-seq-analysis.html#cb338-28"></a>  </span>
<span id="cb338-29"><a href="rna-seq-analysis.html#cb338-29"></a>  resultTopGO.elim &lt;-<span class="st"> </span><span class="kw">runTest</span>(tgd, <span class="dt">algorithm =</span> <span class="st">&quot;elim&quot;</span>, <span class="dt">statistic =</span> <span class="st">&quot;Fisher&quot;</span> )</span>
<span id="cb338-30"><a href="rna-seq-analysis.html#cb338-30"></a>  resultTopGO.classic &lt;-<span class="st"> </span><span class="kw">runTest</span>(tgd, <span class="dt">algorithm =</span> <span class="st">&quot;classic&quot;</span>, <span class="dt">statistic =</span> <span class="st">&quot;Fisher&quot;</span> )</span>
<span id="cb338-31"><a href="rna-seq-analysis.html#cb338-31"></a>  tab[[e]] &lt;-<span class="st"> </span><span class="kw">GenTable</span>( tgd, <span class="dt">Fisher.elim =</span> resultTopGO.elim, </span>
<span id="cb338-32"><a href="rna-seq-analysis.html#cb338-32"></a>                        <span class="dt">Fisher.classic =</span> resultTopGO.classic,</span>
<span id="cb338-33"><a href="rna-seq-analysis.html#cb338-33"></a>                        <span class="dt">orderBy =</span> <span class="st">&quot;Fisher.classic&quot;</span>,  <span class="dt">topNodes =</span>  <span class="dv">100</span>)</span>
<span id="cb338-34"><a href="rna-seq-analysis.html#cb338-34"></a>  tab[[e]]<span class="op">$</span>FDR&lt;-<span class="kw">p.adjust</span>(tab[[e]]<span class="op">$</span>Fisher.classic,<span class="dt">method =</span> <span class="st">&quot;fdr&quot;</span>) <span class="co">#I correct the pvalues </span></span>
<span id="cb338-35"><a href="rna-seq-analysis.html#cb338-35"></a>  tab[[e]]<span class="op">$</span>ontology=onts[e]</span>
<span id="cb338-36"><a href="rna-seq-analysis.html#cb338-36"></a>}</span>
<span id="cb338-37"><a href="rna-seq-analysis.html#cb338-37"></a></span>
<span id="cb338-38"><a href="rna-seq-analysis.html#cb338-38"></a>allGO =<span class="st"> </span><span class="kw">rbind</span> (<span class="kw">as.data.frame</span>(tab[[<span class="dv">2</span>]]),<span class="kw">as.data.frame</span>(tab[[<span class="dv">1</span>]]),<span class="kw">as.data.frame</span>(tab[[<span class="dv">3</span>]]))</span>
<span id="cb338-39"><a href="rna-seq-analysis.html#cb338-39"></a><span class="kw">write.table</span>(allGO, <span class="kw">paste0</span>(<span class="st">&quot;GOterm.txt&quot;</span>),<span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span> , <span class="dt">row.names=</span>F)</span></code></pre></div>
</div>
<div id="extra-negative-binomial-model" class="section level2" number="6.5">
<h2><span class="header-section-number">6.5</span> Extra: negative binomial model</h2>
<p>The model would be then: counts Kij for gene i, sample j are modeled using a negative binomial distribution with fitted mean μij and a gene-specific dispersion parameter αi. The fitted mean is composed of a sample-specific size factor sj and a parameter qij proportional to the expected true concentration of fragments for sample j.</p>
<p><span class="math display">\[
K_{ij}∼NB(μ_{ij},α_i)
\]</span>
<span class="math display">\[
μij=s_jq_{ij}
\]</span></p>
</div>
<div id="activity" class="section level2" number="6.6">
<h2><span class="header-section-number">6.6</span> Activity</h2>
<blockquote>
<p>Finish the analysis for the other cell-types.
Find a way to compare the results (hint: look at dendrograms)</p>
</blockquote>
</div>
<div id="resources-and-bibliography" class="section level2" number="6.7">
<h2><span class="header-section-number">6.7</span> Resources and Bibliography</h2>
<p>Dubowy C, Sehgal A. Circadian Rhythms and Sleep in Drosophila melanogaster. Genetics. 2017;205(4):1373-1397. <a href="doi:10.1534/genetics.115.185157" class="uri">doi:10.1534/genetics.115.185157</a></p>
<p>MEIRELES-FILHO, Antonio Carlos Alves and KYRIACOU, Charalambos Panayiotis. Circadian rhythms in insect disease vectors. Mem. Inst. Oswaldo Cruz [online]. 2013, vol.108, suppl.1 [cited 2020-07-08], pp.48-58.</p>
<p>Abruzzi KC, Zadina A, Luo W, Wiyanto E, Rahman R, Guo F, et al. (2017) RNA-seq analysis of Drosophila clock and non-clock neurons reveals neuron-specific cycling and novel candidate neuropeptides. PLoS Genet 13(2): e1006613. <a href="https://doi.org/10.1371/journal.pgen.1006613" class="uri">https://doi.org/10.1371/journal.pgen.1006613</a></p>
<p>Schubert FK, Hagedorn N, Yoshii T, Helfrich-Förster C, Rieger D. Neuroanatomical details of the lateral neurons of Drosophila melanogaster support their functional role in the circadian system. J Comp Neurol. 2018;526(7):1209-1231. <a href="doi:10.1002/cne.24406" class="uri">doi:10.1002/cne.24406</a></p>
<p>DeSeq2 documentations: <a href="http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html" class="uri">http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html</a></p>
<p>Introduction to implementation steps of MetaCycle
Gang Wu, Ron Anafi, Michael Hughes, Karl Kornacker, and John Hogenesch
2015-12-04
<a href="https://cran.r-project.org/web/packages/MetaCycle/vignettes/implementation.html" class="uri">https://cran.r-project.org/web/packages/MetaCycle/vignettes/implementation.html</a></p>
<p>Gene Ontology overview <a href="http://geneontology.org/docs/ontology-documentation/" class="uri">http://geneontology.org/docs/ontology-documentation/</a></p>
</div>
</div>
<div class="psyteachr_footer">
  
</div>
<script>

/* update total correct if #total_correct exists */
update_total_correct = function() {
  if (t = document.getElementById("total_correct")) {
    t.innerHTML =
      document.getElementsByClassName("correct").length + " of " +
      document.getElementsByClassName("solveme").length + " correct";
  }
}

/* solution button toggling function */
b_func = function() {
  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "");
  }
  
  if (my_answer !== "" & real_answers.includes(my_answer)) {
    cl.add("correct");
  } else {
    cl.remove("correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol){
    var tol = JSON.parse(this.dataset.tol);  
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("correct");
    } else {
      cl.remove("correct");
    }  
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("correct");
    }  
  }
  
  update_total_correct();
}

window.onload = function() {
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('solution')) {
      buttons[i].onclick = b_func;
    }
  }
  
  /* set up solveme inputs */
  var solveme = document.getElementsByClassName("solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off"); 
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";
    
    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;
    
    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;
  }
  
  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="chip-sequencing-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="circadian-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"],
"google": false
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": {},
"toc": {
"collapse": "section",
"scroll_highlight": true
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
