<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Lab 3 Visualization and Statistical test | Genomics with R for biologists</title>
  <meta name="description" content="This is the index for the coursebook" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Lab 3 Visualization and Statistical test | Genomics with R for biologists" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is the index for the coursebook" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Lab 3 Visualization and Statistical test | Genomics with R for biologists" />
  
  <meta name="twitter:description" content="This is the index for the coursebook" />
  

<meta name="author" content="Ines Lucia Patop: inespatop@brandeis.edu" />


<meta name="date" content="2020-07-05" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="data-manipulation.html"/>
<link rel="next" href="sequencing-techniques-and-preprocessing.html"/>
<script src="libs/header-attrs-2.2/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="include/psyteachr.css" type="text/css" />
<link rel="stylesheet" href="include/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Genomics with R for biologists</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Overview</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#course-aims"><i class="fa fa-check"></i><b>0.1</b> Course Aims</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#intended-learning-outcomes"><i class="fa fa-check"></i><b>0.2</b> Intended Learning Outcomes</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#acknowledgments"><i class="fa fa-check"></i><b>0.3</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html"><i class="fa fa-check"></i><b>1</b> Introduction to R and RStudio</a>
<ul>
<li class="chapter" data-level="1.1" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#objectives"><i class="fa fa-check"></i><b>1.1</b> Objectives</a></li>
<li class="chapter" data-level="1.2" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#introduction"><i class="fa fa-check"></i><b>1.2</b> Introduction</a></li>
<li class="chapter" data-level="1.3" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#rstudio-world"><i class="fa fa-check"></i><b>1.3</b> RStudio world</a></li>
<li class="chapter" data-level="1.4" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#create-a-project"><i class="fa fa-check"></i><b>1.4</b> Create a project</a></li>
<li class="chapter" data-level="1.5" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#useful-shortcuts-in-rstudio"><i class="fa fa-check"></i><b>1.5</b> Useful shortcuts in RStudio:</a></li>
<li class="chapter" data-level="1.6" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#installing-and-loading-packages"><i class="fa fa-check"></i><b>1.6</b> Installing and loading packages</a></li>
<li class="chapter" data-level="1.7" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#creating-objects"><i class="fa fa-check"></i><b>1.7</b> Creating objects</a></li>
<li class="chapter" data-level="1.8" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#r-as-a-calculator"><i class="fa fa-check"></i><b>1.8</b> R as a calculator</a></li>
<li class="chapter" data-level="1.9" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#functions"><i class="fa fa-check"></i><b>1.9</b> Functions</a></li>
<li class="chapter" data-level="1.10" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#loops"><i class="fa fa-check"></i><b>1.10</b> Loops</a></li>
<li class="chapter" data-level="1.11" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#conditions"><i class="fa fa-check"></i><b>1.11</b> Conditions</a></li>
<li class="chapter" data-level="1.12" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#character-objects"><i class="fa fa-check"></i><b>1.12</b> Character objects</a></li>
<li class="chapter" data-level="1.13" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#identify"><i class="fa fa-check"></i><b>1.13</b> Identify</a></li>
<li class="chapter" data-level="1.14" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#modify"><i class="fa fa-check"></i><b>1.14</b> Modify</a></li>
<li class="chapter" data-level="1.15" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#useful-resources"><i class="fa fa-check"></i><b>1.15</b> Useful resources</a></li>
<li class="chapter" data-level="1.16" data-path="introduction-to-r-and-rstudio.html"><a href="introduction-to-r-and-rstudio.html#activity"><i class="fa fa-check"></i><b>1.16</b> Activity:</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="data-manipulation.html"><a href="data-manipulation.html"><i class="fa fa-check"></i><b>2</b> Data manipulation</a>
<ul>
<li class="chapter" data-level="2.1" data-path="data-manipulation.html"><a href="data-manipulation.html#objectives-1"><i class="fa fa-check"></i><b>2.1</b> Objectives</a></li>
<li class="chapter" data-level="2.2" data-path="data-manipulation.html"><a href="data-manipulation.html#introduction-1"><i class="fa fa-check"></i><b>2.2</b> Introduction</a></li>
<li class="chapter" data-level="2.3" data-path="data-manipulation.html"><a href="data-manipulation.html#load-data"><i class="fa fa-check"></i><b>2.3</b> Load data</a></li>
<li class="chapter" data-level="2.4" data-path="data-manipulation.html"><a href="data-manipulation.html#data-exploration"><i class="fa fa-check"></i><b>2.4</b> Data exploration</a></li>
<li class="chapter" data-level="2.5" data-path="data-manipulation.html"><a href="data-manipulation.html#subsetting"><i class="fa fa-check"></i><b>2.5</b> Subsetting</a></li>
<li class="chapter" data-level="2.6" data-path="data-manipulation.html"><a href="data-manipulation.html#activity-1"><i class="fa fa-check"></i><b>2.6</b> Activity:</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html"><i class="fa fa-check"></i><b>3</b> Visualization and Statistical test</a>
<ul>
<li class="chapter" data-level="3.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#objectives-2"><i class="fa fa-check"></i><b>3.1</b> Objectives</a></li>
<li class="chapter" data-level="3.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#introduction-2"><i class="fa fa-check"></i><b>3.2</b> Introduction</a></li>
<li class="chapter" data-level="3.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#plots"><i class="fa fa-check"></i><b>3.3</b> Plots</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#line-and-points"><i class="fa fa-check"></i><b>3.3.1</b> Line and points</a></li>
<li class="chapter" data-level="3.3.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#historgam-and-density-plots"><i class="fa fa-check"></i><b>3.3.2</b> Historgam and density plots</a></li>
<li class="chapter" data-level="3.3.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#boxplot"><i class="fa fa-check"></i><b>3.3.3</b> Boxplot</a></li>
<li class="chapter" data-level="3.3.4" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#saving-plots"><i class="fa fa-check"></i><b>3.3.4</b> Saving plots</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#statistical-test"><i class="fa fa-check"></i><b>3.4</b> Statistical test</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#descriptive-statistics"><i class="fa fa-check"></i><b>3.4.1</b> Descriptive statistics</a></li>
<li class="chapter" data-level="3.4.2" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#t-testwilcoxon"><i class="fa fa-check"></i><b>3.4.2</b> T-test/Wilcoxon</a></li>
<li class="chapter" data-level="3.4.3" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#anovakruskalwallis"><i class="fa fa-check"></i><b>3.4.3</b> Anova/Kruskal–Wallis</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#activity-2"><i class="fa fa-check"></i><b>3.5</b> Activity:</a></li>
<li class="chapter" data-level="3.6" data-path="visualization-and-statistical-test.html"><a href="visualization-and-statistical-test.html#resources"><i class="fa fa-check"></i><b>3.6</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html"><i class="fa fa-check"></i><b>4</b> Sequencing techniques and preprocessing</a>
<ul>
<li class="chapter" data-level="4.1" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#objectives-3"><i class="fa fa-check"></i><b>4.1</b> Objectives</a></li>
<li class="chapter" data-level="4.2" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#introduction-3"><i class="fa fa-check"></i><b>4.2</b> Introduction</a></li>
<li class="chapter" data-level="4.3" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#different-type-of-data-dna-rna-splicing-single-end-paired-end"><i class="fa fa-check"></i><b>4.3</b> Different type of data (DNA, RNA, splicing, single-end paired-end)</a></li>
<li class="chapter" data-level="4.4" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#from-the-sequencer-machine-to-the-sequence"><i class="fa fa-check"></i><b>4.4</b> From the sequencer machine to the sequence</a></li>
<li class="chapter" data-level="4.5" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#general-processing"><i class="fa fa-check"></i><b>4.5</b> General processing</a>
<ul>
<li class="chapter" data-level="4.5.1" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#visualization-of-alignment-igv"><i class="fa fa-check"></i><b>4.5.1</b> Visualization of alignment (IGV)</a></li>
<li class="chapter" data-level="4.5.2" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#gene-counts-read-esat-feature-counts-peak-calling"><i class="fa fa-check"></i><b>4.5.2</b> Gene counts / read (esat feature counts) /peak calling</a></li>
</ul></li>
<li class="chapter" data-level="4.6" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#using-the-terminal"><i class="fa fa-check"></i><b>4.6</b> Using the terminal</a></li>
<li class="chapter" data-level="4.7" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#a-bit-of-shell-commands"><i class="fa fa-check"></i><b>4.7</b> A bit of shell commands</a></li>
<li class="chapter" data-level="4.8" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#resources-1"><i class="fa fa-check"></i><b>4.8</b> Resources</a></li>
<li class="chapter" data-level="4.9" data-path="sequencing-techniques-and-preprocessing.html"><a href="sequencing-techniques-and-preprocessing.html#bibliography"><i class="fa fa-check"></i><b>4.9</b> Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html"><i class="fa fa-check"></i><b>5</b> Chip Sequencing Analysis</a>
<ul>
<li class="chapter" data-level="5.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#objectives-4"><i class="fa fa-check"></i><b>5.1</b> Objectives</a></li>
<li class="chapter" data-level="5.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#introduction-4"><i class="fa fa-check"></i><b>5.2</b> Introduction</a></li>
<li class="chapter" data-level="5.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#data-pre-processing-peak-calling"><i class="fa fa-check"></i><b>5.3</b> Data pre-processing (peak-calling):</a></li>
<li class="chapter" data-level="5.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#exploring-the-results"><i class="fa fa-check"></i><b>5.4</b> Exploring the results:</a></li>
<li class="chapter" data-level="5.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#annotating-the-peaks."><i class="fa fa-check"></i><b>5.5</b> Annotating the peaks.</a>
<ul>
<li class="chapter" data-level="5.5.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#read-the-annotation-file"><i class="fa fa-check"></i><b>5.5.1</b> Read the annotation file</a></li>
<li class="chapter" data-level="5.5.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#get-the-genes-from-this-annotation-data-base"><i class="fa fa-check"></i><b>5.5.2</b> Get the genes from this annotation data base</a></li>
<li class="chapter" data-level="5.5.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#get-the-genomic-ranges-of-the-peaks-from-macs-and-find-which-gene-is-closer-to-them-ie.-annotate-them"><i class="fa fa-check"></i><b>5.5.3</b> Get the genomic ranges of the peaks from MACS and find which gene is closer to them (ie. Annotate them)</a></li>
<li class="chapter" data-level="5.5.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#change-the-gene-format."><i class="fa fa-check"></i><b>5.5.4</b> Change the gene format.</a></li>
<li class="chapter" data-level="5.5.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-and-export-the-table"><i class="fa fa-check"></i><b>5.5.5</b> Explore and export the table</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#annotate-with-chipseeker"><i class="fa fa-check"></i><b>5.6</b> Annotate with <code>ChIPseeker</code></a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-and-export-the-output"><i class="fa fa-check"></i><b>5.6.1</b> Explore and Export the output</a></li>
<li class="chapter" data-level="5.6.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#positive-controls."><i class="fa fa-check"></i><b>5.6.2</b> Positive controls.</a></li>
<li class="chapter" data-level="5.6.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#plots-1"><i class="fa fa-check"></i><b>5.6.3</b> Plots</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#sequence-motif-analysis"><i class="fa fa-check"></i><b>5.7</b> Sequence Motif analysis</a>
<ul>
<li class="chapter" data-level="5.7.1" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#put-the-peak-sequences-in-fasta-format"><i class="fa fa-check"></i><b>5.7.1</b> Put the peak sequences in fasta format</a></li>
<li class="chapter" data-level="5.7.2" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-the-data"><i class="fa fa-check"></i><b>5.7.2</b> Explore the data</a></li>
<li class="chapter" data-level="5.7.3" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#find-the-motifs"><i class="fa fa-check"></i><b>5.7.3</b> Find the MOTIFS</a></li>
<li class="chapter" data-level="5.7.4" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#explore-the-results"><i class="fa fa-check"></i><b>5.7.4</b> Explore the results</a></li>
<li class="chapter" data-level="5.7.5" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#plot-the-results"><i class="fa fa-check"></i><b>5.7.5</b> Plot the results</a></li>
<li class="chapter" data-level="5.7.6" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#plotting-with-ggplot2"><i class="fa fa-check"></i><b>5.7.6</b> Plotting with Ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="5.8" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#session-info-all-the-packages-installed."><i class="fa fa-check"></i><b>5.8</b> Session info: all the packages installed.</a></li>
<li class="chapter" data-level="5.9" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#extra-how-to-solve-some-common-annotation-issues."><i class="fa fa-check"></i><b>5.9</b> Extra: How to solve some common annotation issues.</a></li>
<li class="chapter" data-level="5.10" data-path="chip-sequencing-analysis.html"><a href="chip-sequencing-analysis.html#extra-detailed-explanation-of-the-macs-output-files"><i class="fa fa-check"></i><b>5.10</b> Extra: Detailed explanation of the MACS output files</a></li>
</ul></li>
<li class="appendix"><span><b>Appendices</b></span></li>
<li class="divider"></li>
<li><a rel="license" href="https://creativecommons.org/licenses/by-sa/4.0/" 
    target="blank"><img alt="Creative Commons License" 
    style="border-width:0" 
    src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><li>
<li><a href="https://psyteachr.github.io/" target="blank">PsyTeachR</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Genomics with R for biologists</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="visualization-and-statistical-test" class="section level1" number="3">
<h1><span class="header-section-number">Lab 3</span> Visualization and Statistical test</h1>
<div id="objectives-2" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Objectives</h2>
<p>After this section you should be able to:</p>
<ol style="list-style-type: decimal">
<li>Plot and explore the data in many ways.</li>
<li>Do statistical tests on the data.</li>
</ol>
</div>
<div id="introduction-2" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Introduction</h2>
<p>We will use the same data we already explored in last chapter. Remember: chickens weight, age (Time) for different diets.</p>
<p>There are many things we can explore in this data. Each question we might want to answer will be better addressed using different plots.
For example:
1. If the chickens are older we expect them to be bigger. This can be visualized using a dotplot.
2. We might want to see the distribution of weight separated by diet. This can be addressed by a histogram.</p>
</div>
<div id="plots" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Plots</h2>
<p>We will use the package <code>ggplot2</code>. It is a very useful and documented package. We will focus on the <code>ggplot</code> function. This function generates plots as <em>layers</em>. This allows us to manipulate the colors, the plot type, etc. I know it can be difficult to understand it at the beginning but after a while it becomes really intuitive.</p>
<p>Important things to consider:
1. We will be able to plot anything that is a <em>column</em> in the data frame.
2. Everything is or can be a layer in the plot.
3. When you decide to color or shape by a factor that separates your data this will impact the plot.</p>
<p>Again,we can plot any column. So lets axplore the columns. It is important to know the class of each column. It is not the same trying to plot a number, than a letter.</p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="visualization-and-statistical-test.html#cb131-1"></a><span class="kw">names</span>(ChickWeight) <span class="co">#names of the columns in the data frame</span></span></code></pre></div>
<pre><code>## [1] &quot;weight&quot; &quot;Time&quot;   &quot;Chick&quot;  &quot;Diet&quot;</code></pre>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="visualization-and-statistical-test.html#cb133-1"></a><span class="kw">head</span>(ChickWeight) <span class="co">#head of the data frame</span></span></code></pre></div>
<pre><code>##   weight Time Chick Diet
## 1     42    0     1    1
## 2     51    2     1    1
## 3     59    4     1    1
## 4     64    6     1    1
## 5     76    8     1    1
## 6     93   10     1    1</code></pre>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="visualization-and-statistical-test.html#cb135-1"></a><span class="kw">str</span>(ChickWeight) <span class="co">#structure of the data frame</span></span></code></pre></div>
<pre><code>## &#39;data.frame&#39;:    578 obs. of  4 variables:
##  $ weight: num  42 51 59 64 76 93 106 125 149 171 ...
##  $ Time  : num  0 2 4 6 8 10 12 14 16 18 ...
##  $ Chick : Ord.factor w/ 50 levels &quot;18&quot;&lt;&quot;16&quot;&lt;&quot;15&quot;&lt;..: 15 15 15 15 15 15 15 15 15 15 ...
##  $ Diet  : Factor w/ 4 levels &quot;1&quot;,&quot;2&quot;,&quot;3&quot;,&quot;4&quot;: 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
<div id="line-and-points" class="section level3" number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Line and points</h3>
<p>To see things as correlations, we usually use points and lines. We will see how to do it using different plot options.</p>
<p>Dot plot with basic qplot (from ggplot but les complex)</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="visualization-and-statistical-test.html#cb137-1"></a><span class="kw">qplot</span>(<span class="dt">data=</span>ChickWeight,<span class="dt">x =</span> weight, <span class="dt">y=</span>Time, <span class="dt">geom =</span> <span class="kw">c</span>(<span class="st">&quot;line&quot;</span>,<span class="st">&quot;point&quot;</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-351"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-35-1.png" alt="Point and line plots" width="100%" />
<p class="caption">
Figure 3.1: Point and line plots
</p>
</div>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb138-1"><a href="visualization-and-statistical-test.html#cb138-1"></a><span class="kw">qplot</span>(<span class="dt">data=</span>ChickWeight,<span class="dt">x =</span> weight, <span class="dt">y=</span>Time, <span class="dt">geom =</span> <span class="kw">c</span>(<span class="st">&quot;line&quot;</span>,<span class="st">&quot;point&quot;</span>), <span class="dt">colour=</span>Diet) <span class="co">#adding the color helps to separate the data</span></span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-352"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-35-2.png" alt="Point and line plots" width="100%" />
<p class="caption">
Figure 3.2: Point and line plots
</p>
</div>
<p>The same using ggplot:</p>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb139-1"><a href="visualization-and-statistical-test.html#cb139-1"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">y =</span> weight, <span class="dt">x=</span>Time,<span class="dt">colour=</span>Diet))<span class="op">+</span><span class="st"> </span><span class="co">#data and basic things about the plot</span></span>
<span id="cb139-2"><a href="visualization-and-statistical-test.html#cb139-2"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="co">#add the type of plot</span></span>
<span id="cb139-3"><a href="visualization-and-statistical-test.html#cb139-3"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>) <span class="co">#add a colot pallet</span></span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-361"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-36-1.png" alt="Point and line plots" width="100%" />
<p class="caption">
Figure 3.3: Point and line plots
</p>
</div>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb140-1"><a href="visualization-and-statistical-test.html#cb140-1"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">y =</span> weight, <span class="dt">x=</span>Time,<span class="dt">colour=</span>Diet))<span class="op">+</span><span class="st"> </span><span class="co">#data and basic things about the plot</span></span>
<span id="cb140-2"><a href="visualization-and-statistical-test.html#cb140-2"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="co">#add the type of plot</span></span>
<span id="cb140-3"><a href="visualization-and-statistical-test.html#cb140-3"></a><span class="st">  </span><span class="kw">geom_smooth</span>() <span class="op">+</span><span class="st"> </span><span class="co">#add a trend line of mean plus se</span></span>
<span id="cb140-4"><a href="visualization-and-statistical-test.html#cb140-4"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-362"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-36-2.png" alt="Point and line plots" width="100%" />
<p class="caption">
Figure 3.4: Point and line plots
</p>
</div>
</div>
<div id="historgam-and-density-plots" class="section level3" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> Historgam and density plots</h3>
<p>Histograms are used to represent the distribution of a variable over the population. If you want to read more you can go to this <a href="https://en.wikipedia.org/wiki/Histogram">link</a>. Other way to represent the same thing is to use cumulative plots we are not going to explore them now but if you are interested in doing them with ggplot you can go to this <a href="https://stackoverflow.com/questions/15844919/cumulative-plot-using-ggplot2">link</a></p>
<p><a href="https://en.wikipedia.org/wiki/Density_estimation">Density plots</a> are similar to histograms but implies a more complex treatment of the data. They look like smooth histogram. They are the probability density function of the variable.</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="visualization-and-statistical-test.html#cb141-1"></a><span class="kw">qplot</span>(<span class="dt">data =</span> ChickWeight,<span class="dt">x=</span>weight, <span class="dt">binwith=</span><span class="dv">10</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-371"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-37-1.png" alt="Histogram and Density plots" width="100%" />
<p class="caption">
Figure 3.5: Histogram and Density plots
</p>
</div>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb142-1"><a href="visualization-and-statistical-test.html#cb142-1"></a><span class="kw">qplot</span>(<span class="dt">data =</span> ChickWeight,<span class="dt">x=</span>weight, <span class="dt">binwith=</span><span class="dv">10</span>, <span class="dt">colour=</span>Diet) <span class="co">#the color separates the data</span></span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-372"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-37-2.png" alt="Histogram and Density plots" width="100%" />
<p class="caption">
Figure 3.6: Histogram and Density plots
</p>
</div>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="visualization-and-statistical-test.html#cb143-1"></a><span class="kw">qplot</span>(<span class="dt">data =</span> ChickWeight,<span class="dt">x=</span>weight, <span class="dt">geom =</span> <span class="st">&quot;density&quot;</span>, <span class="dt">colour=</span>Diet)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-373"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-37-3.png" alt="Histogram and Density plots" width="100%" />
<p class="caption">
Figure 3.7: Histogram and Density plots
</p>
</div>
<p>With ggplot</p>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="visualization-and-statistical-test.html#cb144-1"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">x=</span>weight,<span class="dt">color=</span>Diet))<span class="op">+</span></span>
<span id="cb144-2"><a href="visualization-and-statistical-test.html#cb144-2"></a><span class="st">  </span><span class="kw">geom_histogram</span>(<span class="dt">fill=</span><span class="st">&quot;white&quot;</span>, <span class="dt">position=</span><span class="st">&quot;identity&quot;</span>)<span class="op">+</span></span>
<span id="cb144-3"><a href="visualization-and-statistical-test.html#cb144-3"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>)<span class="co">#this is selecting the color scheme, try taking it out, or mofyfying it</span></span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-381"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-38-1.png" alt="Point and line plots" width="100%" />
<p class="caption">
Figure 3.8: Point and line plots
</p>
</div>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="visualization-and-statistical-test.html#cb145-1"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">x=</span>weight,<span class="dt">fill=</span>Diet))<span class="op">+</span></span>
<span id="cb145-2"><a href="visualization-and-statistical-test.html#cb145-2"></a><span class="st">  </span><span class="kw">geom_density</span>( <span class="dt">alpha=</span><span class="fl">0.5</span>)<span class="op">+</span><span class="st"> </span><span class="co">#the density plot with the option to modify the transparency of the polot solor, it goes between 0 and 1. Try modifying it.</span></span>
<span id="cb145-3"><a href="visualization-and-statistical-test.html#cb145-3"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-382"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-38-2.png" alt="Point and line plots" width="100%" />
<p class="caption">
Figure 3.9: Point and line plots
</p>
</div>
</div>
<div id="boxplot" class="section level3" number="3.3.3">
<h3><span class="header-section-number">3.3.3</span> Boxplot</h3>
<p>Boxplots are a nice way to visualize the data distribution and to get and intuition of how this is different between conditions. As you can see in this figure, it summarizes a LOT of information:</p>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-39"></span>
<img src="images/boxplot.jpg" alt="Boxplot description. Figrue affapted from https://www.simplypsychology.org/boxplots.html" width="500px" />
<p class="caption">
Figure 3.10: Boxplot description. Figrue affapted from <a href="https://www.simplypsychology.org/boxplots.html" class="uri">https://www.simplypsychology.org/boxplots.html</a>
</p>
</div>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="visualization-and-statistical-test.html#cb146-1"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">y=</span>weight,<span class="dt">x=</span><span class="kw">as.factor</span>(Time),<span class="dt">fill=</span>Diet))<span class="op">+</span><span class="st"> </span><span class="co">#Note how the x, y and color changes</span></span>
<span id="cb146-2"><a href="visualization-and-statistical-test.html#cb146-2"></a><span class="st">  </span><span class="kw">geom_boxplot</span>()<span class="op">+</span><span class="st"> </span><span class="co">#this is adding the boxplot</span></span>
<span id="cb146-3"><a href="visualization-and-statistical-test.html#cb146-3"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>) </span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-40"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-40-1.png" alt="Boxplot" width="100%" />
<p class="caption">
Figure 3.11: Boxplot
</p>
</div>
<p>What happens if we do not use the <code>as.factor</code>? Again, a reminder that the data type is important!</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="visualization-and-statistical-test.html#cb147-1"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">y=</span>weight,<span class="dt">x=</span>Time,<span class="dt">fill=</span>Diet))<span class="op">+</span><span class="st"> </span></span>
<span id="cb147-2"><a href="visualization-and-statistical-test.html#cb147-2"></a><span class="st">  </span><span class="kw">geom_boxplot</span>()<span class="op">+</span><span class="st"> </span><span class="co">#this is adding the boxplot</span></span>
<span id="cb147-3"><a href="visualization-and-statistical-test.html#cb147-3"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-41"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-41-1.png" alt="Boxplot" width="100%" />
<p class="caption">
Figure 3.12: Boxplot
</p>
</div>
<p>It seems interesting to separate this by age (Time). This is achieved by another layer named <a href="https://plotly.com/ggplot2/facet_wrap/">facet</a>.</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="visualization-and-statistical-test.html#cb148-1"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">x=</span>weight,<span class="dt">fill=</span>Diet))<span class="op">+</span></span>
<span id="cb148-2"><a href="visualization-and-statistical-test.html#cb148-2"></a><span class="st">  </span><span class="kw">geom_density</span>( <span class="dt">alpha=</span><span class="fl">0.5</span>)<span class="op">+</span><span class="st"> </span></span>
<span id="cb148-3"><a href="visualization-and-statistical-test.html#cb148-3"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>)<span class="op">+</span></span>
<span id="cb148-4"><a href="visualization-and-statistical-test.html#cb148-4"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Time,<span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>) <span class="co">#This will separate the data into panels given the time, try looking for the meaning of the scale option</span></span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-421"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-42-1.png" alt="Plot separating by age of the chicken" width="100%" />
<p class="caption">
Figure 3.13: Plot separating by age of the chicken
</p>
</div>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="visualization-and-statistical-test.html#cb149-1"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">y=</span>weight,<span class="dt">x=</span><span class="kw">as.factor</span>(Time),<span class="dt">fill=</span>Diet))<span class="op">+</span></span>
<span id="cb149-2"><a href="visualization-and-statistical-test.html#cb149-2"></a><span class="st">  </span><span class="kw">geom_boxplot</span>()<span class="op">+</span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>)<span class="op">+</span></span>
<span id="cb149-3"><a href="visualization-and-statistical-test.html#cb149-3"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Time,<span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-422"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-42-2.png" alt="Plot separating by age of the chicken" width="100%" />
<p class="caption">
Figure 3.14: Plot separating by age of the chicken
</p>
</div>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="visualization-and-statistical-test.html#cb150-1"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">y=</span>weight,<span class="dt">x=</span><span class="kw">as.factor</span>(Time),<span class="dt">fill=</span>Diet))<span class="op">+</span></span>
<span id="cb150-2"><a href="visualization-and-statistical-test.html#cb150-2"></a><span class="st">  </span><span class="kw">geom_violin</span>()<span class="op">+</span></span>
<span id="cb150-3"><a href="visualization-and-statistical-test.html#cb150-3"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>)<span class="op">+</span></span>
<span id="cb150-4"><a href="visualization-and-statistical-test.html#cb150-4"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Time,<span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-423"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-42-3.png" alt="Plot separating by age of the chicken" width="100%" />
<p class="caption">
Figure 3.15: Plot separating by age of the chicken
</p>
</div>
</div>
<div id="saving-plots" class="section level3" number="3.3.4">
<h3><span class="header-section-number">3.3.4</span> Saving plots</h3>
<p>Imagine you want now to save some of these plots. You can use the button <strong>export</strong> in RStudio. But you can also use the <code>pdf</code> function.</p>
<p>This function allows us to determine the width and height of the plots. Check what happens if you modify the option in the plots below.</p>
<p>These pdf files will be saved on your working directory with the name, width and height determined in the function.</p>
<p>Important things:</p>
<ol style="list-style-type: decimal">
<li>Do not forget to put the “.pdf” at the end of the file name. What do you think it will happen if you forget it?</li>
<li>When you finish running the plots that you want to be in the pdf file, you have to run <code>dev.off()</code>. This will close the plot. If you forget this, you will not be able to open the plot.</li>
</ol>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="visualization-and-statistical-test.html#cb151-1"></a><span class="kw">pdf</span>(<span class="st">&quot;densityplot.pdf&quot;</span>,<span class="dt">width =</span> <span class="dv">20</span>, <span class="dt">height =</span> <span class="dv">20</span>) <span class="co">#save the plot as a pdf, control width and height of the pdf</span></span>
<span id="cb151-2"><a href="visualization-and-statistical-test.html#cb151-2"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">x=</span>weight,<span class="dt">fill=</span>Diet))<span class="op">+</span></span>
<span id="cb151-3"><a href="visualization-and-statistical-test.html#cb151-3"></a><span class="st">  </span><span class="kw">geom_density</span>( <span class="dt">alpha=</span><span class="fl">0.5</span>)<span class="op">+</span></span>
<span id="cb151-4"><a href="visualization-and-statistical-test.html#cb151-4"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>)<span class="op">+</span></span>
<span id="cb151-5"><a href="visualization-and-statistical-test.html#cb151-5"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Time,<span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>)</span>
<span id="cb151-6"><a href="visualization-and-statistical-test.html#cb151-6"></a></span>
<span id="cb151-7"><a href="visualization-and-statistical-test.html#cb151-7"></a><span class="kw">dev.off</span>() <span class="co">#end the plot</span></span>
<span id="cb151-8"><a href="visualization-and-statistical-test.html#cb151-8"></a></span>
<span id="cb151-9"><a href="visualization-and-statistical-test.html#cb151-9"></a></span>
<span id="cb151-10"><a href="visualization-and-statistical-test.html#cb151-10"></a><span class="kw">pdf</span>(<span class="st">&quot;density_and_violin.plot.pdf&quot;</span>,<span class="dt">width =</span> <span class="dv">20</span>, <span class="dt">height =</span> <span class="dv">20</span>) <span class="co">#save the plot as a pdf, control width and height of the pdf</span></span>
<span id="cb151-11"><a href="visualization-and-statistical-test.html#cb151-11"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">x=</span>weight,<span class="dt">fill=</span>Diet))<span class="op">+</span></span>
<span id="cb151-12"><a href="visualization-and-statistical-test.html#cb151-12"></a><span class="st">  </span><span class="kw">geom_density</span>( <span class="dt">alpha=</span><span class="fl">0.5</span>)<span class="op">+</span></span>
<span id="cb151-13"><a href="visualization-and-statistical-test.html#cb151-13"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>)<span class="op">+</span></span>
<span id="cb151-14"><a href="visualization-and-statistical-test.html#cb151-14"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Time,<span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>)</span>
<span id="cb151-15"><a href="visualization-and-statistical-test.html#cb151-15"></a></span>
<span id="cb151-16"><a href="visualization-and-statistical-test.html#cb151-16"></a><span class="kw">ggplot</span>(<span class="dt">data =</span> ChickWeight, <span class="kw">aes</span>(<span class="dt">y=</span>weight,<span class="dt">x=</span><span class="kw">as.factor</span>(Time),<span class="dt">fill=</span>Diet))<span class="op">+</span></span>
<span id="cb151-17"><a href="visualization-and-statistical-test.html#cb151-17"></a><span class="st">  </span><span class="kw">geom_violin</span>()<span class="op">+</span></span>
<span id="cb151-18"><a href="visualization-and-statistical-test.html#cb151-18"></a><span class="st">  </span><span class="kw">scale_colour_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Set1&quot;</span>)<span class="op">+</span></span>
<span id="cb151-19"><a href="visualization-and-statistical-test.html#cb151-19"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Time,<span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>)</span>
<span id="cb151-20"><a href="visualization-and-statistical-test.html#cb151-20"></a></span>
<span id="cb151-21"><a href="visualization-and-statistical-test.html#cb151-21"></a><span class="kw">dev.off</span>() <span class="co">#end the plot</span></span></code></pre></div>
<pre><code>## quartz_off_screen 
##                 2 
## quartz_off_screen 
##                 2</code></pre>
</div>
</div>
<div id="statistical-test" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> Statistical test</h2>
<div id="descriptive-statistics" class="section level3" number="3.4.1">
<h3><span class="header-section-number">3.4.1</span> Descriptive statistics</h3>
<p>We already saw a way to get the descriptive stats from a table by using <code>summary</code>.</p>
<p>We will try to compare the weight of chickens under different diets without considering the age.</p>
<p>First, we will now do a mean and SD table for each diet. There is one function that can do this for us.
<code>ddply</code> is a function that first divides the data by a variable written as .(Var) and then perform an specific function. With the indication of “transform” this will create a new column in out data</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="visualization-and-statistical-test.html#cb153-1"></a>stat_ChickWeight&lt;-<span class="kw">ddply</span>(ChickWeight,  .(Diet),  transform,  <span class="dt">Mean=</span><span class="kw">mean</span>(weight,<span class="dt">na.rm =</span> T), <span class="dt">SD=</span><span class="kw">sd</span>(weight,<span class="dt">na.rm =</span> T))</span>
<span id="cb153-2"><a href="visualization-and-statistical-test.html#cb153-2"></a></span>
<span id="cb153-3"><a href="visualization-and-statistical-test.html#cb153-3"></a><span class="kw">head</span>(stat_ChickWeight)</span></code></pre></div>
<pre><code>##   weight Time Chick Diet     Mean       SD
## 1     42    0     1    1 102.6455 56.65655
## 2     51    2     1    1 102.6455 56.65655
## 3     59    4     1    1 102.6455 56.65655
## 4     64    6     1    1 102.6455 56.65655
## 5     76    8     1    1 102.6455 56.65655
## 6     93   10     1    1 102.6455 56.65655</code></pre>
<blockquote>
<p>Is this what we wanted?
What happens if instead of “transform” we use “summarize”? Check <code>?ddply</code> for more detail.</p>
</blockquote>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="visualization-and-statistical-test.html#cb155-1"></a>statWeight_ChickWeight&lt;-<span class="kw">ddply</span>(ChickWeight,  .(Diet),  summarise,  <span class="dt">Mean=</span><span class="kw">mean</span>(weight,<span class="dt">na.rm =</span> T), <span class="dt">SD=</span><span class="kw">sd</span>(weight,<span class="dt">na.rm =</span> T))</span>
<span id="cb155-2"><a href="visualization-and-statistical-test.html#cb155-2"></a></span>
<span id="cb155-3"><a href="visualization-and-statistical-test.html#cb155-3"></a><span class="kw">head</span>(statWeight_ChickWeight)</span></code></pre></div>
<pre><code>##   Diet     Mean       SD
## 1    1 102.6455 56.65655
## 2    2 122.6167 71.60749
## 3    3 142.9500 86.54176
## 4    4 135.2627 68.82871</code></pre>
<p>This is usefull for ploting, here a good plot:</p>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="visualization-and-statistical-test.html#cb157-1"></a><span class="kw">ggplot</span>(statWeight_ChickWeight, <span class="kw">aes</span>(<span class="dt">x=</span>Diet, <span class="dt">y=</span>Mean, <span class="dt">fill=</span>Diet)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb157-2"><a href="visualization-and-statistical-test.html#cb157-2"></a><span class="st">   </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>, <span class="dt">position=</span><span class="kw">position_dodge</span>()) <span class="op">+</span></span>
<span id="cb157-3"><a href="visualization-and-statistical-test.html#cb157-3"></a><span class="st">  </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin=</span>Mean<span class="op">-</span>SD, <span class="dt">ymax=</span>Mean<span class="op">+</span>SD), <span class="dt">width=</span>.<span class="dv">2</span>,</span>
<span id="cb157-4"><a href="visualization-and-statistical-test.html#cb157-4"></a>                 <span class="dt">position=</span><span class="kw">position_dodge</span>(.<span class="dv">9</span>))</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-46"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-46-1.png" alt="Boxplot" width="100%" />
<p class="caption">
Figure 3.16: Boxplot
</p>
</div>
</div>
<div id="t-testwilcoxon" class="section level3" number="3.4.2">
<h3><span class="header-section-number">3.4.2</span> T-test/Wilcoxon</h3>
<p>To compare means we can do a T test but to do this we need to test the assumptions of this test: <strong>Normality</strong> of the data and Homoscedasticity (ie, the variance is similar between the two groups we want to compare)</p>
<p><em>Question</em> : Is there any significant difference in the weights between diet 1 and 3?
Preliminary test to check independent t-test assumptions</p>
<p><em>Assumption 1</em>: Are the two samples independents?
Yes, they are two different samples</p>
<p><em>Assumption 2</em>: Are the data from each of the 2 groups follow a normal distribution?</p>
<p>Shapiro-Wilk normality test for the different diets</p>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="visualization-and-statistical-test.html#cb158-1"></a><span class="kw">shapiro.test</span>(ChickWeight<span class="op">$</span>weight[ChickWeight<span class="op">$</span>Diet<span class="op">==</span><span class="dv">1</span>])</span></code></pre></div>
<pre><code>## 
##  Shapiro-Wilk normality test
## 
## data:  ChickWeight$weight[ChickWeight$Diet == 1]
## W = 0.89336, p-value = 2.211e-11</code></pre>
<p>The function with allows us to do a simpler writing</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="visualization-and-statistical-test.html#cb160-1"></a><span class="kw">with</span>(ChickWeight, <span class="kw">shapiro.test</span>(weight[Diet <span class="op">==</span><span class="st"> </span><span class="dv">1</span>]))</span></code></pre></div>
<pre><code>## 
##  Shapiro-Wilk normality test
## 
## data:  weight[Diet == 1]
## W = 0.89336, p-value = 2.211e-11</code></pre>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="visualization-and-statistical-test.html#cb162-1"></a><span class="kw">with</span>(ChickWeight, <span class="kw">shapiro.test</span>(weight[Diet <span class="op">==</span><span class="st"> </span><span class="dv">2</span>]))</span></code></pre></div>
<pre><code>## 
##  Shapiro-Wilk normality test
## 
## data:  weight[Diet == 2]
## W = 0.90399, p-value = 3.159e-07</code></pre>
<p>pvalue &lt; 0.05, these are not normally distributed. We can NOT use t-test here. If we remember the histograms, this makes sense.</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="visualization-and-statistical-test.html#cb164-1"></a><span class="kw">qplot</span>(<span class="dt">data=</span>ChickWeight, <span class="dt">x =</span> weight, <span class="dt">facets =</span> <span class="st">&quot;Diet&quot;</span>,<span class="dt">geom =</span> <span class="st">&quot;density&quot;</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-50"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-50-1.png" alt="**CAPTION THIS FIGURE!!**" width="100%" />
<p class="caption">
Figure 3.17: <strong>CAPTION THIS FIGURE!!</strong>
</p>
</div>
<p><em>Assumption 3</em>: Do the two populations have the same variances?</p>
<p>We’ll use F-test to test for homogeneity in variances. This is implemented by a function named <code>var.test</code>. This will require you to have which variable you want to test and separated by which variable.
This is clearly also not homoscedastic.</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="visualization-and-statistical-test.html#cb165-1"></a><span class="kw">var.test</span>(weight<span class="op">~</span><span class="st"> </span>Diet, <span class="dt">data =</span> ChickWeight[ChickWeight<span class="op">$</span>Diet <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>),])</span></code></pre></div>
<pre><code>## 
##  F test to compare two variances
## 
## data:  weight by Diet
## F = 0.62601, num df = 219, denom df = 119, p-value = 0.002928
## alternative hypothesis: true ratio of variances is not equal to 1
## 95 percent confidence interval:
##  0.4525703 0.8530014
## sample estimates:
## ratio of variances 
##           0.626013</code></pre>
<blockquote>
<p>What happens if you try to run this <code>var.test(weight~ Diet, data = ChickWeight)</code>?</p>
</blockquote>
<p>We will use then Wilcoxon.</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="visualization-and-statistical-test.html#cb167-1"></a><span class="kw">wilcox.test</span>(weight<span class="op">~</span><span class="st"> </span>Diet, <span class="dt">data =</span> ChickWeight[ChickWeight<span class="op">$</span>Diet <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>),],<span class="dt">exact =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<pre><code>## 
##  Wilcoxon rank sum test with continuity correction
## 
## data:  weight by Diet
## W = 11213, p-value = 0.02181
## alternative hypothesis: true location shift is not equal to 0</code></pre>
<blockquote>
<p>Are them different?</p>
</blockquote>
</div>
<div id="anovakruskalwallis" class="section level3" number="3.4.3">
<h3><span class="header-section-number">3.4.3</span> Anova/Kruskal–Wallis</h3>
<p>Another way to test differences is to do an <a href="https://en.wikipedia.org/wiki/Analysis_of_variance">ANOVA</a> or its non-parametric alternative <a href="https://en.wikipedia.org/wiki/Kruskal–Wallis_one-way_analysis_of_variance">Kruskal–Wallis</a>. We already know that this data cannot be analyzed using parametric test as anova. But let’s explore just for fun.</p>
<p>Let’s check all the diets together</p>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="visualization-and-statistical-test.html#cb169-1"></a><span class="co"># Compute the analysis of variance</span></span>
<span id="cb169-2"><a href="visualization-and-statistical-test.html#cb169-2"></a>res.aov &lt;-<span class="st"> </span><span class="kw">aov</span>(weight<span class="op">~</span><span class="st"> </span>Diet, <span class="dt">data =</span> ChickWeight)</span>
<span id="cb169-3"><a href="visualization-and-statistical-test.html#cb169-3"></a><span class="co"># Summary of the analysis</span></span>
<span id="cb169-4"><a href="visualization-and-statistical-test.html#cb169-4"></a><span class="kw">summary</span>(res.aov)</span></code></pre></div>
<pre><code>##              Df  Sum Sq Mean Sq F value   Pr(&gt;F)    
## Diet          3  155863   51954   10.81 6.43e-07 ***
## Residuals   574 2758693    4806                     
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>We see that the diet is a significant component of the variance of the data. Now we should know from where it is coming. We need to do a multiple pairwise-comparison. We will use Tukey Honest Significant Difference for this.</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="visualization-and-statistical-test.html#cb171-1"></a><span class="kw">TukeyHSD</span>(res.aov)</span></code></pre></div>
<pre><code>##   Tukey multiple comparisons of means
##     95% family-wise confidence level
## 
## Fit: aov(formula = weight ~ Diet, data = ChickWeight)
## 
## $Diet
##          diff         lwr      upr     p adj
## 2-1 19.971212  -0.2998092 40.24223 0.0552271
## 3-1 40.304545  20.0335241 60.57557 0.0000025
## 4-1 32.617257  12.2353820 52.99913 0.0002501
## 3-2 20.333333  -2.7268370 43.39350 0.1058474
## 4-2 12.646045 -10.5116315 35.80372 0.4954239
## 4-3 -7.687288 -30.8449649 15.47039 0.8277810</code></pre>
<p>However, you should be screaming at me now: I did not check the assumptions!</p>
<ol style="list-style-type: decimal">
<li><em>Homogeneity of variances</em>:</li>
</ol>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="visualization-and-statistical-test.html#cb173-1"></a><span class="kw">plot</span>(res.aov, <span class="dv">1</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-55"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-55-1.png" alt="**CAPTION THIS FIGURE!!**" width="100%" />
<p class="caption">
Figure 3.18: <strong>CAPTION THIS FIGURE!!</strong>
</p>
</div>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="visualization-and-statistical-test.html#cb174-1"></a><span class="kw">leveneTest</span>(weight<span class="op">~</span><span class="st"> </span>Diet, <span class="dt">data =</span> ChickWeight)</span></code></pre></div>
<pre><code>## Levene&#39;s Test for Homogeneity of Variance (center = median)
##        Df F value    Pr(&gt;F)    
## group   3  9.6001 3.418e-06 ***
##       574                      
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>We have just found what we already know. The variance is not homogeneous. The variance across groups is statistically significantly different.</p>
<ol start="2" style="list-style-type: decimal">
<li><em>Normality of the residuals</em>:</li>
</ol>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="visualization-and-statistical-test.html#cb176-1"></a><span class="kw">plot</span>(res.aov, <span class="dv">2</span>)</span></code></pre></div>
<div class="figure" style="text-align: center"><span id="fig:unnamed-chunk-57"></span>
<img src="02-introduction-to-R_files/figure-html/unnamed-chunk-57-1.png" alt="**CAPTION THIS FIGURE!!**" width="100%" />
<p class="caption">
Figure 3.19: <strong>CAPTION THIS FIGURE!!</strong>
</p>
</div>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="visualization-and-statistical-test.html#cb177-1"></a><span class="co"># Extract the residuals</span></span>
<span id="cb177-2"><a href="visualization-and-statistical-test.html#cb177-2"></a>aov_residuals &lt;-<span class="st"> </span><span class="kw">residuals</span>(<span class="dt">object =</span> res.aov )</span>
<span id="cb177-3"><a href="visualization-and-statistical-test.html#cb177-3"></a><span class="co"># Run Shapiro-Wilk test</span></span>
<span id="cb177-4"><a href="visualization-and-statistical-test.html#cb177-4"></a><span class="kw">shapiro.test</span>(<span class="dt">x =</span> aov_residuals )</span></code></pre></div>
<pre><code>## 
##  Shapiro-Wilk normality test
## 
## data:  aov_residuals
## W = 0.94065, p-value = 2.014e-14</code></pre>
<p>Non parametric then:</p>
<div class="sourceCode" id="cb179"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb179-1"><a href="visualization-and-statistical-test.html#cb179-1"></a><span class="kw">kruskal.test</span>(weight<span class="op">~</span><span class="st"> </span>Diet, <span class="dt">data =</span> ChickWeight)</span></code></pre></div>
<pre><code>## 
##  Kruskal-Wallis rank sum test
## 
## data:  weight by Diet
## Kruskal-Wallis chi-squared = 24.45, df = 3, p-value = 2.012e-05</code></pre>
<blockquote>
<p>What would you conclude about this data?</p>
</blockquote>
</div>
</div>
<div id="activity-2" class="section level2" number="3.5">
<h2><span class="header-section-number">3.5</span> Activity:</h2>
<p><strong>Find another thing you want to test with this data. Solve this in a graphical and statistical way. Save the plots</strong></p>
</div>
<div id="resources" class="section level2" number="3.6">
<h2><span class="header-section-number">3.6</span> Resources</h2>
<ul>
<li>Statistics with R: <a href="https://cran.r-project.org/doc/contrib/Seefeld_StatsRBio.pdf" class="uri">https://cran.r-project.org/doc/contrib/Seefeld_StatsRBio.pdf</a></li>
<li>Stat and plots with R: <a href="http://www.sthda.com/english/" class="uri">http://www.sthda.com/english/</a></li>
</ul>

</div>
</div>
<div class="psyteachr_footer">
  
</div>
<script>

/* update total correct if #total_correct exists */
update_total_correct = function() {
  if (t = document.getElementById("total_correct")) {
    t.innerHTML =
      document.getElementsByClassName("correct").length + " of " +
      document.getElementsByClassName("solveme").length + " correct";
  }
}

/* solution button toggling function */
b_func = function() {
  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "");
  }
  
  if (my_answer !== "" & real_answers.includes(my_answer)) {
    cl.add("correct");
  } else {
    cl.remove("correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol){
    var tol = JSON.parse(this.dataset.tol);  
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("correct");
    } else {
      cl.remove("correct");
    }  
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("correct");
    }  
  }
  
  update_total_correct();
}

window.onload = function() {
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('solution')) {
      buttons[i].onclick = b_func;
    }
  }
  
  /* set up solveme inputs */
  var solveme = document.getElementsByClassName("solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off"); 
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";
    
    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;
    
    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;
  }
  
  update_total_correct();
}

</script>
            </section>

          </div>
        </div>
      </div>
<a href="data-manipulation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="sequencing-techniques-and-preprocessing.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"],
"google": false
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": {},
"toc": {
"collapse": "section",
"scroll_highlight": true
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
